// renovate/dist/datasource/datasource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Datasource = void 0;
const external_host_error_1 = require("../types/errors/external-host-error");
const http_1 = require("../util/http");
class Datasource {
    constructor(id) {
        this.id = id;
        this.customRegistrySupport = true;
        this.registryStrategy = 'first';
        this.http = new http_1.Http(id);
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    handleSpecificErrors(err) { }
    handleGenericErrors(err) {
        var _a, _b, _c, _d;
        // istanbul ignore if: not easy testable with nock
        if (err instanceof external_host_error_1.ExternalHostError) {
            throw err;
        }
        this.handleSpecificErrors(err);
        if (((_a = err.response) === null || _a === void 0 ? void 0 : _a.statusCode) !== undefined) {
            if (((_b = err.response) === null || _b === void 0 ? void 0 : _b.statusCode) === 429 ||
                (((_c = err.response) === null || _c === void 0 ? void 0 : _c.statusCode) >= 500 && ((_d = err.response) === null || _d === void 0 ? void 0 : _d.statusCode) < 600)) {
                throw new external_host_error_1.ExternalHostError(err);
            }
        }
        throw err;
    }
}
exports.Datasource = Datasource;
