// renovate/dist/manager/pip_setup/extract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = void 0;
const specifier_1 = require("@renovate/pep440/lib/specifier");
const parser_utils_1 = require("@renovatebot/parser-utils");
const pypi_1 = require("../../datasource/pypi");
const types_1 = require("../../types");
const regex_1 = require("../../util/regex");
const python = parser_utils_1.lang.createLang('python');
// Optimize regex memory usage when we don't need named groups
function cleanupNamedGroups(regexSource) {
    return regexSource.replace(/\(\?<\w+>/g, '(?:');
}
const rangePattern = cleanupNamedGroups(specifier_1.RANGE_PATTERN);
const versionPattern = `(?:${rangePattern}(?:\\s*,\\s*${rangePattern})*)`;
const depNamePattern = '(?:[a-zA-Z][-_a-zA-Z0-9]*[a-zA-Z0-9])';
const depPattern = [
    '^',
    `(?<depName>${depNamePattern})`,
    `(?<extra>(?:\\[\\s*(?:${depNamePattern}(?:\\s*,\\s*${depNamePattern})*\\s*)\\])?)`,
    `(?<currentValue>${versionPattern})`,
].join('\\s*');
const extractRegex = (0, regex_1.regEx)(depPattern);
// Extract dependency string
function depStringHandler(ctx, token) {
    const depStr = token.value;
    const match = extractRegex.exec(depStr);
    const { depName, currentValue } = match.groups;
    const dep = {
        depName,
        currentValue,
        managerData: {
            lineNumber: token.line - 1,
        },
        datasource: pypi_1.PypiDatasource.id,
    };
    return { ...ctx, deps: [...ctx.deps, dep] };
}
// Add `skip-reason` for dependencies annotated
// with "# renovate: ignore" comment
function depSkipHandler(ctx) {
    const dep = ctx.deps[ctx.deps.length - 1];
    const deps = ctx.deps.slice(0, -1);
    deps.push({ ...dep, skipReason: types_1.SkipReason.Ignored });
    return { ...ctx, deps };
}
const incompleteDepString = parser_utils_1.query
    .str(new RegExp(cleanupNamedGroups(depPattern)))
    .op(/^\+|\*$/);
const depString = parser_utils_1.query
    .str(new RegExp(cleanupNamedGroups(depPattern)), depStringHandler)
    .opt(parser_utils_1.query
    .opt(parser_utils_1.query.op(','))
    .comment(/^#\s*renovate\s*:\s*ignore\s*$/, depSkipHandler));
const query = parser_utils_1.query.alt(incompleteDepString, depString);
function extractPackageFile(content, _packageFile, _config) {
    var _a;
    const res = python.query(content, query, { deps: [] });
    return ((_a = res === null || res === void 0 ? void 0 : res.deps) === null || _a === void 0 ? void 0 : _a.length) ? res : null;
}
exports.extractPackageFile = extractPackageFile;
