// renovate/dist/datasource/adoptium-java/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdoptiumJavaDatasource = void 0;
const tslib_1 = require("tslib");
const logger_1 = require("../../logger");
const external_host_error_1 = require("../../types/errors/external-host-error");
const decorator_1 = require("../../util/cache/package/decorator");
const types_1 = require("../../util/http/types");
const datasource_1 = require("../datasource");
const common_1 = require("./common");
class AdoptiumJavaDatasource extends datasource_1.Datasource {
    constructor() {
        super(common_1.datasource);
        this.customRegistrySupport = false;
        this.defaultRegistryUrls = [common_1.defaultRegistryUrl];
        this.caching = true;
    }
    async getReleases({ registryUrl, lookupName, }) {
        var _a, _b;
        let page = 0;
        const imageType = (0, common_1.getImageType)(lookupName);
        logger_1.logger.trace({ registryUrl, lookupName, imageType }, 'fetching java release');
        const url = `${registryUrl}v3/info/release_versions?page_size=${common_1.pageSize}&image_type=${imageType}&project=jdk&release_type=ga&sort_method=DATE&sort_order=DESC&vendor=adoptium`;
        const result = {
            homepage: 'https://adoptium.net',
            releases: [],
        };
        let resp;
        try {
            do {
                resp = (await this.http.getJson(`${url}&page=${page}`)).body;
                result.releases.push(...resp.versions.map(({ semver }) => ({ version: semver })));
                page += 1;
            } while (page < 50 && resp.versions.length === common_1.pageSize);
        }
        catch (err) {
            // istanbul ignore else: not testable with nock
            if (err instanceof types_1.HttpError) {
                if (((_a = err.response) === null || _a === void 0 ? void 0 : _a.statusCode) === 404 && page > 0) {
                    // no more pages
                    return result;
                }
                if (((_b = err.response) === null || _b === void 0 ? void 0 : _b.statusCode) !== 404) {
                    throw new external_host_error_1.ExternalHostError(err);
                }
            }
            this.handleGenericErrors(err);
        }
        return result.releases.length ? result : null;
    }
}
AdoptiumJavaDatasource.id = common_1.datasource;
(0, tslib_1.__decorate)([
    (0, decorator_1.cache)({
        namespace: `datasource-${common_1.datasource}`,
        key: ({ registryUrl, lookupName }) => `${registryUrl}:${(0, common_1.getImageType)(lookupName)}`,
    })
], AdoptiumJavaDatasource.prototype, "getReleases", null);
exports.AdoptiumJavaDatasource = AdoptiumJavaDatasource;
