// renovate/dist/versioning/docker/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.api = exports.supportsRanges = exports.urls = exports.displayName = exports.id = void 0;
const tslib_1 = require("tslib");
const regex_1 = require("../../util/regex");
const generic = (0, tslib_1.__importStar)(require("../loose/generic"));
exports.id = 'docker';
exports.displayName = 'Docker';
exports.urls = [
    'https://docs.docker.com/engine/reference/commandline/tag/',
];
exports.supportsRanges = false;
const versionPattern = (0, regex_1.regEx)(/^(?<version>\d+(?:\.\d+)*)(?<prerelease>.*)$/);
const commitHashPattern = (0, regex_1.regEx)(/^[a-f0-9]{7,40}$/);
const numericPattern = (0, regex_1.regEx)(/^[0-9]+$/);
function parse(version) {
    if (commitHashPattern.test(version) && !numericPattern.test(version)) {
        return null;
    }
    const versionPieces = version.replace((0, regex_1.regEx)(/^v/), '').split('-');
    const prefix = versionPieces.shift();
    const suffix = versionPieces.join('-');
    const m = versionPattern.exec(prefix);
    if (!(m === null || m === void 0 ? void 0 : m.groups)) {
        return null;
    }
    const { version: ver, prerelease } = m.groups;
    const release = ver.split('.').map(Number);
    return { release, suffix, prerelease };
}
function valueToVersion(value) {
    // Remove any suffix after '-', e.g. '-alpine'
    return value ? value.split('-')[0] : value;
}
function compare(version1, version2) {
    const parsed1 = parse(version1);
    const parsed2 = parse(version2);
    // istanbul ignore if
    if (!(parsed1 && parsed2)) {
        return 1;
    }
    const length = Math.max(parsed1.release.length, parsed2.release.length);
    for (let i = 0; i < length; i += 1) {
        const part1 = parsed1.release[i];
        const part2 = parsed2.release[i];
        // shorter is bigger 2.1 > 2.1.1
        if (part1 === undefined) {
            return 1;
        }
        if (part2 === undefined) {
            return -1;
        }
        if (part1 !== part2) {
            return part1 - part2;
        }
    }
    if (parsed1.prerelease !== parsed2.prerelease) {
        // unstable is lower
        if (!parsed1.prerelease && parsed2.prerelease) {
            return 1;
        }
        if (parsed1.prerelease && !parsed2.prerelease) {
            return -1;
        }
        // alphabetic order
        return parsed1.prerelease.localeCompare(parsed2.prerelease);
    }
    // equals
    return parsed2.suffix.localeCompare(parsed1.suffix);
}
function isCompatible(version, range) {
    const parsed1 = parse(version);
    const parsed2 = parse(range);
    return (parsed1.suffix === parsed2.suffix &&
        parsed1.release.length === parsed2.release.length);
}
exports.api = {
    ...generic.create({
        parse,
        compare,
    }),
    isCompatible,
    valueToVersion,
};
exports.default = exports.api;
