// renovate/dist/manager/circleci/extract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = void 0;
const tslib_1 = require("tslib");
const orb_1 = require("../../datasource/orb");
const logger_1 = require("../../logger");
const regex_1 = require("../../util/regex");
const npmVersioning = (0, tslib_1.__importStar)(require("../../versioning/npm"));
const extract_1 = require("../dockerfile/extract");
function extractPackageFile(content) {
    var _a, _b;
    const deps = [];
    try {
        const lines = content.split('\n');
        for (let lineNumber = 0; lineNumber < lines.length; lineNumber += 1) {
            const line = lines[lineNumber];
            const orbs = (0, regex_1.regEx)(/^\s*orbs:\s*$/).exec(line); // TODO #12071
            if (orbs) {
                logger_1.logger.trace(`Matched orbs on line ${lineNumber}`);
                let foundOrbOrNoop;
                do {
                    foundOrbOrNoop = false;
                    const orbLine = lines[lineNumber + 1];
                    logger_1.logger.trace(`orbLine: "${orbLine}"`);
                    const yamlNoop = (0, regex_1.regEx)(/^\s*(#|$)/).exec(orbLine); // TODO #12071
                    if (yamlNoop) {
                        logger_1.logger.debug('orbNoop');
                        foundOrbOrNoop = true;
                        lineNumber += 1;
                        continue;
                    }
                    const orbMatch = (0, regex_1.regEx)(/^\s+([^:]+):\s(.+)$/).exec(orbLine); // TODO #12071
                    if (orbMatch) {
                        logger_1.logger.trace('orbMatch');
                        foundOrbOrNoop = true;
                        lineNumber += 1;
                        const depName = orbMatch[1];
                        const [orbName, currentValue] = orbMatch[2].split('@');
                        const dep = {
                            depType: 'orb',
                            depName,
                            currentValue,
                            datasource: orb_1.OrbDatasource.id,
                            lookupName: orbName,
                            commitMessageTopic: '{{{depName}}} orb',
                            versioning: npmVersioning.id,
                            rangeStrategy: 'pin',
                        };
                        deps.push(dep);
                    }
                } while (foundOrbOrNoop);
            }
            const match = (0, regex_1.regEx)(/^\s*-? image:\s*'?"?([^\s'"]+)'?"?\s*$/).exec(line); // TODO #12071
            if (match) {
                const currentFrom = match[1];
                const dep = (0, extract_1.getDep)(currentFrom);
                logger_1.logger.debug({
                    depName: dep.depName,
                    currentValue: dep.currentValue,
                    currentDigest: dep.currentDigest,
                }, 'CircleCI docker image');
                dep.depType = 'docker';
                dep.versioning = 'docker';
                if (!((_a = dep.depName) === null || _a === void 0 ? void 0 : _a.startsWith('ubuntu-')) &&
                    !((_b = dep.depName) === null || _b === void 0 ? void 0 : _b.startsWith('windows-server-'))) {
                    deps.push(dep);
                }
            }
        }
    }
    catch (err) /* istanbul ignore next */ {
        logger_1.logger.warn({ err }, 'Error extracting circleci images');
    }
    if (!deps.length) {
        return null;
    }
    return { deps };
}
exports.extractPackageFile = extractPackageFile;
