// renovate/dist/manager/helmv3/artifacts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateArtifacts = void 0;
const shlex_1 = require("shlex");
const error_messages_1 = require("../../constants/error-messages");
const logger_1 = require("../../logger");
const exec_1 = require("../../util/exec");
const fs_1 = require("../../util/fs");
async function helmCommands(manifestPath, aliases) {
    const execOptions = {
        docker: {
            image: 'helm',
        },
        extraEnv: {
            HELM_EXPERIMENTAL_OCI: '1',
        },
    };
    const cmd = [];
    if (aliases) {
        Object.entries(aliases).forEach(([alias, url]) => cmd.push(`helm repo add ${(0, shlex_1.quote)(alias)} ${(0, shlex_1.quote)(url)}`));
    }
    cmd.push(`helm dependency update ${(0, shlex_1.quote)((0, fs_1.getSubDirectory)(manifestPath))}`);
    await (0, exec_1.exec)(cmd, execOptions);
}
async function updateArtifacts({ packageFileName, updatedDeps, newPackageFileContent, config, }) {
    logger_1.logger.debug(`helmv3.updateArtifacts(${packageFileName})`);
    const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
    if (!isLockFileMaintenance &&
        (updatedDeps === undefined || updatedDeps.length < 1)) {
        logger_1.logger.debug('No updated helmv3 deps - returning null');
        return null;
    }
    const lockFileName = (0, fs_1.getSiblingFileName)(packageFileName, 'Chart.lock');
    const existingLockFileContent = await (0, fs_1.readLocalFile)(lockFileName);
    if (!existingLockFileContent) {
        logger_1.logger.debug('No Chart.lock found');
        return null;
    }
    try {
        await (0, fs_1.writeLocalFile)(packageFileName, newPackageFileContent);
        logger_1.logger.debug('Updating ' + lockFileName);
        await helmCommands(packageFileName, config.aliases);
        logger_1.logger.debug('Returning updated Chart.lock');
        const newHelmLockContent = await (0, fs_1.readLocalFile)(lockFileName);
        if (existingLockFileContent === newHelmLockContent) {
            logger_1.logger.debug('Chart.lock is unchanged');
            return null;
        }
        return [
            {
                file: {
                    name: lockFileName,
                    contents: newHelmLockContent,
                },
            },
        ];
    }
    catch (err) {
        // istanbul ignore if
        if (err.message === error_messages_1.TEMPORARY_ERROR) {
            throw err;
        }
        logger_1.logger.debug({ err }, 'Failed to update Helm lock file');
        return [
            {
                artifactError: {
                    lockFile: lockFileName,
                    stderr: err.message,
                },
            },
        ];
    }
}
exports.updateArtifacts = updateArtifacts;
