// renovate/dist/versioning/ubuntu/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.api = exports.supportsRanges = exports.urls = exports.displayName = exports.id = void 0;
const regex_1 = require("../../util/regex");
exports.id = 'ubuntu';
exports.displayName = 'Ubuntu';
exports.urls = ['https://changelogs.ubuntu.com/meta-release'];
exports.supportsRanges = false;
// #12509
const temporarilyUnstable = ['22.04'];
// validation
function isValid(input) {
    return (typeof input === 'string' &&
        (0, regex_1.regEx)(/^(0[4-5]|[6-9]|[1-9][0-9])\.[0-9][0-9](\.[0-9]{1,2})?$/).test(input));
}
function isVersion(input) {
    return isValid(input);
}
function isCompatible(version, _range) {
    return isValid(version);
}
function isSingleVersion(version) {
    return isValid(version) ? true : null;
}
function isStable(version) {
    if (!isValid(version)) {
        return false;
    }
    if (temporarilyUnstable.includes(version)) {
        return false;
    }
    return (0, regex_1.regEx)(/^\d?[02468]\.04/).test(version);
}
// digestion of version
function getMajor(version) {
    if (isValid(version)) {
        const [major] = version.split('.') || [];
        return parseInt(major, 10);
    }
    return null;
}
function getMinor(version) {
    if (isValid(version)) {
        const [, minor] = version.split('.') || [];
        return parseInt(minor, 10);
    }
    return null;
}
function getPatch(version) {
    if (isValid(version)) {
        const [, , patch] = version.split('.') || [];
        return patch ? parseInt(patch, 10) : null;
    }
    return null;
}
// comparison
function equals(version, other) {
    return !!isVersion(version) && !!isVersion(other) && version === other;
}
function isGreaterThan(version, other) {
    var _a, _b, _c, _d, _e, _f;
    const xMajor = (_a = getMajor(version)) !== null && _a !== void 0 ? _a : 0;
    const yMajor = (_b = getMajor(other)) !== null && _b !== void 0 ? _b : 0;
    if (xMajor > yMajor) {
        return true;
    }
    if (xMajor < yMajor) {
        return false;
    }
    const xMinor = (_c = getMinor(version)) !== null && _c !== void 0 ? _c : 0;
    const yMinor = (_d = getMinor(other)) !== null && _d !== void 0 ? _d : 0;
    if (xMinor > yMinor) {
        return true;
    }
    if (xMinor < yMinor) {
        return false;
    }
    const xPatch = (_e = getPatch(version)) !== null && _e !== void 0 ? _e : 0;
    const yPatch = (_f = getPatch(other)) !== null && _f !== void 0 ? _f : 0;
    return xPatch > yPatch;
}
function getSatisfyingVersion(versions, range) {
    return versions.find((version) => equals(version, range)) ? range : null;
}
function minSatisfyingVersion(versions, range) {
    return getSatisfyingVersion(versions, range);
}
function getNewValue(newValueConfig) {
    return newValueConfig.newVersion;
}
function sortVersions(version, other) {
    if (equals(version, other)) {
        return 0;
    }
    if (isGreaterThan(version, other)) {
        return 1;
    }
    return -1;
}
function matches(version, range) {
    return equals(version, range);
}
exports.api = {
    isCompatible,
    isSingleVersion,
    isStable,
    isValid,
    isVersion,
    getMajor,
    getMinor,
    getPatch,
    equals,
    isGreaterThan,
    getSatisfyingVersion,
    minSatisfyingVersion,
    getNewValue,
    sortVersions,
    matches,
};
exports.default = exports.api;
