// renovate/dist/workers/repository/model/commit-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommitMessage = void 0;
class CommitMessage {
    constructor(message = '') {
        this.message = '';
        this.prefix = '';
        this.setMessage(message);
    }
    static formatPrefix(prefix) {
        if (!prefix) {
            return '';
        }
        if (prefix.endsWith(CommitMessage.SEPARATOR)) {
            return prefix;
        }
        return `${prefix}${CommitMessage.SEPARATOR}`;
    }
    setMessage(message) {
        this.message = (message || '').trim();
    }
    setCustomPrefix(prefix) {
        this.prefix = (prefix !== null && prefix !== void 0 ? prefix : '').trim();
    }
    setSemanticPrefix(type, scope) {
        this.prefix = (type !== null && type !== void 0 ? type : '').trim();
        if (scope === null || scope === void 0 ? void 0 : scope.trim()) {
            this.prefix += `(${scope.trim()})`;
        }
    }
    toString() {
        const prefix = CommitMessage.formatPrefix(this.prefix);
        const message = this.formatMessage();
        return [prefix, message].join(' ').trim();
    }
    formatMessage() {
        if (this.prefix) {
            return this.message;
        }
        return this.message.charAt(0).toUpperCase() + this.message.slice(1);
    }
}
exports.CommitMessage = CommitMessage;
CommitMessage.SEPARATOR = ':';
