// renovate/dist/datasource/npm/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customRegistrySupport = exports.defaultVersioning = exports.id = exports.setNpmrc = exports.getNpmrc = exports.getReleases = exports.resetCache = exports.resetMemCache = void 0;
const tslib_1 = require("tslib");
const npmVersioning = (0, tslib_1.__importStar)(require("../../versioning/npm"));
var get_1 = require("./get");
Object.defineProperty(exports, "resetMemCache", { enumerable: true, get: function () { return get_1.resetMemCache; } });
Object.defineProperty(exports, "resetCache", { enumerable: true, get: function () { return get_1.resetCache; } });
var releases_1 = require("./releases");
Object.defineProperty(exports, "getReleases", { enumerable: true, get: function () { return releases_1.getReleases; } });
var npmrc_1 = require("./npmrc");
Object.defineProperty(exports, "getNpmrc", { enumerable: true, get: function () { return npmrc_1.getNpmrc; } });
Object.defineProperty(exports, "setNpmrc", { enumerable: true, get: function () { return npmrc_1.setNpmrc; } });
var common_1 = require("./common");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return common_1.id; } });
exports.defaultVersioning = npmVersioning.id;
exports.customRegistrySupport = false;
