// renovate/dist/config/presets/internal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPreset = exports.groups = void 0;
const tslib_1 = require("tslib");
const compatibilityPreset = (0, tslib_1.__importStar)(require("./compatibility"));
const configPreset = (0, tslib_1.__importStar)(require("./config"));
const defaultPreset = (0, tslib_1.__importStar)(require("./default"));
const dockerPreset = (0, tslib_1.__importStar)(require("./docker"));
const groupPreset = (0, tslib_1.__importStar)(require("./group"));
const helpersPreset = (0, tslib_1.__importStar)(require("./helpers"));
const monorepoPreset = (0, tslib_1.__importStar)(require("./monorepo"));
const npm = (0, tslib_1.__importStar)(require("./npm"));
const packagesPreset = (0, tslib_1.__importStar)(require("./packages"));
const previewPreset = (0, tslib_1.__importStar)(require("./preview"));
const regexManagersPreset = (0, tslib_1.__importStar)(require("./regex-managers"));
const replacements = (0, tslib_1.__importStar)(require("./replacements"));
const schedulePreset = (0, tslib_1.__importStar)(require("./schedule"));
const workaroundsPreset = (0, tslib_1.__importStar)(require("./workarounds"));
exports.groups = {
    compatibility: compatibilityPreset.presets,
    config: configPreset.presets,
    default: defaultPreset.presets,
    docker: dockerPreset.presets,
    group: groupPreset.presets,
    helpers: helpersPreset.presets,
    monorepo: monorepoPreset.presets,
    npm: npm.presets,
    packages: packagesPreset.presets,
    preview: previewPreset.presets,
    regexManagers: regexManagersPreset.presets,
    replacements: replacements.presets,
    schedule: schedulePreset.presets,
    workarounds: workaroundsPreset.presets,
};
function getPreset({ packageName: pkgName, presetName, }) {
    return exports.groups[pkgName]
        ? exports.groups[pkgName][presetName]
        : /* istanbul ignore next */ undefined;
}
exports.getPreset = getPreset;
