// @aws-sdk/client-ec2/dist-cjs/models/models_4.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisableSerialConsoleAccessResult = exports.DisableSerialConsoleAccessRequest = exports.DisableImageDeprecationResult = exports.DisableImageDeprecationRequest = exports.DisableFastSnapshotRestoresResult = exports.DisableFastSnapshotRestoreErrorItem = exports.DisableFastSnapshotRestoreStateErrorItem = exports.DisableFastSnapshotRestoreStateError = exports.DisableFastSnapshotRestoreSuccessItem = exports.DisableFastSnapshotRestoresRequest = exports.DisableEbsEncryptionByDefaultResult = exports.DisableEbsEncryptionByDefaultRequest = exports.DetachVpnGatewayRequest = exports.DetachVolumeRequest = exports.DetachNetworkInterfaceRequest = exports.DetachInternetGatewayRequest = exports.DetachClassicLinkVpcResult = exports.DetachClassicLinkVpcRequest = exports.DescribeVpnGatewaysResult = exports.DescribeVpnGatewaysRequest = exports.DescribeVpnConnectionsResult = exports.DescribeVpnConnectionsRequest = exports.DescribeVpcsResult = exports.DescribeVpcsRequest = exports.DescribeVpcPeeringConnectionsResult = exports.DescribeVpcPeeringConnectionsRequest = exports.DescribeVpcEndpointServicesResult = exports.ServiceDetail = exports.PrivateDnsDetails = exports.DescribeVpcEndpointServicesRequest = exports.DescribeVpcEndpointServicePermissionsResult = exports.DescribeVpcEndpointServicePermissionsRequest = exports.DescribeVpcEndpointServiceConfigurationsResult = exports.DescribeVpcEndpointServiceConfigurationsRequest = exports.DescribeVpcEndpointsResult = exports.DescribeVpcEndpointsRequest = exports.DescribeVpcEndpointConnectionsResult = exports.VpcEndpointConnection = exports.DescribeVpcEndpointConnectionsRequest = exports.DescribeVpcEndpointConnectionNotificationsResult = exports.DescribeVpcEndpointConnectionNotificationsRequest = exports.DescribeVpcClassicLinkDnsSupportResult = exports.ClassicLinkDnsSupport = exports.DescribeVpcClassicLinkDnsSupportRequest = exports.DescribeVpcClassicLinkResult = exports.VpcClassicLink = exports.DescribeVpcClassicLinkRequest = exports.DescribeVpcAttributeResult = exports.DescribeVpcAttributeRequest = exports.DescribeVolumeStatusResult = void 0;
exports.ExportClientVpnClientCertificateRevocationListRequest = exports.EnableVpcClassicLinkDnsSupportResult = exports.EnableVpcClassicLinkDnsSupportRequest = exports.EnableVpcClassicLinkResult = exports.EnableVpcClassicLinkRequest = exports.EnableVolumeIORequest = exports.EnableVgwRoutePropagationRequest = exports.EnableTransitGatewayRouteTablePropagationResult = exports.EnableTransitGatewayRouteTablePropagationRequest = exports.EnableSerialConsoleAccessResult = exports.EnableSerialConsoleAccessRequest = exports.EnableImageDeprecationResult = exports.EnableImageDeprecationRequest = exports.EnableFastSnapshotRestoresResult = exports.EnableFastSnapshotRestoreErrorItem = exports.EnableFastSnapshotRestoreStateErrorItem = exports.EnableFastSnapshotRestoreStateError = exports.EnableFastSnapshotRestoreSuccessItem = exports.EnableFastSnapshotRestoresRequest = exports.EnableEbsEncryptionByDefaultResult = exports.EnableEbsEncryptionByDefaultRequest = exports.DisassociateVpcCidrBlockResult = exports.DisassociateVpcCidrBlockRequest = exports.DisassociateTrunkInterfaceResult = exports.DisassociateTrunkInterfaceRequest = exports.DisassociateTransitGatewayRouteTableResult = exports.DisassociateTransitGatewayRouteTableRequest = exports.DisassociateTransitGatewayMulticastDomainResult = exports.DisassociateTransitGatewayMulticastDomainRequest = exports.DisassociateSubnetCidrBlockResult = exports.DisassociateSubnetCidrBlockRequest = exports.DisassociateRouteTableRequest = exports.DisassociateInstanceEventWindowResult = exports.DisassociateInstanceEventWindowRequest = exports.InstanceEventWindowDisassociationRequest = exports.DisassociateIamInstanceProfileResult = exports.DisassociateIamInstanceProfileRequest = exports.DisassociateEnclaveCertificateIamRoleResult = exports.DisassociateEnclaveCertificateIamRoleRequest = exports.DisassociateClientVpnTargetNetworkResult = exports.DisassociateClientVpnTargetNetworkRequest = exports.DisassociateAddressRequest = exports.DisableVpcClassicLinkDnsSupportResult = exports.DisableVpcClassicLinkDnsSupportRequest = exports.DisableVpcClassicLinkResult = exports.DisableVpcClassicLinkRequest = exports.DisableVgwRoutePropagationRequest = exports.DisableTransitGatewayRouteTablePropagationResult = exports.TransitGatewayPropagation = exports.DisableTransitGatewayRouteTablePropagationRequest = void 0;
exports.PrefixListEntry = exports.GetManagedPrefixListEntriesRequest = exports.GetManagedPrefixListAssociationsResult = exports.PrefixListAssociation = exports.GetManagedPrefixListAssociationsRequest = exports.GetLaunchTemplateDataResult = exports.GetLaunchTemplateDataRequest = exports.GetHostReservationPurchasePreviewResult = exports.Purchase = exports.GetHostReservationPurchasePreviewRequest = exports.GetGroupsForCapacityReservationResult = exports.CapacityReservationGroup = exports.GetGroupsForCapacityReservationRequest = exports.GetFlowLogsIntegrationTemplateResult = exports.GetFlowLogsIntegrationTemplateRequest = exports.IntegrateServices = exports.AthenaIntegration = exports.PartitionLoadFrequency = exports.GetEbsEncryptionByDefaultResult = exports.GetEbsEncryptionByDefaultRequest = exports.GetEbsDefaultKmsKeyIdResult = exports.GetEbsDefaultKmsKeyIdRequest = exports.GetDefaultCreditSpecificationResult = exports.InstanceFamilyCreditSpecification = exports.GetDefaultCreditSpecificationRequest = exports.GetConsoleScreenshotResult = exports.GetConsoleScreenshotRequest = exports.GetConsoleOutputResult = exports.GetConsoleOutputRequest = exports.GetCoipPoolUsageResult = exports.CoipAddressUsage = exports.GetCoipPoolUsageRequest = exports.GetCapacityReservationUsageResult = exports.InstanceUsage = exports.GetCapacityReservationUsageRequest = exports.GetAssociatedIpv6PoolCidrsResult = exports.Ipv6CidrAssociation = exports.GetAssociatedIpv6PoolCidrsRequest = exports.GetAssociatedEnclaveCertificateIamRolesResult = exports.AssociatedRole = exports.GetAssociatedEnclaveCertificateIamRolesRequest = exports.ExportTransitGatewayRoutesResult = exports.ExportTransitGatewayRoutesRequest = exports.ExportImageResult = exports.ExportImageRequest = exports.ExportTaskS3LocationRequest = exports.ExportClientVpnClientConfigurationResult = exports.ExportClientVpnClientConfigurationRequest = exports.ExportClientVpnClientCertificateRevocationListResult = exports.ClientCertificateRevocationListStatus = void 0;
exports.SnapshotDiskContainer = exports.ImportKeyPairResult = exports.ImportKeyPairRequest = exports.ImportInstanceResult = exports.ImportInstanceRequest = exports.ImportInstanceLaunchSpecification = exports.UserData = exports.DiskImage = exports.VolumeDetail = exports.DiskImageDetail = exports.ImportImageResult = exports.ImportImageRequest = exports.ImportImageLicenseConfigurationRequest = exports.ImageDiskContainer = exports.UserBucket = exports.ClientData = exports.ImportClientVpnClientCertificateRevocationListResult = exports.ImportClientVpnClientCertificateRevocationListRequest = exports.GetVpnConnectionDeviceTypesResult = exports.VpnConnectionDeviceType = exports.GetVpnConnectionDeviceTypesRequest = exports.GetVpnConnectionDeviceSampleConfigurationResult = exports.GetVpnConnectionDeviceSampleConfigurationRequest = exports.GetTransitGatewayRouteTablePropagationsResult = exports.TransitGatewayRouteTablePropagation = exports.GetTransitGatewayRouteTablePropagationsRequest = exports.GetTransitGatewayRouteTableAssociationsResult = exports.TransitGatewayRouteTableAssociation = exports.GetTransitGatewayRouteTableAssociationsRequest = exports.GetTransitGatewayPrefixListReferencesResult = exports.GetTransitGatewayPrefixListReferencesRequest = exports.GetTransitGatewayMulticastDomainAssociationsResult = exports.TransitGatewayMulticastDomainAssociation = exports.GetTransitGatewayMulticastDomainAssociationsRequest = exports.GetTransitGatewayAttachmentPropagationsResult = exports.TransitGatewayAttachmentPropagation = exports.GetTransitGatewayAttachmentPropagationsRequest = exports.GetSubnetCidrReservationsResult = exports.GetSubnetCidrReservationsRequest = exports.GetSerialConsoleAccessStatusResult = exports.GetSerialConsoleAccessStatusRequest = exports.GetReservedInstancesExchangeQuoteResult = exports.TargetReservationValue = exports.TargetConfiguration = exports.ReservedInstanceReservationValue = exports.ReservationValue = exports.GetReservedInstancesExchangeQuoteRequest = exports.GetPasswordDataResult = exports.GetPasswordDataRequest = exports.GetManagedPrefixListEntriesResult = void 0;
exports.ModifyInstancePlacementRequest = exports.ModifyInstanceMetadataOptionsResult = exports.ModifyInstanceMetadataOptionsRequest = exports.ModifyInstanceEventWindowResult = exports.ModifyInstanceEventWindowRequest = exports.ModifyInstanceEventStartTimeResult = exports.ModifyInstanceEventStartTimeRequest = exports.ModifyInstanceCreditSpecificationResult = exports.UnsuccessfulInstanceCreditSpecificationItem = exports.UnsuccessfulInstanceCreditSpecificationItemError = exports.UnsuccessfulInstanceCreditSpecificationErrorCode = exports.SuccessfulInstanceCreditSpecificationItem = exports.ModifyInstanceCreditSpecificationRequest = exports.InstanceCreditSpecificationRequest = exports.ModifyInstanceCapacityReservationAttributesResult = exports.ModifyInstanceCapacityReservationAttributesRequest = exports.CapacityReservationSpecification = exports.ModifyInstanceAttributeRequest = exports.BlobAttributeValue = exports.InstanceBlockDeviceMappingSpecification = exports.EbsInstanceBlockDeviceSpecification = exports.ModifyImageAttributeRequest = exports.LaunchPermissionModifications = exports.ModifyIdFormatRequest = exports.ModifyIdentityIdFormatRequest = exports.ModifyHostsResult = exports.ModifyHostsRequest = exports.ModifyFpgaImageAttributeResult = exports.ModifyFpgaImageAttributeRequest = exports.LoadPermissionModifications = exports.LoadPermissionRequest = exports.ModifyFleetResult = exports.ModifyFleetRequest = exports.ModifyEbsDefaultKmsKeyIdResult = exports.ModifyEbsDefaultKmsKeyIdRequest = exports.ModifyDefaultCreditSpecificationResult = exports.ModifyDefaultCreditSpecificationRequest = exports.ModifyClientVpnEndpointResult = exports.ModifyClientVpnEndpointRequest = exports.DnsServersOptionsModifyStructure = exports.ModifyCapacityReservationResult = exports.ModifyCapacityReservationRequest = exports.ModifyAvailabilityZoneGroupResult = exports.ModifyAvailabilityZoneGroupRequest = exports.ModifyAddressAttributeResult = exports.ModifyAddressAttributeRequest = exports.ImportVolumeResult = exports.ImportVolumeRequest = exports.ImportSnapshotResult = exports.ImportSnapshotRequest = void 0;
exports.ModifyVpcEndpointConnectionNotificationResult = exports.ModifyVpcEndpointConnectionNotificationRequest = exports.ModifyVpcEndpointResult = exports.ModifyVpcEndpointRequest = exports.ModifyVpcAttributeRequest = exports.ModifyVolumeAttributeRequest = exports.ModifyVolumeResult = exports.ModifyVolumeRequest = exports.ModifyTransitGatewayVpcAttachmentResult = exports.ModifyTransitGatewayVpcAttachmentRequest = exports.ModifyTransitGatewayVpcAttachmentRequestOptions = exports.ModifyTransitGatewayPrefixListReferenceResult = exports.ModifyTransitGatewayPrefixListReferenceRequest = exports.ModifyTransitGatewayResult = exports.ModifyTransitGatewayRequest = exports.ModifyTransitGatewayOptions = exports.ModifyTrafficMirrorSessionResult = exports.ModifyTrafficMirrorSessionRequest = exports.ModifyTrafficMirrorFilterRuleResult = exports.ModifyTrafficMirrorFilterRuleRequest = exports.ModifyTrafficMirrorFilterNetworkServicesResult = exports.ModifyTrafficMirrorFilterNetworkServicesRequest = exports.ModifySubnetAttributeRequest = exports.ModifySpotFleetRequestResponse = exports.ModifySpotFleetRequestRequest = exports.ModifySnapshotAttributeRequest = exports.CreateVolumePermissionModifications = exports.ModifySecurityGroupRulesResult = exports.ModifySecurityGroupRulesRequest = exports.SecurityGroupRuleUpdate = exports.SecurityGroupRuleRequest = exports.ModifyReservedInstancesResult = exports.ModifyReservedInstancesRequest = exports.ModifyNetworkInterfaceAttributeRequest = exports.NetworkInterfaceAttachmentChanges = exports.ModifyManagedPrefixListResult = exports.ModifyManagedPrefixListRequest = exports.RemovePrefixListEntry = exports.ModifyLaunchTemplateResult = exports.ModifyLaunchTemplateRequest = exports.ModifyInstancePlacementResult = void 0;
const smithy_client_1 = require("@aws-sdk/smithy-client");
var DescribeVolumeStatusResult;
(function (DescribeVolumeStatusResult) {
    /**
     * @internal
     */
    DescribeVolumeStatusResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVolumeStatusResult = exports.DescribeVolumeStatusResult || (exports.DescribeVolumeStatusResult = {}));
var DescribeVpcAttributeRequest;
(function (DescribeVpcAttributeRequest) {
    /**
     * @internal
     */
    DescribeVpcAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcAttributeRequest = exports.DescribeVpcAttributeRequest || (exports.DescribeVpcAttributeRequest = {}));
var DescribeVpcAttributeResult;
(function (DescribeVpcAttributeResult) {
    /**
     * @internal
     */
    DescribeVpcAttributeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcAttributeResult = exports.DescribeVpcAttributeResult || (exports.DescribeVpcAttributeResult = {}));
var DescribeVpcClassicLinkRequest;
(function (DescribeVpcClassicLinkRequest) {
    /**
     * @internal
     */
    DescribeVpcClassicLinkRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcClassicLinkRequest = exports.DescribeVpcClassicLinkRequest || (exports.DescribeVpcClassicLinkRequest = {}));
var VpcClassicLink;
(function (VpcClassicLink) {
    /**
     * @internal
     */
    VpcClassicLink.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(VpcClassicLink = exports.VpcClassicLink || (exports.VpcClassicLink = {}));
var DescribeVpcClassicLinkResult;
(function (DescribeVpcClassicLinkResult) {
    /**
     * @internal
     */
    DescribeVpcClassicLinkResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcClassicLinkResult = exports.DescribeVpcClassicLinkResult || (exports.DescribeVpcClassicLinkResult = {}));
var DescribeVpcClassicLinkDnsSupportRequest;
(function (DescribeVpcClassicLinkDnsSupportRequest) {
    /**
     * @internal
     */
    DescribeVpcClassicLinkDnsSupportRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcClassicLinkDnsSupportRequest = exports.DescribeVpcClassicLinkDnsSupportRequest || (exports.DescribeVpcClassicLinkDnsSupportRequest = {}));
var ClassicLinkDnsSupport;
(function (ClassicLinkDnsSupport) {
    /**
     * @internal
     */
    ClassicLinkDnsSupport.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ClassicLinkDnsSupport = exports.ClassicLinkDnsSupport || (exports.ClassicLinkDnsSupport = {}));
var DescribeVpcClassicLinkDnsSupportResult;
(function (DescribeVpcClassicLinkDnsSupportResult) {
    /**
     * @internal
     */
    DescribeVpcClassicLinkDnsSupportResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcClassicLinkDnsSupportResult = exports.DescribeVpcClassicLinkDnsSupportResult || (exports.DescribeVpcClassicLinkDnsSupportResult = {}));
var DescribeVpcEndpointConnectionNotificationsRequest;
(function (DescribeVpcEndpointConnectionNotificationsRequest) {
    /**
     * @internal
     */
    DescribeVpcEndpointConnectionNotificationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointConnectionNotificationsRequest = exports.DescribeVpcEndpointConnectionNotificationsRequest || (exports.DescribeVpcEndpointConnectionNotificationsRequest = {}));
var DescribeVpcEndpointConnectionNotificationsResult;
(function (DescribeVpcEndpointConnectionNotificationsResult) {
    /**
     * @internal
     */
    DescribeVpcEndpointConnectionNotificationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointConnectionNotificationsResult = exports.DescribeVpcEndpointConnectionNotificationsResult || (exports.DescribeVpcEndpointConnectionNotificationsResult = {}));
var DescribeVpcEndpointConnectionsRequest;
(function (DescribeVpcEndpointConnectionsRequest) {
    /**
     * @internal
     */
    DescribeVpcEndpointConnectionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointConnectionsRequest = exports.DescribeVpcEndpointConnectionsRequest || (exports.DescribeVpcEndpointConnectionsRequest = {}));
var VpcEndpointConnection;
(function (VpcEndpointConnection) {
    /**
     * @internal
     */
    VpcEndpointConnection.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(VpcEndpointConnection = exports.VpcEndpointConnection || (exports.VpcEndpointConnection = {}));
var DescribeVpcEndpointConnectionsResult;
(function (DescribeVpcEndpointConnectionsResult) {
    /**
     * @internal
     */
    DescribeVpcEndpointConnectionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointConnectionsResult = exports.DescribeVpcEndpointConnectionsResult || (exports.DescribeVpcEndpointConnectionsResult = {}));
var DescribeVpcEndpointsRequest;
(function (DescribeVpcEndpointsRequest) {
    /**
     * @internal
     */
    DescribeVpcEndpointsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointsRequest = exports.DescribeVpcEndpointsRequest || (exports.DescribeVpcEndpointsRequest = {}));
var DescribeVpcEndpointsResult;
(function (DescribeVpcEndpointsResult) {
    /**
     * @internal
     */
    DescribeVpcEndpointsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointsResult = exports.DescribeVpcEndpointsResult || (exports.DescribeVpcEndpointsResult = {}));
var DescribeVpcEndpointServiceConfigurationsRequest;
(function (DescribeVpcEndpointServiceConfigurationsRequest) {
    /**
     * @internal
     */
    DescribeVpcEndpointServiceConfigurationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointServiceConfigurationsRequest = exports.DescribeVpcEndpointServiceConfigurationsRequest || (exports.DescribeVpcEndpointServiceConfigurationsRequest = {}));
var DescribeVpcEndpointServiceConfigurationsResult;
(function (DescribeVpcEndpointServiceConfigurationsResult) {
    /**
     * @internal
     */
    DescribeVpcEndpointServiceConfigurationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointServiceConfigurationsResult = exports.DescribeVpcEndpointServiceConfigurationsResult || (exports.DescribeVpcEndpointServiceConfigurationsResult = {}));
var DescribeVpcEndpointServicePermissionsRequest;
(function (DescribeVpcEndpointServicePermissionsRequest) {
    /**
     * @internal
     */
    DescribeVpcEndpointServicePermissionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointServicePermissionsRequest = exports.DescribeVpcEndpointServicePermissionsRequest || (exports.DescribeVpcEndpointServicePermissionsRequest = {}));
var DescribeVpcEndpointServicePermissionsResult;
(function (DescribeVpcEndpointServicePermissionsResult) {
    /**
     * @internal
     */
    DescribeVpcEndpointServicePermissionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointServicePermissionsResult = exports.DescribeVpcEndpointServicePermissionsResult || (exports.DescribeVpcEndpointServicePermissionsResult = {}));
var DescribeVpcEndpointServicesRequest;
(function (DescribeVpcEndpointServicesRequest) {
    /**
     * @internal
     */
    DescribeVpcEndpointServicesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointServicesRequest = exports.DescribeVpcEndpointServicesRequest || (exports.DescribeVpcEndpointServicesRequest = {}));
var PrivateDnsDetails;
(function (PrivateDnsDetails) {
    /**
     * @internal
     */
    PrivateDnsDetails.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PrivateDnsDetails = exports.PrivateDnsDetails || (exports.PrivateDnsDetails = {}));
var ServiceDetail;
(function (ServiceDetail) {
    /**
     * @internal
     */
    ServiceDetail.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ServiceDetail = exports.ServiceDetail || (exports.ServiceDetail = {}));
var DescribeVpcEndpointServicesResult;
(function (DescribeVpcEndpointServicesResult) {
    /**
     * @internal
     */
    DescribeVpcEndpointServicesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcEndpointServicesResult = exports.DescribeVpcEndpointServicesResult || (exports.DescribeVpcEndpointServicesResult = {}));
var DescribeVpcPeeringConnectionsRequest;
(function (DescribeVpcPeeringConnectionsRequest) {
    /**
     * @internal
     */
    DescribeVpcPeeringConnectionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcPeeringConnectionsRequest = exports.DescribeVpcPeeringConnectionsRequest || (exports.DescribeVpcPeeringConnectionsRequest = {}));
var DescribeVpcPeeringConnectionsResult;
(function (DescribeVpcPeeringConnectionsResult) {
    /**
     * @internal
     */
    DescribeVpcPeeringConnectionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcPeeringConnectionsResult = exports.DescribeVpcPeeringConnectionsResult || (exports.DescribeVpcPeeringConnectionsResult = {}));
var DescribeVpcsRequest;
(function (DescribeVpcsRequest) {
    /**
     * @internal
     */
    DescribeVpcsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcsRequest = exports.DescribeVpcsRequest || (exports.DescribeVpcsRequest = {}));
var DescribeVpcsResult;
(function (DescribeVpcsResult) {
    /**
     * @internal
     */
    DescribeVpcsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpcsResult = exports.DescribeVpcsResult || (exports.DescribeVpcsResult = {}));
var DescribeVpnConnectionsRequest;
(function (DescribeVpnConnectionsRequest) {
    /**
     * @internal
     */
    DescribeVpnConnectionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpnConnectionsRequest = exports.DescribeVpnConnectionsRequest || (exports.DescribeVpnConnectionsRequest = {}));
var DescribeVpnConnectionsResult;
(function (DescribeVpnConnectionsResult) {
    /**
     * @internal
     */
    DescribeVpnConnectionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpnConnectionsResult = exports.DescribeVpnConnectionsResult || (exports.DescribeVpnConnectionsResult = {}));
var DescribeVpnGatewaysRequest;
(function (DescribeVpnGatewaysRequest) {
    /**
     * @internal
     */
    DescribeVpnGatewaysRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpnGatewaysRequest = exports.DescribeVpnGatewaysRequest || (exports.DescribeVpnGatewaysRequest = {}));
var DescribeVpnGatewaysResult;
(function (DescribeVpnGatewaysResult) {
    /**
     * @internal
     */
    DescribeVpnGatewaysResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DescribeVpnGatewaysResult = exports.DescribeVpnGatewaysResult || (exports.DescribeVpnGatewaysResult = {}));
var DetachClassicLinkVpcRequest;
(function (DetachClassicLinkVpcRequest) {
    /**
     * @internal
     */
    DetachClassicLinkVpcRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DetachClassicLinkVpcRequest = exports.DetachClassicLinkVpcRequest || (exports.DetachClassicLinkVpcRequest = {}));
var DetachClassicLinkVpcResult;
(function (DetachClassicLinkVpcResult) {
    /**
     * @internal
     */
    DetachClassicLinkVpcResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DetachClassicLinkVpcResult = exports.DetachClassicLinkVpcResult || (exports.DetachClassicLinkVpcResult = {}));
var DetachInternetGatewayRequest;
(function (DetachInternetGatewayRequest) {
    /**
     * @internal
     */
    DetachInternetGatewayRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DetachInternetGatewayRequest = exports.DetachInternetGatewayRequest || (exports.DetachInternetGatewayRequest = {}));
var DetachNetworkInterfaceRequest;
(function (DetachNetworkInterfaceRequest) {
    /**
     * @internal
     */
    DetachNetworkInterfaceRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DetachNetworkInterfaceRequest = exports.DetachNetworkInterfaceRequest || (exports.DetachNetworkInterfaceRequest = {}));
var DetachVolumeRequest;
(function (DetachVolumeRequest) {
    /**
     * @internal
     */
    DetachVolumeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DetachVolumeRequest = exports.DetachVolumeRequest || (exports.DetachVolumeRequest = {}));
var DetachVpnGatewayRequest;
(function (DetachVpnGatewayRequest) {
    /**
     * @internal
     */
    DetachVpnGatewayRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DetachVpnGatewayRequest = exports.DetachVpnGatewayRequest || (exports.DetachVpnGatewayRequest = {}));
var DisableEbsEncryptionByDefaultRequest;
(function (DisableEbsEncryptionByDefaultRequest) {
    /**
     * @internal
     */
    DisableEbsEncryptionByDefaultRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableEbsEncryptionByDefaultRequest = exports.DisableEbsEncryptionByDefaultRequest || (exports.DisableEbsEncryptionByDefaultRequest = {}));
var DisableEbsEncryptionByDefaultResult;
(function (DisableEbsEncryptionByDefaultResult) {
    /**
     * @internal
     */
    DisableEbsEncryptionByDefaultResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableEbsEncryptionByDefaultResult = exports.DisableEbsEncryptionByDefaultResult || (exports.DisableEbsEncryptionByDefaultResult = {}));
var DisableFastSnapshotRestoresRequest;
(function (DisableFastSnapshotRestoresRequest) {
    /**
     * @internal
     */
    DisableFastSnapshotRestoresRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableFastSnapshotRestoresRequest = exports.DisableFastSnapshotRestoresRequest || (exports.DisableFastSnapshotRestoresRequest = {}));
var DisableFastSnapshotRestoreSuccessItem;
(function (DisableFastSnapshotRestoreSuccessItem) {
    /**
     * @internal
     */
    DisableFastSnapshotRestoreSuccessItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableFastSnapshotRestoreSuccessItem = exports.DisableFastSnapshotRestoreSuccessItem || (exports.DisableFastSnapshotRestoreSuccessItem = {}));
var DisableFastSnapshotRestoreStateError;
(function (DisableFastSnapshotRestoreStateError) {
    /**
     * @internal
     */
    DisableFastSnapshotRestoreStateError.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableFastSnapshotRestoreStateError = exports.DisableFastSnapshotRestoreStateError || (exports.DisableFastSnapshotRestoreStateError = {}));
var DisableFastSnapshotRestoreStateErrorItem;
(function (DisableFastSnapshotRestoreStateErrorItem) {
    /**
     * @internal
     */
    DisableFastSnapshotRestoreStateErrorItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableFastSnapshotRestoreStateErrorItem = exports.DisableFastSnapshotRestoreStateErrorItem || (exports.DisableFastSnapshotRestoreStateErrorItem = {}));
var DisableFastSnapshotRestoreErrorItem;
(function (DisableFastSnapshotRestoreErrorItem) {
    /**
     * @internal
     */
    DisableFastSnapshotRestoreErrorItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableFastSnapshotRestoreErrorItem = exports.DisableFastSnapshotRestoreErrorItem || (exports.DisableFastSnapshotRestoreErrorItem = {}));
var DisableFastSnapshotRestoresResult;
(function (DisableFastSnapshotRestoresResult) {
    /**
     * @internal
     */
    DisableFastSnapshotRestoresResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableFastSnapshotRestoresResult = exports.DisableFastSnapshotRestoresResult || (exports.DisableFastSnapshotRestoresResult = {}));
var DisableImageDeprecationRequest;
(function (DisableImageDeprecationRequest) {
    /**
     * @internal
     */
    DisableImageDeprecationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableImageDeprecationRequest = exports.DisableImageDeprecationRequest || (exports.DisableImageDeprecationRequest = {}));
var DisableImageDeprecationResult;
(function (DisableImageDeprecationResult) {
    /**
     * @internal
     */
    DisableImageDeprecationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableImageDeprecationResult = exports.DisableImageDeprecationResult || (exports.DisableImageDeprecationResult = {}));
var DisableSerialConsoleAccessRequest;
(function (DisableSerialConsoleAccessRequest) {
    /**
     * @internal
     */
    DisableSerialConsoleAccessRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableSerialConsoleAccessRequest = exports.DisableSerialConsoleAccessRequest || (exports.DisableSerialConsoleAccessRequest = {}));
var DisableSerialConsoleAccessResult;
(function (DisableSerialConsoleAccessResult) {
    /**
     * @internal
     */
    DisableSerialConsoleAccessResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableSerialConsoleAccessResult = exports.DisableSerialConsoleAccessResult || (exports.DisableSerialConsoleAccessResult = {}));
var DisableTransitGatewayRouteTablePropagationRequest;
(function (DisableTransitGatewayRouteTablePropagationRequest) {
    /**
     * @internal
     */
    DisableTransitGatewayRouteTablePropagationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableTransitGatewayRouteTablePropagationRequest = exports.DisableTransitGatewayRouteTablePropagationRequest || (exports.DisableTransitGatewayRouteTablePropagationRequest = {}));
var TransitGatewayPropagation;
(function (TransitGatewayPropagation) {
    /**
     * @internal
     */
    TransitGatewayPropagation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayPropagation = exports.TransitGatewayPropagation || (exports.TransitGatewayPropagation = {}));
var DisableTransitGatewayRouteTablePropagationResult;
(function (DisableTransitGatewayRouteTablePropagationResult) {
    /**
     * @internal
     */
    DisableTransitGatewayRouteTablePropagationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableTransitGatewayRouteTablePropagationResult = exports.DisableTransitGatewayRouteTablePropagationResult || (exports.DisableTransitGatewayRouteTablePropagationResult = {}));
var DisableVgwRoutePropagationRequest;
(function (DisableVgwRoutePropagationRequest) {
    /**
     * @internal
     */
    DisableVgwRoutePropagationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableVgwRoutePropagationRequest = exports.DisableVgwRoutePropagationRequest || (exports.DisableVgwRoutePropagationRequest = {}));
var DisableVpcClassicLinkRequest;
(function (DisableVpcClassicLinkRequest) {
    /**
     * @internal
     */
    DisableVpcClassicLinkRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableVpcClassicLinkRequest = exports.DisableVpcClassicLinkRequest || (exports.DisableVpcClassicLinkRequest = {}));
var DisableVpcClassicLinkResult;
(function (DisableVpcClassicLinkResult) {
    /**
     * @internal
     */
    DisableVpcClassicLinkResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableVpcClassicLinkResult = exports.DisableVpcClassicLinkResult || (exports.DisableVpcClassicLinkResult = {}));
var DisableVpcClassicLinkDnsSupportRequest;
(function (DisableVpcClassicLinkDnsSupportRequest) {
    /**
     * @internal
     */
    DisableVpcClassicLinkDnsSupportRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableVpcClassicLinkDnsSupportRequest = exports.DisableVpcClassicLinkDnsSupportRequest || (exports.DisableVpcClassicLinkDnsSupportRequest = {}));
var DisableVpcClassicLinkDnsSupportResult;
(function (DisableVpcClassicLinkDnsSupportResult) {
    /**
     * @internal
     */
    DisableVpcClassicLinkDnsSupportResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisableVpcClassicLinkDnsSupportResult = exports.DisableVpcClassicLinkDnsSupportResult || (exports.DisableVpcClassicLinkDnsSupportResult = {}));
var DisassociateAddressRequest;
(function (DisassociateAddressRequest) {
    /**
     * @internal
     */
    DisassociateAddressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateAddressRequest = exports.DisassociateAddressRequest || (exports.DisassociateAddressRequest = {}));
var DisassociateClientVpnTargetNetworkRequest;
(function (DisassociateClientVpnTargetNetworkRequest) {
    /**
     * @internal
     */
    DisassociateClientVpnTargetNetworkRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateClientVpnTargetNetworkRequest = exports.DisassociateClientVpnTargetNetworkRequest || (exports.DisassociateClientVpnTargetNetworkRequest = {}));
var DisassociateClientVpnTargetNetworkResult;
(function (DisassociateClientVpnTargetNetworkResult) {
    /**
     * @internal
     */
    DisassociateClientVpnTargetNetworkResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateClientVpnTargetNetworkResult = exports.DisassociateClientVpnTargetNetworkResult || (exports.DisassociateClientVpnTargetNetworkResult = {}));
var DisassociateEnclaveCertificateIamRoleRequest;
(function (DisassociateEnclaveCertificateIamRoleRequest) {
    /**
     * @internal
     */
    DisassociateEnclaveCertificateIamRoleRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateEnclaveCertificateIamRoleRequest = exports.DisassociateEnclaveCertificateIamRoleRequest || (exports.DisassociateEnclaveCertificateIamRoleRequest = {}));
var DisassociateEnclaveCertificateIamRoleResult;
(function (DisassociateEnclaveCertificateIamRoleResult) {
    /**
     * @internal
     */
    DisassociateEnclaveCertificateIamRoleResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateEnclaveCertificateIamRoleResult = exports.DisassociateEnclaveCertificateIamRoleResult || (exports.DisassociateEnclaveCertificateIamRoleResult = {}));
var DisassociateIamInstanceProfileRequest;
(function (DisassociateIamInstanceProfileRequest) {
    /**
     * @internal
     */
    DisassociateIamInstanceProfileRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateIamInstanceProfileRequest = exports.DisassociateIamInstanceProfileRequest || (exports.DisassociateIamInstanceProfileRequest = {}));
var DisassociateIamInstanceProfileResult;
(function (DisassociateIamInstanceProfileResult) {
    /**
     * @internal
     */
    DisassociateIamInstanceProfileResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateIamInstanceProfileResult = exports.DisassociateIamInstanceProfileResult || (exports.DisassociateIamInstanceProfileResult = {}));
var InstanceEventWindowDisassociationRequest;
(function (InstanceEventWindowDisassociationRequest) {
    /**
     * @internal
     */
    InstanceEventWindowDisassociationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceEventWindowDisassociationRequest = exports.InstanceEventWindowDisassociationRequest || (exports.InstanceEventWindowDisassociationRequest = {}));
var DisassociateInstanceEventWindowRequest;
(function (DisassociateInstanceEventWindowRequest) {
    /**
     * @internal
     */
    DisassociateInstanceEventWindowRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateInstanceEventWindowRequest = exports.DisassociateInstanceEventWindowRequest || (exports.DisassociateInstanceEventWindowRequest = {}));
var DisassociateInstanceEventWindowResult;
(function (DisassociateInstanceEventWindowResult) {
    /**
     * @internal
     */
    DisassociateInstanceEventWindowResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateInstanceEventWindowResult = exports.DisassociateInstanceEventWindowResult || (exports.DisassociateInstanceEventWindowResult = {}));
var DisassociateRouteTableRequest;
(function (DisassociateRouteTableRequest) {
    /**
     * @internal
     */
    DisassociateRouteTableRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateRouteTableRequest = exports.DisassociateRouteTableRequest || (exports.DisassociateRouteTableRequest = {}));
var DisassociateSubnetCidrBlockRequest;
(function (DisassociateSubnetCidrBlockRequest) {
    /**
     * @internal
     */
    DisassociateSubnetCidrBlockRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateSubnetCidrBlockRequest = exports.DisassociateSubnetCidrBlockRequest || (exports.DisassociateSubnetCidrBlockRequest = {}));
var DisassociateSubnetCidrBlockResult;
(function (DisassociateSubnetCidrBlockResult) {
    /**
     * @internal
     */
    DisassociateSubnetCidrBlockResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateSubnetCidrBlockResult = exports.DisassociateSubnetCidrBlockResult || (exports.DisassociateSubnetCidrBlockResult = {}));
var DisassociateTransitGatewayMulticastDomainRequest;
(function (DisassociateTransitGatewayMulticastDomainRequest) {
    /**
     * @internal
     */
    DisassociateTransitGatewayMulticastDomainRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateTransitGatewayMulticastDomainRequest = exports.DisassociateTransitGatewayMulticastDomainRequest || (exports.DisassociateTransitGatewayMulticastDomainRequest = {}));
var DisassociateTransitGatewayMulticastDomainResult;
(function (DisassociateTransitGatewayMulticastDomainResult) {
    /**
     * @internal
     */
    DisassociateTransitGatewayMulticastDomainResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateTransitGatewayMulticastDomainResult = exports.DisassociateTransitGatewayMulticastDomainResult || (exports.DisassociateTransitGatewayMulticastDomainResult = {}));
var DisassociateTransitGatewayRouteTableRequest;
(function (DisassociateTransitGatewayRouteTableRequest) {
    /**
     * @internal
     */
    DisassociateTransitGatewayRouteTableRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateTransitGatewayRouteTableRequest = exports.DisassociateTransitGatewayRouteTableRequest || (exports.DisassociateTransitGatewayRouteTableRequest = {}));
var DisassociateTransitGatewayRouteTableResult;
(function (DisassociateTransitGatewayRouteTableResult) {
    /**
     * @internal
     */
    DisassociateTransitGatewayRouteTableResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateTransitGatewayRouteTableResult = exports.DisassociateTransitGatewayRouteTableResult || (exports.DisassociateTransitGatewayRouteTableResult = {}));
var DisassociateTrunkInterfaceRequest;
(function (DisassociateTrunkInterfaceRequest) {
    /**
     * @internal
     */
    DisassociateTrunkInterfaceRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateTrunkInterfaceRequest = exports.DisassociateTrunkInterfaceRequest || (exports.DisassociateTrunkInterfaceRequest = {}));
var DisassociateTrunkInterfaceResult;
(function (DisassociateTrunkInterfaceResult) {
    /**
     * @internal
     */
    DisassociateTrunkInterfaceResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateTrunkInterfaceResult = exports.DisassociateTrunkInterfaceResult || (exports.DisassociateTrunkInterfaceResult = {}));
var DisassociateVpcCidrBlockRequest;
(function (DisassociateVpcCidrBlockRequest) {
    /**
     * @internal
     */
    DisassociateVpcCidrBlockRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateVpcCidrBlockRequest = exports.DisassociateVpcCidrBlockRequest || (exports.DisassociateVpcCidrBlockRequest = {}));
var DisassociateVpcCidrBlockResult;
(function (DisassociateVpcCidrBlockResult) {
    /**
     * @internal
     */
    DisassociateVpcCidrBlockResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DisassociateVpcCidrBlockResult = exports.DisassociateVpcCidrBlockResult || (exports.DisassociateVpcCidrBlockResult = {}));
var EnableEbsEncryptionByDefaultRequest;
(function (EnableEbsEncryptionByDefaultRequest) {
    /**
     * @internal
     */
    EnableEbsEncryptionByDefaultRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableEbsEncryptionByDefaultRequest = exports.EnableEbsEncryptionByDefaultRequest || (exports.EnableEbsEncryptionByDefaultRequest = {}));
var EnableEbsEncryptionByDefaultResult;
(function (EnableEbsEncryptionByDefaultResult) {
    /**
     * @internal
     */
    EnableEbsEncryptionByDefaultResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableEbsEncryptionByDefaultResult = exports.EnableEbsEncryptionByDefaultResult || (exports.EnableEbsEncryptionByDefaultResult = {}));
var EnableFastSnapshotRestoresRequest;
(function (EnableFastSnapshotRestoresRequest) {
    /**
     * @internal
     */
    EnableFastSnapshotRestoresRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableFastSnapshotRestoresRequest = exports.EnableFastSnapshotRestoresRequest || (exports.EnableFastSnapshotRestoresRequest = {}));
var EnableFastSnapshotRestoreSuccessItem;
(function (EnableFastSnapshotRestoreSuccessItem) {
    /**
     * @internal
     */
    EnableFastSnapshotRestoreSuccessItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableFastSnapshotRestoreSuccessItem = exports.EnableFastSnapshotRestoreSuccessItem || (exports.EnableFastSnapshotRestoreSuccessItem = {}));
var EnableFastSnapshotRestoreStateError;
(function (EnableFastSnapshotRestoreStateError) {
    /**
     * @internal
     */
    EnableFastSnapshotRestoreStateError.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableFastSnapshotRestoreStateError = exports.EnableFastSnapshotRestoreStateError || (exports.EnableFastSnapshotRestoreStateError = {}));
var EnableFastSnapshotRestoreStateErrorItem;
(function (EnableFastSnapshotRestoreStateErrorItem) {
    /**
     * @internal
     */
    EnableFastSnapshotRestoreStateErrorItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableFastSnapshotRestoreStateErrorItem = exports.EnableFastSnapshotRestoreStateErrorItem || (exports.EnableFastSnapshotRestoreStateErrorItem = {}));
var EnableFastSnapshotRestoreErrorItem;
(function (EnableFastSnapshotRestoreErrorItem) {
    /**
     * @internal
     */
    EnableFastSnapshotRestoreErrorItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableFastSnapshotRestoreErrorItem = exports.EnableFastSnapshotRestoreErrorItem || (exports.EnableFastSnapshotRestoreErrorItem = {}));
var EnableFastSnapshotRestoresResult;
(function (EnableFastSnapshotRestoresResult) {
    /**
     * @internal
     */
    EnableFastSnapshotRestoresResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableFastSnapshotRestoresResult = exports.EnableFastSnapshotRestoresResult || (exports.EnableFastSnapshotRestoresResult = {}));
var EnableImageDeprecationRequest;
(function (EnableImageDeprecationRequest) {
    /**
     * @internal
     */
    EnableImageDeprecationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableImageDeprecationRequest = exports.EnableImageDeprecationRequest || (exports.EnableImageDeprecationRequest = {}));
var EnableImageDeprecationResult;
(function (EnableImageDeprecationResult) {
    /**
     * @internal
     */
    EnableImageDeprecationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableImageDeprecationResult = exports.EnableImageDeprecationResult || (exports.EnableImageDeprecationResult = {}));
var EnableSerialConsoleAccessRequest;
(function (EnableSerialConsoleAccessRequest) {
    /**
     * @internal
     */
    EnableSerialConsoleAccessRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableSerialConsoleAccessRequest = exports.EnableSerialConsoleAccessRequest || (exports.EnableSerialConsoleAccessRequest = {}));
var EnableSerialConsoleAccessResult;
(function (EnableSerialConsoleAccessResult) {
    /**
     * @internal
     */
    EnableSerialConsoleAccessResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableSerialConsoleAccessResult = exports.EnableSerialConsoleAccessResult || (exports.EnableSerialConsoleAccessResult = {}));
var EnableTransitGatewayRouteTablePropagationRequest;
(function (EnableTransitGatewayRouteTablePropagationRequest) {
    /**
     * @internal
     */
    EnableTransitGatewayRouteTablePropagationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableTransitGatewayRouteTablePropagationRequest = exports.EnableTransitGatewayRouteTablePropagationRequest || (exports.EnableTransitGatewayRouteTablePropagationRequest = {}));
var EnableTransitGatewayRouteTablePropagationResult;
(function (EnableTransitGatewayRouteTablePropagationResult) {
    /**
     * @internal
     */
    EnableTransitGatewayRouteTablePropagationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableTransitGatewayRouteTablePropagationResult = exports.EnableTransitGatewayRouteTablePropagationResult || (exports.EnableTransitGatewayRouteTablePropagationResult = {}));
var EnableVgwRoutePropagationRequest;
(function (EnableVgwRoutePropagationRequest) {
    /**
     * @internal
     */
    EnableVgwRoutePropagationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableVgwRoutePropagationRequest = exports.EnableVgwRoutePropagationRequest || (exports.EnableVgwRoutePropagationRequest = {}));
var EnableVolumeIORequest;
(function (EnableVolumeIORequest) {
    /**
     * @internal
     */
    EnableVolumeIORequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableVolumeIORequest = exports.EnableVolumeIORequest || (exports.EnableVolumeIORequest = {}));
var EnableVpcClassicLinkRequest;
(function (EnableVpcClassicLinkRequest) {
    /**
     * @internal
     */
    EnableVpcClassicLinkRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableVpcClassicLinkRequest = exports.EnableVpcClassicLinkRequest || (exports.EnableVpcClassicLinkRequest = {}));
var EnableVpcClassicLinkResult;
(function (EnableVpcClassicLinkResult) {
    /**
     * @internal
     */
    EnableVpcClassicLinkResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableVpcClassicLinkResult = exports.EnableVpcClassicLinkResult || (exports.EnableVpcClassicLinkResult = {}));
var EnableVpcClassicLinkDnsSupportRequest;
(function (EnableVpcClassicLinkDnsSupportRequest) {
    /**
     * @internal
     */
    EnableVpcClassicLinkDnsSupportRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableVpcClassicLinkDnsSupportRequest = exports.EnableVpcClassicLinkDnsSupportRequest || (exports.EnableVpcClassicLinkDnsSupportRequest = {}));
var EnableVpcClassicLinkDnsSupportResult;
(function (EnableVpcClassicLinkDnsSupportResult) {
    /**
     * @internal
     */
    EnableVpcClassicLinkDnsSupportResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnableVpcClassicLinkDnsSupportResult = exports.EnableVpcClassicLinkDnsSupportResult || (exports.EnableVpcClassicLinkDnsSupportResult = {}));
var ExportClientVpnClientCertificateRevocationListRequest;
(function (ExportClientVpnClientCertificateRevocationListRequest) {
    /**
     * @internal
     */
    ExportClientVpnClientCertificateRevocationListRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportClientVpnClientCertificateRevocationListRequest = exports.ExportClientVpnClientCertificateRevocationListRequest || (exports.ExportClientVpnClientCertificateRevocationListRequest = {}));
var ClientCertificateRevocationListStatus;
(function (ClientCertificateRevocationListStatus) {
    /**
     * @internal
     */
    ClientCertificateRevocationListStatus.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ClientCertificateRevocationListStatus = exports.ClientCertificateRevocationListStatus || (exports.ClientCertificateRevocationListStatus = {}));
var ExportClientVpnClientCertificateRevocationListResult;
(function (ExportClientVpnClientCertificateRevocationListResult) {
    /**
     * @internal
     */
    ExportClientVpnClientCertificateRevocationListResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportClientVpnClientCertificateRevocationListResult = exports.ExportClientVpnClientCertificateRevocationListResult || (exports.ExportClientVpnClientCertificateRevocationListResult = {}));
var ExportClientVpnClientConfigurationRequest;
(function (ExportClientVpnClientConfigurationRequest) {
    /**
     * @internal
     */
    ExportClientVpnClientConfigurationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportClientVpnClientConfigurationRequest = exports.ExportClientVpnClientConfigurationRequest || (exports.ExportClientVpnClientConfigurationRequest = {}));
var ExportClientVpnClientConfigurationResult;
(function (ExportClientVpnClientConfigurationResult) {
    /**
     * @internal
     */
    ExportClientVpnClientConfigurationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportClientVpnClientConfigurationResult = exports.ExportClientVpnClientConfigurationResult || (exports.ExportClientVpnClientConfigurationResult = {}));
var ExportTaskS3LocationRequest;
(function (ExportTaskS3LocationRequest) {
    /**
     * @internal
     */
    ExportTaskS3LocationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportTaskS3LocationRequest = exports.ExportTaskS3LocationRequest || (exports.ExportTaskS3LocationRequest = {}));
var ExportImageRequest;
(function (ExportImageRequest) {
    /**
     * @internal
     */
    ExportImageRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportImageRequest = exports.ExportImageRequest || (exports.ExportImageRequest = {}));
var ExportImageResult;
(function (ExportImageResult) {
    /**
     * @internal
     */
    ExportImageResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportImageResult = exports.ExportImageResult || (exports.ExportImageResult = {}));
var ExportTransitGatewayRoutesRequest;
(function (ExportTransitGatewayRoutesRequest) {
    /**
     * @internal
     */
    ExportTransitGatewayRoutesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportTransitGatewayRoutesRequest = exports.ExportTransitGatewayRoutesRequest || (exports.ExportTransitGatewayRoutesRequest = {}));
var ExportTransitGatewayRoutesResult;
(function (ExportTransitGatewayRoutesResult) {
    /**
     * @internal
     */
    ExportTransitGatewayRoutesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ExportTransitGatewayRoutesResult = exports.ExportTransitGatewayRoutesResult || (exports.ExportTransitGatewayRoutesResult = {}));
var GetAssociatedEnclaveCertificateIamRolesRequest;
(function (GetAssociatedEnclaveCertificateIamRolesRequest) {
    /**
     * @internal
     */
    GetAssociatedEnclaveCertificateIamRolesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetAssociatedEnclaveCertificateIamRolesRequest = exports.GetAssociatedEnclaveCertificateIamRolesRequest || (exports.GetAssociatedEnclaveCertificateIamRolesRequest = {}));
var AssociatedRole;
(function (AssociatedRole) {
    /**
     * @internal
     */
    AssociatedRole.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(AssociatedRole = exports.AssociatedRole || (exports.AssociatedRole = {}));
var GetAssociatedEnclaveCertificateIamRolesResult;
(function (GetAssociatedEnclaveCertificateIamRolesResult) {
    /**
     * @internal
     */
    GetAssociatedEnclaveCertificateIamRolesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetAssociatedEnclaveCertificateIamRolesResult = exports.GetAssociatedEnclaveCertificateIamRolesResult || (exports.GetAssociatedEnclaveCertificateIamRolesResult = {}));
var GetAssociatedIpv6PoolCidrsRequest;
(function (GetAssociatedIpv6PoolCidrsRequest) {
    /**
     * @internal
     */
    GetAssociatedIpv6PoolCidrsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetAssociatedIpv6PoolCidrsRequest = exports.GetAssociatedIpv6PoolCidrsRequest || (exports.GetAssociatedIpv6PoolCidrsRequest = {}));
var Ipv6CidrAssociation;
(function (Ipv6CidrAssociation) {
    /**
     * @internal
     */
    Ipv6CidrAssociation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(Ipv6CidrAssociation = exports.Ipv6CidrAssociation || (exports.Ipv6CidrAssociation = {}));
var GetAssociatedIpv6PoolCidrsResult;
(function (GetAssociatedIpv6PoolCidrsResult) {
    /**
     * @internal
     */
    GetAssociatedIpv6PoolCidrsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetAssociatedIpv6PoolCidrsResult = exports.GetAssociatedIpv6PoolCidrsResult || (exports.GetAssociatedIpv6PoolCidrsResult = {}));
var GetCapacityReservationUsageRequest;
(function (GetCapacityReservationUsageRequest) {
    /**
     * @internal
     */
    GetCapacityReservationUsageRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetCapacityReservationUsageRequest = exports.GetCapacityReservationUsageRequest || (exports.GetCapacityReservationUsageRequest = {}));
var InstanceUsage;
(function (InstanceUsage) {
    /**
     * @internal
     */
    InstanceUsage.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceUsage = exports.InstanceUsage || (exports.InstanceUsage = {}));
var GetCapacityReservationUsageResult;
(function (GetCapacityReservationUsageResult) {
    /**
     * @internal
     */
    GetCapacityReservationUsageResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetCapacityReservationUsageResult = exports.GetCapacityReservationUsageResult || (exports.GetCapacityReservationUsageResult = {}));
var GetCoipPoolUsageRequest;
(function (GetCoipPoolUsageRequest) {
    /**
     * @internal
     */
    GetCoipPoolUsageRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetCoipPoolUsageRequest = exports.GetCoipPoolUsageRequest || (exports.GetCoipPoolUsageRequest = {}));
var CoipAddressUsage;
(function (CoipAddressUsage) {
    /**
     * @internal
     */
    CoipAddressUsage.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(CoipAddressUsage = exports.CoipAddressUsage || (exports.CoipAddressUsage = {}));
var GetCoipPoolUsageResult;
(function (GetCoipPoolUsageResult) {
    /**
     * @internal
     */
    GetCoipPoolUsageResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetCoipPoolUsageResult = exports.GetCoipPoolUsageResult || (exports.GetCoipPoolUsageResult = {}));
var GetConsoleOutputRequest;
(function (GetConsoleOutputRequest) {
    /**
     * @internal
     */
    GetConsoleOutputRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetConsoleOutputRequest = exports.GetConsoleOutputRequest || (exports.GetConsoleOutputRequest = {}));
var GetConsoleOutputResult;
(function (GetConsoleOutputResult) {
    /**
     * @internal
     */
    GetConsoleOutputResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetConsoleOutputResult = exports.GetConsoleOutputResult || (exports.GetConsoleOutputResult = {}));
var GetConsoleScreenshotRequest;
(function (GetConsoleScreenshotRequest) {
    /**
     * @internal
     */
    GetConsoleScreenshotRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetConsoleScreenshotRequest = exports.GetConsoleScreenshotRequest || (exports.GetConsoleScreenshotRequest = {}));
var GetConsoleScreenshotResult;
(function (GetConsoleScreenshotResult) {
    /**
     * @internal
     */
    GetConsoleScreenshotResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetConsoleScreenshotResult = exports.GetConsoleScreenshotResult || (exports.GetConsoleScreenshotResult = {}));
var GetDefaultCreditSpecificationRequest;
(function (GetDefaultCreditSpecificationRequest) {
    /**
     * @internal
     */
    GetDefaultCreditSpecificationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetDefaultCreditSpecificationRequest = exports.GetDefaultCreditSpecificationRequest || (exports.GetDefaultCreditSpecificationRequest = {}));
var InstanceFamilyCreditSpecification;
(function (InstanceFamilyCreditSpecification) {
    /**
     * @internal
     */
    InstanceFamilyCreditSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceFamilyCreditSpecification = exports.InstanceFamilyCreditSpecification || (exports.InstanceFamilyCreditSpecification = {}));
var GetDefaultCreditSpecificationResult;
(function (GetDefaultCreditSpecificationResult) {
    /**
     * @internal
     */
    GetDefaultCreditSpecificationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetDefaultCreditSpecificationResult = exports.GetDefaultCreditSpecificationResult || (exports.GetDefaultCreditSpecificationResult = {}));
var GetEbsDefaultKmsKeyIdRequest;
(function (GetEbsDefaultKmsKeyIdRequest) {
    /**
     * @internal
     */
    GetEbsDefaultKmsKeyIdRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetEbsDefaultKmsKeyIdRequest = exports.GetEbsDefaultKmsKeyIdRequest || (exports.GetEbsDefaultKmsKeyIdRequest = {}));
var GetEbsDefaultKmsKeyIdResult;
(function (GetEbsDefaultKmsKeyIdResult) {
    /**
     * @internal
     */
    GetEbsDefaultKmsKeyIdResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetEbsDefaultKmsKeyIdResult = exports.GetEbsDefaultKmsKeyIdResult || (exports.GetEbsDefaultKmsKeyIdResult = {}));
var GetEbsEncryptionByDefaultRequest;
(function (GetEbsEncryptionByDefaultRequest) {
    /**
     * @internal
     */
    GetEbsEncryptionByDefaultRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetEbsEncryptionByDefaultRequest = exports.GetEbsEncryptionByDefaultRequest || (exports.GetEbsEncryptionByDefaultRequest = {}));
var GetEbsEncryptionByDefaultResult;
(function (GetEbsEncryptionByDefaultResult) {
    /**
     * @internal
     */
    GetEbsEncryptionByDefaultResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetEbsEncryptionByDefaultResult = exports.GetEbsEncryptionByDefaultResult || (exports.GetEbsEncryptionByDefaultResult = {}));
var PartitionLoadFrequency;
(function (PartitionLoadFrequency) {
    PartitionLoadFrequency["DAILY"] = "daily";
    PartitionLoadFrequency["MONTHLY"] = "monthly";
    PartitionLoadFrequency["NONE"] = "none";
    PartitionLoadFrequency["WEEKLY"] = "weekly";
})(PartitionLoadFrequency = exports.PartitionLoadFrequency || (exports.PartitionLoadFrequency = {}));
var AthenaIntegration;
(function (AthenaIntegration) {
    /**
     * @internal
     */
    AthenaIntegration.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(AthenaIntegration = exports.AthenaIntegration || (exports.AthenaIntegration = {}));
var IntegrateServices;
(function (IntegrateServices) {
    /**
     * @internal
     */
    IntegrateServices.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(IntegrateServices = exports.IntegrateServices || (exports.IntegrateServices = {}));
var GetFlowLogsIntegrationTemplateRequest;
(function (GetFlowLogsIntegrationTemplateRequest) {
    /**
     * @internal
     */
    GetFlowLogsIntegrationTemplateRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetFlowLogsIntegrationTemplateRequest = exports.GetFlowLogsIntegrationTemplateRequest || (exports.GetFlowLogsIntegrationTemplateRequest = {}));
var GetFlowLogsIntegrationTemplateResult;
(function (GetFlowLogsIntegrationTemplateResult) {
    /**
     * @internal
     */
    GetFlowLogsIntegrationTemplateResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetFlowLogsIntegrationTemplateResult = exports.GetFlowLogsIntegrationTemplateResult || (exports.GetFlowLogsIntegrationTemplateResult = {}));
var GetGroupsForCapacityReservationRequest;
(function (GetGroupsForCapacityReservationRequest) {
    /**
     * @internal
     */
    GetGroupsForCapacityReservationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetGroupsForCapacityReservationRequest = exports.GetGroupsForCapacityReservationRequest || (exports.GetGroupsForCapacityReservationRequest = {}));
var CapacityReservationGroup;
(function (CapacityReservationGroup) {
    /**
     * @internal
     */
    CapacityReservationGroup.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(CapacityReservationGroup = exports.CapacityReservationGroup || (exports.CapacityReservationGroup = {}));
var GetGroupsForCapacityReservationResult;
(function (GetGroupsForCapacityReservationResult) {
    /**
     * @internal
     */
    GetGroupsForCapacityReservationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetGroupsForCapacityReservationResult = exports.GetGroupsForCapacityReservationResult || (exports.GetGroupsForCapacityReservationResult = {}));
var GetHostReservationPurchasePreviewRequest;
(function (GetHostReservationPurchasePreviewRequest) {
    /**
     * @internal
     */
    GetHostReservationPurchasePreviewRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetHostReservationPurchasePreviewRequest = exports.GetHostReservationPurchasePreviewRequest || (exports.GetHostReservationPurchasePreviewRequest = {}));
var Purchase;
(function (Purchase) {
    /**
     * @internal
     */
    Purchase.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(Purchase = exports.Purchase || (exports.Purchase = {}));
var GetHostReservationPurchasePreviewResult;
(function (GetHostReservationPurchasePreviewResult) {
    /**
     * @internal
     */
    GetHostReservationPurchasePreviewResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetHostReservationPurchasePreviewResult = exports.GetHostReservationPurchasePreviewResult || (exports.GetHostReservationPurchasePreviewResult = {}));
var GetLaunchTemplateDataRequest;
(function (GetLaunchTemplateDataRequest) {
    /**
     * @internal
     */
    GetLaunchTemplateDataRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetLaunchTemplateDataRequest = exports.GetLaunchTemplateDataRequest || (exports.GetLaunchTemplateDataRequest = {}));
var GetLaunchTemplateDataResult;
(function (GetLaunchTemplateDataResult) {
    /**
     * @internal
     */
    GetLaunchTemplateDataResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetLaunchTemplateDataResult = exports.GetLaunchTemplateDataResult || (exports.GetLaunchTemplateDataResult = {}));
var GetManagedPrefixListAssociationsRequest;
(function (GetManagedPrefixListAssociationsRequest) {
    /**
     * @internal
     */
    GetManagedPrefixListAssociationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetManagedPrefixListAssociationsRequest = exports.GetManagedPrefixListAssociationsRequest || (exports.GetManagedPrefixListAssociationsRequest = {}));
var PrefixListAssociation;
(function (PrefixListAssociation) {
    /**
     * @internal
     */
    PrefixListAssociation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PrefixListAssociation = exports.PrefixListAssociation || (exports.PrefixListAssociation = {}));
var GetManagedPrefixListAssociationsResult;
(function (GetManagedPrefixListAssociationsResult) {
    /**
     * @internal
     */
    GetManagedPrefixListAssociationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetManagedPrefixListAssociationsResult = exports.GetManagedPrefixListAssociationsResult || (exports.GetManagedPrefixListAssociationsResult = {}));
var GetManagedPrefixListEntriesRequest;
(function (GetManagedPrefixListEntriesRequest) {
    /**
     * @internal
     */
    GetManagedPrefixListEntriesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetManagedPrefixListEntriesRequest = exports.GetManagedPrefixListEntriesRequest || (exports.GetManagedPrefixListEntriesRequest = {}));
var PrefixListEntry;
(function (PrefixListEntry) {
    /**
     * @internal
     */
    PrefixListEntry.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PrefixListEntry = exports.PrefixListEntry || (exports.PrefixListEntry = {}));
var GetManagedPrefixListEntriesResult;
(function (GetManagedPrefixListEntriesResult) {
    /**
     * @internal
     */
    GetManagedPrefixListEntriesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetManagedPrefixListEntriesResult = exports.GetManagedPrefixListEntriesResult || (exports.GetManagedPrefixListEntriesResult = {}));
var GetPasswordDataRequest;
(function (GetPasswordDataRequest) {
    /**
     * @internal
     */
    GetPasswordDataRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetPasswordDataRequest = exports.GetPasswordDataRequest || (exports.GetPasswordDataRequest = {}));
var GetPasswordDataResult;
(function (GetPasswordDataResult) {
    /**
     * @internal
     */
    GetPasswordDataResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetPasswordDataResult = exports.GetPasswordDataResult || (exports.GetPasswordDataResult = {}));
var GetReservedInstancesExchangeQuoteRequest;
(function (GetReservedInstancesExchangeQuoteRequest) {
    /**
     * @internal
     */
    GetReservedInstancesExchangeQuoteRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetReservedInstancesExchangeQuoteRequest = exports.GetReservedInstancesExchangeQuoteRequest || (exports.GetReservedInstancesExchangeQuoteRequest = {}));
var ReservationValue;
(function (ReservationValue) {
    /**
     * @internal
     */
    ReservationValue.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReservationValue = exports.ReservationValue || (exports.ReservationValue = {}));
var ReservedInstanceReservationValue;
(function (ReservedInstanceReservationValue) {
    /**
     * @internal
     */
    ReservedInstanceReservationValue.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReservedInstanceReservationValue = exports.ReservedInstanceReservationValue || (exports.ReservedInstanceReservationValue = {}));
var TargetConfiguration;
(function (TargetConfiguration) {
    /**
     * @internal
     */
    TargetConfiguration.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TargetConfiguration = exports.TargetConfiguration || (exports.TargetConfiguration = {}));
var TargetReservationValue;
(function (TargetReservationValue) {
    /**
     * @internal
     */
    TargetReservationValue.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TargetReservationValue = exports.TargetReservationValue || (exports.TargetReservationValue = {}));
var GetReservedInstancesExchangeQuoteResult;
(function (GetReservedInstancesExchangeQuoteResult) {
    /**
     * @internal
     */
    GetReservedInstancesExchangeQuoteResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetReservedInstancesExchangeQuoteResult = exports.GetReservedInstancesExchangeQuoteResult || (exports.GetReservedInstancesExchangeQuoteResult = {}));
var GetSerialConsoleAccessStatusRequest;
(function (GetSerialConsoleAccessStatusRequest) {
    /**
     * @internal
     */
    GetSerialConsoleAccessStatusRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetSerialConsoleAccessStatusRequest = exports.GetSerialConsoleAccessStatusRequest || (exports.GetSerialConsoleAccessStatusRequest = {}));
var GetSerialConsoleAccessStatusResult;
(function (GetSerialConsoleAccessStatusResult) {
    /**
     * @internal
     */
    GetSerialConsoleAccessStatusResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetSerialConsoleAccessStatusResult = exports.GetSerialConsoleAccessStatusResult || (exports.GetSerialConsoleAccessStatusResult = {}));
var GetSubnetCidrReservationsRequest;
(function (GetSubnetCidrReservationsRequest) {
    /**
     * @internal
     */
    GetSubnetCidrReservationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetSubnetCidrReservationsRequest = exports.GetSubnetCidrReservationsRequest || (exports.GetSubnetCidrReservationsRequest = {}));
var GetSubnetCidrReservationsResult;
(function (GetSubnetCidrReservationsResult) {
    /**
     * @internal
     */
    GetSubnetCidrReservationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetSubnetCidrReservationsResult = exports.GetSubnetCidrReservationsResult || (exports.GetSubnetCidrReservationsResult = {}));
var GetTransitGatewayAttachmentPropagationsRequest;
(function (GetTransitGatewayAttachmentPropagationsRequest) {
    /**
     * @internal
     */
    GetTransitGatewayAttachmentPropagationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayAttachmentPropagationsRequest = exports.GetTransitGatewayAttachmentPropagationsRequest || (exports.GetTransitGatewayAttachmentPropagationsRequest = {}));
var TransitGatewayAttachmentPropagation;
(function (TransitGatewayAttachmentPropagation) {
    /**
     * @internal
     */
    TransitGatewayAttachmentPropagation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayAttachmentPropagation = exports.TransitGatewayAttachmentPropagation || (exports.TransitGatewayAttachmentPropagation = {}));
var GetTransitGatewayAttachmentPropagationsResult;
(function (GetTransitGatewayAttachmentPropagationsResult) {
    /**
     * @internal
     */
    GetTransitGatewayAttachmentPropagationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayAttachmentPropagationsResult = exports.GetTransitGatewayAttachmentPropagationsResult || (exports.GetTransitGatewayAttachmentPropagationsResult = {}));
var GetTransitGatewayMulticastDomainAssociationsRequest;
(function (GetTransitGatewayMulticastDomainAssociationsRequest) {
    /**
     * @internal
     */
    GetTransitGatewayMulticastDomainAssociationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayMulticastDomainAssociationsRequest = exports.GetTransitGatewayMulticastDomainAssociationsRequest || (exports.GetTransitGatewayMulticastDomainAssociationsRequest = {}));
var TransitGatewayMulticastDomainAssociation;
(function (TransitGatewayMulticastDomainAssociation) {
    /**
     * @internal
     */
    TransitGatewayMulticastDomainAssociation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayMulticastDomainAssociation = exports.TransitGatewayMulticastDomainAssociation || (exports.TransitGatewayMulticastDomainAssociation = {}));
var GetTransitGatewayMulticastDomainAssociationsResult;
(function (GetTransitGatewayMulticastDomainAssociationsResult) {
    /**
     * @internal
     */
    GetTransitGatewayMulticastDomainAssociationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayMulticastDomainAssociationsResult = exports.GetTransitGatewayMulticastDomainAssociationsResult || (exports.GetTransitGatewayMulticastDomainAssociationsResult = {}));
var GetTransitGatewayPrefixListReferencesRequest;
(function (GetTransitGatewayPrefixListReferencesRequest) {
    /**
     * @internal
     */
    GetTransitGatewayPrefixListReferencesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayPrefixListReferencesRequest = exports.GetTransitGatewayPrefixListReferencesRequest || (exports.GetTransitGatewayPrefixListReferencesRequest = {}));
var GetTransitGatewayPrefixListReferencesResult;
(function (GetTransitGatewayPrefixListReferencesResult) {
    /**
     * @internal
     */
    GetTransitGatewayPrefixListReferencesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayPrefixListReferencesResult = exports.GetTransitGatewayPrefixListReferencesResult || (exports.GetTransitGatewayPrefixListReferencesResult = {}));
var GetTransitGatewayRouteTableAssociationsRequest;
(function (GetTransitGatewayRouteTableAssociationsRequest) {
    /**
     * @internal
     */
    GetTransitGatewayRouteTableAssociationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayRouteTableAssociationsRequest = exports.GetTransitGatewayRouteTableAssociationsRequest || (exports.GetTransitGatewayRouteTableAssociationsRequest = {}));
var TransitGatewayRouteTableAssociation;
(function (TransitGatewayRouteTableAssociation) {
    /**
     * @internal
     */
    TransitGatewayRouteTableAssociation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayRouteTableAssociation = exports.TransitGatewayRouteTableAssociation || (exports.TransitGatewayRouteTableAssociation = {}));
var GetTransitGatewayRouteTableAssociationsResult;
(function (GetTransitGatewayRouteTableAssociationsResult) {
    /**
     * @internal
     */
    GetTransitGatewayRouteTableAssociationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayRouteTableAssociationsResult = exports.GetTransitGatewayRouteTableAssociationsResult || (exports.GetTransitGatewayRouteTableAssociationsResult = {}));
var GetTransitGatewayRouteTablePropagationsRequest;
(function (GetTransitGatewayRouteTablePropagationsRequest) {
    /**
     * @internal
     */
    GetTransitGatewayRouteTablePropagationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayRouteTablePropagationsRequest = exports.GetTransitGatewayRouteTablePropagationsRequest || (exports.GetTransitGatewayRouteTablePropagationsRequest = {}));
var TransitGatewayRouteTablePropagation;
(function (TransitGatewayRouteTablePropagation) {
    /**
     * @internal
     */
    TransitGatewayRouteTablePropagation.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayRouteTablePropagation = exports.TransitGatewayRouteTablePropagation || (exports.TransitGatewayRouteTablePropagation = {}));
var GetTransitGatewayRouteTablePropagationsResult;
(function (GetTransitGatewayRouteTablePropagationsResult) {
    /**
     * @internal
     */
    GetTransitGatewayRouteTablePropagationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetTransitGatewayRouteTablePropagationsResult = exports.GetTransitGatewayRouteTablePropagationsResult || (exports.GetTransitGatewayRouteTablePropagationsResult = {}));
var GetVpnConnectionDeviceSampleConfigurationRequest;
(function (GetVpnConnectionDeviceSampleConfigurationRequest) {
    /**
     * @internal
     */
    GetVpnConnectionDeviceSampleConfigurationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetVpnConnectionDeviceSampleConfigurationRequest = exports.GetVpnConnectionDeviceSampleConfigurationRequest || (exports.GetVpnConnectionDeviceSampleConfigurationRequest = {}));
var GetVpnConnectionDeviceSampleConfigurationResult;
(function (GetVpnConnectionDeviceSampleConfigurationResult) {
    /**
     * @internal
     */
    GetVpnConnectionDeviceSampleConfigurationResult.filterSensitiveLog = (obj) => ({
        ...obj,
        ...(obj.VpnConnectionDeviceSampleConfiguration && { VpnConnectionDeviceSampleConfiguration: smithy_client_1.SENSITIVE_STRING }),
    });
})(GetVpnConnectionDeviceSampleConfigurationResult = exports.GetVpnConnectionDeviceSampleConfigurationResult || (exports.GetVpnConnectionDeviceSampleConfigurationResult = {}));
var GetVpnConnectionDeviceTypesRequest;
(function (GetVpnConnectionDeviceTypesRequest) {
    /**
     * @internal
     */
    GetVpnConnectionDeviceTypesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetVpnConnectionDeviceTypesRequest = exports.GetVpnConnectionDeviceTypesRequest || (exports.GetVpnConnectionDeviceTypesRequest = {}));
var VpnConnectionDeviceType;
(function (VpnConnectionDeviceType) {
    /**
     * @internal
     */
    VpnConnectionDeviceType.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(VpnConnectionDeviceType = exports.VpnConnectionDeviceType || (exports.VpnConnectionDeviceType = {}));
var GetVpnConnectionDeviceTypesResult;
(function (GetVpnConnectionDeviceTypesResult) {
    /**
     * @internal
     */
    GetVpnConnectionDeviceTypesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetVpnConnectionDeviceTypesResult = exports.GetVpnConnectionDeviceTypesResult || (exports.GetVpnConnectionDeviceTypesResult = {}));
var ImportClientVpnClientCertificateRevocationListRequest;
(function (ImportClientVpnClientCertificateRevocationListRequest) {
    /**
     * @internal
     */
    ImportClientVpnClientCertificateRevocationListRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportClientVpnClientCertificateRevocationListRequest = exports.ImportClientVpnClientCertificateRevocationListRequest || (exports.ImportClientVpnClientCertificateRevocationListRequest = {}));
var ImportClientVpnClientCertificateRevocationListResult;
(function (ImportClientVpnClientCertificateRevocationListResult) {
    /**
     * @internal
     */
    ImportClientVpnClientCertificateRevocationListResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportClientVpnClientCertificateRevocationListResult = exports.ImportClientVpnClientCertificateRevocationListResult || (exports.ImportClientVpnClientCertificateRevocationListResult = {}));
var ClientData;
(function (ClientData) {
    /**
     * @internal
     */
    ClientData.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ClientData = exports.ClientData || (exports.ClientData = {}));
var UserBucket;
(function (UserBucket) {
    /**
     * @internal
     */
    UserBucket.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UserBucket = exports.UserBucket || (exports.UserBucket = {}));
var ImageDiskContainer;
(function (ImageDiskContainer) {
    /**
     * @internal
     */
    ImageDiskContainer.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImageDiskContainer = exports.ImageDiskContainer || (exports.ImageDiskContainer = {}));
var ImportImageLicenseConfigurationRequest;
(function (ImportImageLicenseConfigurationRequest) {
    /**
     * @internal
     */
    ImportImageLicenseConfigurationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportImageLicenseConfigurationRequest = exports.ImportImageLicenseConfigurationRequest || (exports.ImportImageLicenseConfigurationRequest = {}));
var ImportImageRequest;
(function (ImportImageRequest) {
    /**
     * @internal
     */
    ImportImageRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportImageRequest = exports.ImportImageRequest || (exports.ImportImageRequest = {}));
var ImportImageResult;
(function (ImportImageResult) {
    /**
     * @internal
     */
    ImportImageResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportImageResult = exports.ImportImageResult || (exports.ImportImageResult = {}));
var DiskImageDetail;
(function (DiskImageDetail) {
    /**
     * @internal
     */
    DiskImageDetail.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DiskImageDetail = exports.DiskImageDetail || (exports.DiskImageDetail = {}));
var VolumeDetail;
(function (VolumeDetail) {
    /**
     * @internal
     */
    VolumeDetail.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(VolumeDetail = exports.VolumeDetail || (exports.VolumeDetail = {}));
var DiskImage;
(function (DiskImage) {
    /**
     * @internal
     */
    DiskImage.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DiskImage = exports.DiskImage || (exports.DiskImage = {}));
var UserData;
(function (UserData) {
    /**
     * @internal
     */
    UserData.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UserData = exports.UserData || (exports.UserData = {}));
var ImportInstanceLaunchSpecification;
(function (ImportInstanceLaunchSpecification) {
    /**
     * @internal
     */
    ImportInstanceLaunchSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
        ...(obj.UserData && { UserData: smithy_client_1.SENSITIVE_STRING }),
    });
})(ImportInstanceLaunchSpecification = exports.ImportInstanceLaunchSpecification || (exports.ImportInstanceLaunchSpecification = {}));
var ImportInstanceRequest;
(function (ImportInstanceRequest) {
    /**
     * @internal
     */
    ImportInstanceRequest.filterSensitiveLog = (obj) => ({
        ...obj,
        ...(obj.LaunchSpecification && {
            LaunchSpecification: ImportInstanceLaunchSpecification.filterSensitiveLog(obj.LaunchSpecification),
        }),
    });
})(ImportInstanceRequest = exports.ImportInstanceRequest || (exports.ImportInstanceRequest = {}));
var ImportInstanceResult;
(function (ImportInstanceResult) {
    /**
     * @internal
     */
    ImportInstanceResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportInstanceResult = exports.ImportInstanceResult || (exports.ImportInstanceResult = {}));
var ImportKeyPairRequest;
(function (ImportKeyPairRequest) {
    /**
     * @internal
     */
    ImportKeyPairRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportKeyPairRequest = exports.ImportKeyPairRequest || (exports.ImportKeyPairRequest = {}));
var ImportKeyPairResult;
(function (ImportKeyPairResult) {
    /**
     * @internal
     */
    ImportKeyPairResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportKeyPairResult = exports.ImportKeyPairResult || (exports.ImportKeyPairResult = {}));
var SnapshotDiskContainer;
(function (SnapshotDiskContainer) {
    /**
     * @internal
     */
    SnapshotDiskContainer.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SnapshotDiskContainer = exports.SnapshotDiskContainer || (exports.SnapshotDiskContainer = {}));
var ImportSnapshotRequest;
(function (ImportSnapshotRequest) {
    /**
     * @internal
     */
    ImportSnapshotRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportSnapshotRequest = exports.ImportSnapshotRequest || (exports.ImportSnapshotRequest = {}));
var ImportSnapshotResult;
(function (ImportSnapshotResult) {
    /**
     * @internal
     */
    ImportSnapshotResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportSnapshotResult = exports.ImportSnapshotResult || (exports.ImportSnapshotResult = {}));
var ImportVolumeRequest;
(function (ImportVolumeRequest) {
    /**
     * @internal
     */
    ImportVolumeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportVolumeRequest = exports.ImportVolumeRequest || (exports.ImportVolumeRequest = {}));
var ImportVolumeResult;
(function (ImportVolumeResult) {
    /**
     * @internal
     */
    ImportVolumeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ImportVolumeResult = exports.ImportVolumeResult || (exports.ImportVolumeResult = {}));
var ModifyAddressAttributeRequest;
(function (ModifyAddressAttributeRequest) {
    /**
     * @internal
     */
    ModifyAddressAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyAddressAttributeRequest = exports.ModifyAddressAttributeRequest || (exports.ModifyAddressAttributeRequest = {}));
var ModifyAddressAttributeResult;
(function (ModifyAddressAttributeResult) {
    /**
     * @internal
     */
    ModifyAddressAttributeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyAddressAttributeResult = exports.ModifyAddressAttributeResult || (exports.ModifyAddressAttributeResult = {}));
var ModifyAvailabilityZoneGroupRequest;
(function (ModifyAvailabilityZoneGroupRequest) {
    /**
     * @internal
     */
    ModifyAvailabilityZoneGroupRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyAvailabilityZoneGroupRequest = exports.ModifyAvailabilityZoneGroupRequest || (exports.ModifyAvailabilityZoneGroupRequest = {}));
var ModifyAvailabilityZoneGroupResult;
(function (ModifyAvailabilityZoneGroupResult) {
    /**
     * @internal
     */
    ModifyAvailabilityZoneGroupResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyAvailabilityZoneGroupResult = exports.ModifyAvailabilityZoneGroupResult || (exports.ModifyAvailabilityZoneGroupResult = {}));
var ModifyCapacityReservationRequest;
(function (ModifyCapacityReservationRequest) {
    /**
     * @internal
     */
    ModifyCapacityReservationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyCapacityReservationRequest = exports.ModifyCapacityReservationRequest || (exports.ModifyCapacityReservationRequest = {}));
var ModifyCapacityReservationResult;
(function (ModifyCapacityReservationResult) {
    /**
     * @internal
     */
    ModifyCapacityReservationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyCapacityReservationResult = exports.ModifyCapacityReservationResult || (exports.ModifyCapacityReservationResult = {}));
var DnsServersOptionsModifyStructure;
(function (DnsServersOptionsModifyStructure) {
    /**
     * @internal
     */
    DnsServersOptionsModifyStructure.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DnsServersOptionsModifyStructure = exports.DnsServersOptionsModifyStructure || (exports.DnsServersOptionsModifyStructure = {}));
var ModifyClientVpnEndpointRequest;
(function (ModifyClientVpnEndpointRequest) {
    /**
     * @internal
     */
    ModifyClientVpnEndpointRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyClientVpnEndpointRequest = exports.ModifyClientVpnEndpointRequest || (exports.ModifyClientVpnEndpointRequest = {}));
var ModifyClientVpnEndpointResult;
(function (ModifyClientVpnEndpointResult) {
    /**
     * @internal
     */
    ModifyClientVpnEndpointResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyClientVpnEndpointResult = exports.ModifyClientVpnEndpointResult || (exports.ModifyClientVpnEndpointResult = {}));
var ModifyDefaultCreditSpecificationRequest;
(function (ModifyDefaultCreditSpecificationRequest) {
    /**
     * @internal
     */
    ModifyDefaultCreditSpecificationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyDefaultCreditSpecificationRequest = exports.ModifyDefaultCreditSpecificationRequest || (exports.ModifyDefaultCreditSpecificationRequest = {}));
var ModifyDefaultCreditSpecificationResult;
(function (ModifyDefaultCreditSpecificationResult) {
    /**
     * @internal
     */
    ModifyDefaultCreditSpecificationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyDefaultCreditSpecificationResult = exports.ModifyDefaultCreditSpecificationResult || (exports.ModifyDefaultCreditSpecificationResult = {}));
var ModifyEbsDefaultKmsKeyIdRequest;
(function (ModifyEbsDefaultKmsKeyIdRequest) {
    /**
     * @internal
     */
    ModifyEbsDefaultKmsKeyIdRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyEbsDefaultKmsKeyIdRequest = exports.ModifyEbsDefaultKmsKeyIdRequest || (exports.ModifyEbsDefaultKmsKeyIdRequest = {}));
var ModifyEbsDefaultKmsKeyIdResult;
(function (ModifyEbsDefaultKmsKeyIdResult) {
    /**
     * @internal
     */
    ModifyEbsDefaultKmsKeyIdResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyEbsDefaultKmsKeyIdResult = exports.ModifyEbsDefaultKmsKeyIdResult || (exports.ModifyEbsDefaultKmsKeyIdResult = {}));
var ModifyFleetRequest;
(function (ModifyFleetRequest) {
    /**
     * @internal
     */
    ModifyFleetRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyFleetRequest = exports.ModifyFleetRequest || (exports.ModifyFleetRequest = {}));
var ModifyFleetResult;
(function (ModifyFleetResult) {
    /**
     * @internal
     */
    ModifyFleetResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyFleetResult = exports.ModifyFleetResult || (exports.ModifyFleetResult = {}));
var LoadPermissionRequest;
(function (LoadPermissionRequest) {
    /**
     * @internal
     */
    LoadPermissionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(LoadPermissionRequest = exports.LoadPermissionRequest || (exports.LoadPermissionRequest = {}));
var LoadPermissionModifications;
(function (LoadPermissionModifications) {
    /**
     * @internal
     */
    LoadPermissionModifications.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(LoadPermissionModifications = exports.LoadPermissionModifications || (exports.LoadPermissionModifications = {}));
var ModifyFpgaImageAttributeRequest;
(function (ModifyFpgaImageAttributeRequest) {
    /**
     * @internal
     */
    ModifyFpgaImageAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyFpgaImageAttributeRequest = exports.ModifyFpgaImageAttributeRequest || (exports.ModifyFpgaImageAttributeRequest = {}));
var ModifyFpgaImageAttributeResult;
(function (ModifyFpgaImageAttributeResult) {
    /**
     * @internal
     */
    ModifyFpgaImageAttributeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyFpgaImageAttributeResult = exports.ModifyFpgaImageAttributeResult || (exports.ModifyFpgaImageAttributeResult = {}));
var ModifyHostsRequest;
(function (ModifyHostsRequest) {
    /**
     * @internal
     */
    ModifyHostsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyHostsRequest = exports.ModifyHostsRequest || (exports.ModifyHostsRequest = {}));
var ModifyHostsResult;
(function (ModifyHostsResult) {
    /**
     * @internal
     */
    ModifyHostsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyHostsResult = exports.ModifyHostsResult || (exports.ModifyHostsResult = {}));
var ModifyIdentityIdFormatRequest;
(function (ModifyIdentityIdFormatRequest) {
    /**
     * @internal
     */
    ModifyIdentityIdFormatRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyIdentityIdFormatRequest = exports.ModifyIdentityIdFormatRequest || (exports.ModifyIdentityIdFormatRequest = {}));
var ModifyIdFormatRequest;
(function (ModifyIdFormatRequest) {
    /**
     * @internal
     */
    ModifyIdFormatRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyIdFormatRequest = exports.ModifyIdFormatRequest || (exports.ModifyIdFormatRequest = {}));
var LaunchPermissionModifications;
(function (LaunchPermissionModifications) {
    /**
     * @internal
     */
    LaunchPermissionModifications.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(LaunchPermissionModifications = exports.LaunchPermissionModifications || (exports.LaunchPermissionModifications = {}));
var ModifyImageAttributeRequest;
(function (ModifyImageAttributeRequest) {
    /**
     * @internal
     */
    ModifyImageAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyImageAttributeRequest = exports.ModifyImageAttributeRequest || (exports.ModifyImageAttributeRequest = {}));
var EbsInstanceBlockDeviceSpecification;
(function (EbsInstanceBlockDeviceSpecification) {
    /**
     * @internal
     */
    EbsInstanceBlockDeviceSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EbsInstanceBlockDeviceSpecification = exports.EbsInstanceBlockDeviceSpecification || (exports.EbsInstanceBlockDeviceSpecification = {}));
var InstanceBlockDeviceMappingSpecification;
(function (InstanceBlockDeviceMappingSpecification) {
    /**
     * @internal
     */
    InstanceBlockDeviceMappingSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceBlockDeviceMappingSpecification = exports.InstanceBlockDeviceMappingSpecification || (exports.InstanceBlockDeviceMappingSpecification = {}));
var BlobAttributeValue;
(function (BlobAttributeValue) {
    /**
     * @internal
     */
    BlobAttributeValue.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(BlobAttributeValue = exports.BlobAttributeValue || (exports.BlobAttributeValue = {}));
var ModifyInstanceAttributeRequest;
(function (ModifyInstanceAttributeRequest) {
    /**
     * @internal
     */
    ModifyInstanceAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceAttributeRequest = exports.ModifyInstanceAttributeRequest || (exports.ModifyInstanceAttributeRequest = {}));
var CapacityReservationSpecification;
(function (CapacityReservationSpecification) {
    /**
     * @internal
     */
    CapacityReservationSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(CapacityReservationSpecification = exports.CapacityReservationSpecification || (exports.CapacityReservationSpecification = {}));
var ModifyInstanceCapacityReservationAttributesRequest;
(function (ModifyInstanceCapacityReservationAttributesRequest) {
    /**
     * @internal
     */
    ModifyInstanceCapacityReservationAttributesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceCapacityReservationAttributesRequest = exports.ModifyInstanceCapacityReservationAttributesRequest || (exports.ModifyInstanceCapacityReservationAttributesRequest = {}));
var ModifyInstanceCapacityReservationAttributesResult;
(function (ModifyInstanceCapacityReservationAttributesResult) {
    /**
     * @internal
     */
    ModifyInstanceCapacityReservationAttributesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceCapacityReservationAttributesResult = exports.ModifyInstanceCapacityReservationAttributesResult || (exports.ModifyInstanceCapacityReservationAttributesResult = {}));
var InstanceCreditSpecificationRequest;
(function (InstanceCreditSpecificationRequest) {
    /**
     * @internal
     */
    InstanceCreditSpecificationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceCreditSpecificationRequest = exports.InstanceCreditSpecificationRequest || (exports.InstanceCreditSpecificationRequest = {}));
var ModifyInstanceCreditSpecificationRequest;
(function (ModifyInstanceCreditSpecificationRequest) {
    /**
     * @internal
     */
    ModifyInstanceCreditSpecificationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceCreditSpecificationRequest = exports.ModifyInstanceCreditSpecificationRequest || (exports.ModifyInstanceCreditSpecificationRequest = {}));
var SuccessfulInstanceCreditSpecificationItem;
(function (SuccessfulInstanceCreditSpecificationItem) {
    /**
     * @internal
     */
    SuccessfulInstanceCreditSpecificationItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SuccessfulInstanceCreditSpecificationItem = exports.SuccessfulInstanceCreditSpecificationItem || (exports.SuccessfulInstanceCreditSpecificationItem = {}));
var UnsuccessfulInstanceCreditSpecificationErrorCode;
(function (UnsuccessfulInstanceCreditSpecificationErrorCode) {
    UnsuccessfulInstanceCreditSpecificationErrorCode["INCORRECT_INSTANCE_STATE"] = "IncorrectInstanceState";
    UnsuccessfulInstanceCreditSpecificationErrorCode["INSTANCE_CREDIT_SPECIFICATION_NOT_SUPPORTED"] = "InstanceCreditSpecification.NotSupported";
    UnsuccessfulInstanceCreditSpecificationErrorCode["INSTANCE_NOT_FOUND"] = "InvalidInstanceID.NotFound";
    UnsuccessfulInstanceCreditSpecificationErrorCode["INVALID_INSTANCE_ID"] = "InvalidInstanceID.Malformed";
})(UnsuccessfulInstanceCreditSpecificationErrorCode = exports.UnsuccessfulInstanceCreditSpecificationErrorCode || (exports.UnsuccessfulInstanceCreditSpecificationErrorCode = {}));
var UnsuccessfulInstanceCreditSpecificationItemError;
(function (UnsuccessfulInstanceCreditSpecificationItemError) {
    /**
     * @internal
     */
    UnsuccessfulInstanceCreditSpecificationItemError.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnsuccessfulInstanceCreditSpecificationItemError = exports.UnsuccessfulInstanceCreditSpecificationItemError || (exports.UnsuccessfulInstanceCreditSpecificationItemError = {}));
var UnsuccessfulInstanceCreditSpecificationItem;
(function (UnsuccessfulInstanceCreditSpecificationItem) {
    /**
     * @internal
     */
    UnsuccessfulInstanceCreditSpecificationItem.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnsuccessfulInstanceCreditSpecificationItem = exports.UnsuccessfulInstanceCreditSpecificationItem || (exports.UnsuccessfulInstanceCreditSpecificationItem = {}));
var ModifyInstanceCreditSpecificationResult;
(function (ModifyInstanceCreditSpecificationResult) {
    /**
     * @internal
     */
    ModifyInstanceCreditSpecificationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceCreditSpecificationResult = exports.ModifyInstanceCreditSpecificationResult || (exports.ModifyInstanceCreditSpecificationResult = {}));
var ModifyInstanceEventStartTimeRequest;
(function (ModifyInstanceEventStartTimeRequest) {
    /**
     * @internal
     */
    ModifyInstanceEventStartTimeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceEventStartTimeRequest = exports.ModifyInstanceEventStartTimeRequest || (exports.ModifyInstanceEventStartTimeRequest = {}));
var ModifyInstanceEventStartTimeResult;
(function (ModifyInstanceEventStartTimeResult) {
    /**
     * @internal
     */
    ModifyInstanceEventStartTimeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceEventStartTimeResult = exports.ModifyInstanceEventStartTimeResult || (exports.ModifyInstanceEventStartTimeResult = {}));
var ModifyInstanceEventWindowRequest;
(function (ModifyInstanceEventWindowRequest) {
    /**
     * @internal
     */
    ModifyInstanceEventWindowRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceEventWindowRequest = exports.ModifyInstanceEventWindowRequest || (exports.ModifyInstanceEventWindowRequest = {}));
var ModifyInstanceEventWindowResult;
(function (ModifyInstanceEventWindowResult) {
    /**
     * @internal
     */
    ModifyInstanceEventWindowResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceEventWindowResult = exports.ModifyInstanceEventWindowResult || (exports.ModifyInstanceEventWindowResult = {}));
var ModifyInstanceMetadataOptionsRequest;
(function (ModifyInstanceMetadataOptionsRequest) {
    /**
     * @internal
     */
    ModifyInstanceMetadataOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceMetadataOptionsRequest = exports.ModifyInstanceMetadataOptionsRequest || (exports.ModifyInstanceMetadataOptionsRequest = {}));
var ModifyInstanceMetadataOptionsResult;
(function (ModifyInstanceMetadataOptionsResult) {
    /**
     * @internal
     */
    ModifyInstanceMetadataOptionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstanceMetadataOptionsResult = exports.ModifyInstanceMetadataOptionsResult || (exports.ModifyInstanceMetadataOptionsResult = {}));
var ModifyInstancePlacementRequest;
(function (ModifyInstancePlacementRequest) {
    /**
     * @internal
     */
    ModifyInstancePlacementRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstancePlacementRequest = exports.ModifyInstancePlacementRequest || (exports.ModifyInstancePlacementRequest = {}));
var ModifyInstancePlacementResult;
(function (ModifyInstancePlacementResult) {
    /**
     * @internal
     */
    ModifyInstancePlacementResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyInstancePlacementResult = exports.ModifyInstancePlacementResult || (exports.ModifyInstancePlacementResult = {}));
var ModifyLaunchTemplateRequest;
(function (ModifyLaunchTemplateRequest) {
    /**
     * @internal
     */
    ModifyLaunchTemplateRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyLaunchTemplateRequest = exports.ModifyLaunchTemplateRequest || (exports.ModifyLaunchTemplateRequest = {}));
var ModifyLaunchTemplateResult;
(function (ModifyLaunchTemplateResult) {
    /**
     * @internal
     */
    ModifyLaunchTemplateResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyLaunchTemplateResult = exports.ModifyLaunchTemplateResult || (exports.ModifyLaunchTemplateResult = {}));
var RemovePrefixListEntry;
(function (RemovePrefixListEntry) {
    /**
     * @internal
     */
    RemovePrefixListEntry.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RemovePrefixListEntry = exports.RemovePrefixListEntry || (exports.RemovePrefixListEntry = {}));
var ModifyManagedPrefixListRequest;
(function (ModifyManagedPrefixListRequest) {
    /**
     * @internal
     */
    ModifyManagedPrefixListRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyManagedPrefixListRequest = exports.ModifyManagedPrefixListRequest || (exports.ModifyManagedPrefixListRequest = {}));
var ModifyManagedPrefixListResult;
(function (ModifyManagedPrefixListResult) {
    /**
     * @internal
     */
    ModifyManagedPrefixListResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyManagedPrefixListResult = exports.ModifyManagedPrefixListResult || (exports.ModifyManagedPrefixListResult = {}));
var NetworkInterfaceAttachmentChanges;
(function (NetworkInterfaceAttachmentChanges) {
    /**
     * @internal
     */
    NetworkInterfaceAttachmentChanges.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(NetworkInterfaceAttachmentChanges = exports.NetworkInterfaceAttachmentChanges || (exports.NetworkInterfaceAttachmentChanges = {}));
var ModifyNetworkInterfaceAttributeRequest;
(function (ModifyNetworkInterfaceAttributeRequest) {
    /**
     * @internal
     */
    ModifyNetworkInterfaceAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyNetworkInterfaceAttributeRequest = exports.ModifyNetworkInterfaceAttributeRequest || (exports.ModifyNetworkInterfaceAttributeRequest = {}));
var ModifyReservedInstancesRequest;
(function (ModifyReservedInstancesRequest) {
    /**
     * @internal
     */
    ModifyReservedInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyReservedInstancesRequest = exports.ModifyReservedInstancesRequest || (exports.ModifyReservedInstancesRequest = {}));
var ModifyReservedInstancesResult;
(function (ModifyReservedInstancesResult) {
    /**
     * @internal
     */
    ModifyReservedInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyReservedInstancesResult = exports.ModifyReservedInstancesResult || (exports.ModifyReservedInstancesResult = {}));
var SecurityGroupRuleRequest;
(function (SecurityGroupRuleRequest) {
    /**
     * @internal
     */
    SecurityGroupRuleRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SecurityGroupRuleRequest = exports.SecurityGroupRuleRequest || (exports.SecurityGroupRuleRequest = {}));
var SecurityGroupRuleUpdate;
(function (SecurityGroupRuleUpdate) {
    /**
     * @internal
     */
    SecurityGroupRuleUpdate.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SecurityGroupRuleUpdate = exports.SecurityGroupRuleUpdate || (exports.SecurityGroupRuleUpdate = {}));
var ModifySecurityGroupRulesRequest;
(function (ModifySecurityGroupRulesRequest) {
    /**
     * @internal
     */
    ModifySecurityGroupRulesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifySecurityGroupRulesRequest = exports.ModifySecurityGroupRulesRequest || (exports.ModifySecurityGroupRulesRequest = {}));
var ModifySecurityGroupRulesResult;
(function (ModifySecurityGroupRulesResult) {
    /**
     * @internal
     */
    ModifySecurityGroupRulesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifySecurityGroupRulesResult = exports.ModifySecurityGroupRulesResult || (exports.ModifySecurityGroupRulesResult = {}));
var CreateVolumePermissionModifications;
(function (CreateVolumePermissionModifications) {
    /**
     * @internal
     */
    CreateVolumePermissionModifications.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(CreateVolumePermissionModifications = exports.CreateVolumePermissionModifications || (exports.CreateVolumePermissionModifications = {}));
var ModifySnapshotAttributeRequest;
(function (ModifySnapshotAttributeRequest) {
    /**
     * @internal
     */
    ModifySnapshotAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifySnapshotAttributeRequest = exports.ModifySnapshotAttributeRequest || (exports.ModifySnapshotAttributeRequest = {}));
var ModifySpotFleetRequestRequest;
(function (ModifySpotFleetRequestRequest) {
    /**
     * @internal
     */
    ModifySpotFleetRequestRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifySpotFleetRequestRequest = exports.ModifySpotFleetRequestRequest || (exports.ModifySpotFleetRequestRequest = {}));
var ModifySpotFleetRequestResponse;
(function (ModifySpotFleetRequestResponse) {
    /**
     * @internal
     */
    ModifySpotFleetRequestResponse.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifySpotFleetRequestResponse = exports.ModifySpotFleetRequestResponse || (exports.ModifySpotFleetRequestResponse = {}));
var ModifySubnetAttributeRequest;
(function (ModifySubnetAttributeRequest) {
    /**
     * @internal
     */
    ModifySubnetAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifySubnetAttributeRequest = exports.ModifySubnetAttributeRequest || (exports.ModifySubnetAttributeRequest = {}));
var ModifyTrafficMirrorFilterNetworkServicesRequest;
(function (ModifyTrafficMirrorFilterNetworkServicesRequest) {
    /**
     * @internal
     */
    ModifyTrafficMirrorFilterNetworkServicesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTrafficMirrorFilterNetworkServicesRequest = exports.ModifyTrafficMirrorFilterNetworkServicesRequest || (exports.ModifyTrafficMirrorFilterNetworkServicesRequest = {}));
var ModifyTrafficMirrorFilterNetworkServicesResult;
(function (ModifyTrafficMirrorFilterNetworkServicesResult) {
    /**
     * @internal
     */
    ModifyTrafficMirrorFilterNetworkServicesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTrafficMirrorFilterNetworkServicesResult = exports.ModifyTrafficMirrorFilterNetworkServicesResult || (exports.ModifyTrafficMirrorFilterNetworkServicesResult = {}));
var ModifyTrafficMirrorFilterRuleRequest;
(function (ModifyTrafficMirrorFilterRuleRequest) {
    /**
     * @internal
     */
    ModifyTrafficMirrorFilterRuleRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTrafficMirrorFilterRuleRequest = exports.ModifyTrafficMirrorFilterRuleRequest || (exports.ModifyTrafficMirrorFilterRuleRequest = {}));
var ModifyTrafficMirrorFilterRuleResult;
(function (ModifyTrafficMirrorFilterRuleResult) {
    /**
     * @internal
     */
    ModifyTrafficMirrorFilterRuleResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTrafficMirrorFilterRuleResult = exports.ModifyTrafficMirrorFilterRuleResult || (exports.ModifyTrafficMirrorFilterRuleResult = {}));
var ModifyTrafficMirrorSessionRequest;
(function (ModifyTrafficMirrorSessionRequest) {
    /**
     * @internal
     */
    ModifyTrafficMirrorSessionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTrafficMirrorSessionRequest = exports.ModifyTrafficMirrorSessionRequest || (exports.ModifyTrafficMirrorSessionRequest = {}));
var ModifyTrafficMirrorSessionResult;
(function (ModifyTrafficMirrorSessionResult) {
    /**
     * @internal
     */
    ModifyTrafficMirrorSessionResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTrafficMirrorSessionResult = exports.ModifyTrafficMirrorSessionResult || (exports.ModifyTrafficMirrorSessionResult = {}));
var ModifyTransitGatewayOptions;
(function (ModifyTransitGatewayOptions) {
    /**
     * @internal
     */
    ModifyTransitGatewayOptions.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayOptions = exports.ModifyTransitGatewayOptions || (exports.ModifyTransitGatewayOptions = {}));
var ModifyTransitGatewayRequest;
(function (ModifyTransitGatewayRequest) {
    /**
     * @internal
     */
    ModifyTransitGatewayRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayRequest = exports.ModifyTransitGatewayRequest || (exports.ModifyTransitGatewayRequest = {}));
var ModifyTransitGatewayResult;
(function (ModifyTransitGatewayResult) {
    /**
     * @internal
     */
    ModifyTransitGatewayResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayResult = exports.ModifyTransitGatewayResult || (exports.ModifyTransitGatewayResult = {}));
var ModifyTransitGatewayPrefixListReferenceRequest;
(function (ModifyTransitGatewayPrefixListReferenceRequest) {
    /**
     * @internal
     */
    ModifyTransitGatewayPrefixListReferenceRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayPrefixListReferenceRequest = exports.ModifyTransitGatewayPrefixListReferenceRequest || (exports.ModifyTransitGatewayPrefixListReferenceRequest = {}));
var ModifyTransitGatewayPrefixListReferenceResult;
(function (ModifyTransitGatewayPrefixListReferenceResult) {
    /**
     * @internal
     */
    ModifyTransitGatewayPrefixListReferenceResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayPrefixListReferenceResult = exports.ModifyTransitGatewayPrefixListReferenceResult || (exports.ModifyTransitGatewayPrefixListReferenceResult = {}));
var ModifyTransitGatewayVpcAttachmentRequestOptions;
(function (ModifyTransitGatewayVpcAttachmentRequestOptions) {
    /**
     * @internal
     */
    ModifyTransitGatewayVpcAttachmentRequestOptions.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayVpcAttachmentRequestOptions = exports.ModifyTransitGatewayVpcAttachmentRequestOptions || (exports.ModifyTransitGatewayVpcAttachmentRequestOptions = {}));
var ModifyTransitGatewayVpcAttachmentRequest;
(function (ModifyTransitGatewayVpcAttachmentRequest) {
    /**
     * @internal
     */
    ModifyTransitGatewayVpcAttachmentRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayVpcAttachmentRequest = exports.ModifyTransitGatewayVpcAttachmentRequest || (exports.ModifyTransitGatewayVpcAttachmentRequest = {}));
var ModifyTransitGatewayVpcAttachmentResult;
(function (ModifyTransitGatewayVpcAttachmentResult) {
    /**
     * @internal
     */
    ModifyTransitGatewayVpcAttachmentResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyTransitGatewayVpcAttachmentResult = exports.ModifyTransitGatewayVpcAttachmentResult || (exports.ModifyTransitGatewayVpcAttachmentResult = {}));
var ModifyVolumeRequest;
(function (ModifyVolumeRequest) {
    /**
     * @internal
     */
    ModifyVolumeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVolumeRequest = exports.ModifyVolumeRequest || (exports.ModifyVolumeRequest = {}));
var ModifyVolumeResult;
(function (ModifyVolumeResult) {
    /**
     * @internal
     */
    ModifyVolumeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVolumeResult = exports.ModifyVolumeResult || (exports.ModifyVolumeResult = {}));
var ModifyVolumeAttributeRequest;
(function (ModifyVolumeAttributeRequest) {
    /**
     * @internal
     */
    ModifyVolumeAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVolumeAttributeRequest = exports.ModifyVolumeAttributeRequest || (exports.ModifyVolumeAttributeRequest = {}));
var ModifyVpcAttributeRequest;
(function (ModifyVpcAttributeRequest) {
    /**
     * @internal
     */
    ModifyVpcAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcAttributeRequest = exports.ModifyVpcAttributeRequest || (exports.ModifyVpcAttributeRequest = {}));
var ModifyVpcEndpointRequest;
(function (ModifyVpcEndpointRequest) {
    /**
     * @internal
     */
    ModifyVpcEndpointRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointRequest = exports.ModifyVpcEndpointRequest || (exports.ModifyVpcEndpointRequest = {}));
var ModifyVpcEndpointResult;
(function (ModifyVpcEndpointResult) {
    /**
     * @internal
     */
    ModifyVpcEndpointResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointResult = exports.ModifyVpcEndpointResult || (exports.ModifyVpcEndpointResult = {}));
var ModifyVpcEndpointConnectionNotificationRequest;
(function (ModifyVpcEndpointConnectionNotificationRequest) {
    /**
     * @internal
     */
    ModifyVpcEndpointConnectionNotificationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointConnectionNotificationRequest = exports.ModifyVpcEndpointConnectionNotificationRequest || (exports.ModifyVpcEndpointConnectionNotificationRequest = {}));
var ModifyVpcEndpointConnectionNotificationResult;
(function (ModifyVpcEndpointConnectionNotificationResult) {
    /**
     * @internal
     */
    ModifyVpcEndpointConnectionNotificationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointConnectionNotificationResult = exports.ModifyVpcEndpointConnectionNotificationResult || (exports.ModifyVpcEndpointConnectionNotificationResult = {}));
