// @aws-sdk/client-ec2/dist-cjs/models/models_5.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RejectTransitGatewayMulticastDomainAssociationsResult = exports.RejectTransitGatewayMulticastDomainAssociationsRequest = exports.RegisterTransitGatewayMulticastGroupSourcesResult = exports.TransitGatewayMulticastRegisteredGroupSources = exports.RegisterTransitGatewayMulticastGroupSourcesRequest = exports.RegisterTransitGatewayMulticastGroupMembersResult = exports.TransitGatewayMulticastRegisteredGroupMembers = exports.RegisterTransitGatewayMulticastGroupMembersRequest = exports.RegisterInstanceEventNotificationAttributesResult = exports.RegisterInstanceEventNotificationAttributesRequest = exports.RegisterInstanceTagAttributeRequest = exports.RegisterImageResult = exports.RegisterImageRequest = exports.RebootInstancesRequest = exports.PurchaseScheduledInstancesResult = exports.PurchaseScheduledInstancesRequest = exports.PurchaseRequest = exports.PurchaseReservedInstancesOfferingResult = exports.PurchaseReservedInstancesOfferingRequest = exports.ReservedInstanceLimitPrice = exports.PurchaseHostReservationResult = exports.PurchaseHostReservationRequest = exports.ProvisionByoipCidrResult = exports.ProvisionByoipCidrRequest = exports.CidrAuthorizationContext = exports.MoveAddressToVpcResult = exports.Status = exports.MoveAddressToVpcRequest = exports.MonitorInstancesResult = exports.InstanceMonitoring = exports.MonitorInstancesRequest = exports.ModifyVpnTunnelOptionsResult = exports.ModifyVpnTunnelOptionsRequest = exports.ModifyVpnTunnelOptionsSpecification = exports.ModifyVpnTunnelCertificateResult = exports.ModifyVpnTunnelCertificateRequest = exports.ModifyVpnConnectionOptionsResult = exports.ModifyVpnConnectionOptionsRequest = exports.ModifyVpnConnectionResult = exports.ModifyVpnConnectionRequest = exports.ModifyVpcTenancyResult = exports.ModifyVpcTenancyRequest = exports.ModifyVpcPeeringConnectionOptionsResult = exports.PeeringConnectionOptions = exports.ModifyVpcPeeringConnectionOptionsRequest = exports.PeeringConnectionOptionsRequest = exports.ModifyVpcEndpointServicePermissionsResult = exports.ModifyVpcEndpointServicePermissionsRequest = exports.ModifyVpcEndpointServiceConfigurationResult = exports.ModifyVpcEndpointServiceConfigurationRequest = void 0;
exports.EnclaveOptionsRequest = exports.ElasticInferenceAccelerator = exports.CpuOptionsRequest = exports.RevokeSecurityGroupIngressResult = exports.RevokeSecurityGroupIngressRequest = exports.RevokeSecurityGroupEgressResult = exports.RevokeSecurityGroupEgressRequest = exports.RevokeClientVpnIngressResult = exports.RevokeClientVpnIngressRequest = exports.RestoreManagedPrefixListVersionResult = exports.RestoreManagedPrefixListVersionRequest = exports.RestoreAddressToClassicResult = exports.RestoreAddressToClassicRequest = exports.ResetSnapshotAttributeRequest = exports.ResetNetworkInterfaceAttributeRequest = exports.ResetInstanceAttributeRequest = exports.ResetImageAttributeRequest = exports.ResetFpgaImageAttributeResult = exports.ResetFpgaImageAttributeRequest = exports.ResetEbsDefaultKmsKeyIdResult = exports.ResetEbsDefaultKmsKeyIdRequest = exports.ResetAddressAttributeResult = exports.ResetAddressAttributeRequest = exports.RequestSpotInstancesResult = exports.RequestSpotInstancesRequest = exports.RequestSpotLaunchSpecification = exports.RequestSpotFleetResponse = exports.RequestSpotFleetRequest = exports.ReportInstanceStatusRequest = exports.ReplaceTransitGatewayRouteResult = exports.ReplaceTransitGatewayRouteRequest = exports.ReplaceRouteTableAssociationResult = exports.ReplaceRouteTableAssociationRequest = exports.ReplaceRouteRequest = exports.ReplaceNetworkAclEntryRequest = exports.ReplaceNetworkAclAssociationResult = exports.ReplaceNetworkAclAssociationRequest = exports.ReplaceIamInstanceProfileAssociationResult = exports.ReplaceIamInstanceProfileAssociationRequest = exports.ReleaseHostsResult = exports.ReleaseHostsRequest = exports.ReleaseAddressRequest = exports.RejectVpcPeeringConnectionResult = exports.RejectVpcPeeringConnectionRequest = exports.RejectVpcEndpointConnectionsResult = exports.RejectVpcEndpointConnectionsRequest = exports.RejectTransitGatewayVpcAttachmentResult = exports.RejectTransitGatewayVpcAttachmentRequest = exports.RejectTransitGatewayPeeringAttachmentResult = exports.RejectTransitGatewayPeeringAttachmentRequest = void 0;
exports.UpdateSecurityGroupRuleDescriptionsIngressResult = exports.UpdateSecurityGroupRuleDescriptionsIngressRequest = exports.UpdateSecurityGroupRuleDescriptionsEgressResult = exports.UpdateSecurityGroupRuleDescriptionsEgressRequest = exports.SecurityGroupRuleDescription = exports.UnmonitorInstancesResult = exports.UnmonitorInstancesRequest = exports.UnassignPrivateIpAddressesRequest = exports.UnassignIpv6AddressesResult = exports.UnassignIpv6AddressesRequest = exports.TerminateInstancesResult = exports.TerminateInstancesRequest = exports.TerminateClientVpnConnectionsResult = exports.TerminateConnectionStatus = exports.TerminateClientVpnConnectionsRequest = exports.StopInstancesResult = exports.StopInstancesRequest = exports.StartVpcEndpointServicePrivateDnsVerificationResult = exports.StartVpcEndpointServicePrivateDnsVerificationRequest = exports.StartNetworkInsightsAnalysisResult = exports.StartNetworkInsightsAnalysisRequest = exports.StartInstancesResult = exports.InstanceStateChange = exports.StartInstancesRequest = exports.SendDiagnosticInterruptRequest = exports.SearchTransitGatewayRoutesResult = exports.SearchTransitGatewayRoutesRequest = exports.SearchTransitGatewayMulticastGroupsResult = exports.TransitGatewayMulticastGroup = exports.SearchTransitGatewayMulticastGroupsRequest = exports.SearchLocalGatewayRoutesResult = exports.SearchLocalGatewayRoutesRequest = exports.RunScheduledInstancesResult = exports.RunScheduledInstancesRequest = exports.ScheduledInstancesLaunchSpecification = exports.ScheduledInstancesPlacement = exports.ScheduledInstancesNetworkInterface = exports.ScheduledInstancesPrivateIpAddressConfig = exports.ScheduledInstancesIpv6Address = exports.ScheduledInstancesMonitoring = exports.ScheduledInstancesIamInstanceProfile = exports.ScheduledInstancesBlockDeviceMapping = exports.ScheduledInstancesEbs = exports.RunInstancesRequest = exports.InstanceMetadataOptionsRequest = exports.LicenseConfigurationRequest = exports.LaunchTemplateSpecification = exports.InstanceMarketOptionsRequest = exports.SpotMarketOptions = exports.HibernationOptionsRequest = void 0;
exports.WithdrawByoipCidrResult = exports.WithdrawByoipCidrRequest = void 0;
var ModifyVpcEndpointServiceConfigurationRequest;
(function (ModifyVpcEndpointServiceConfigurationRequest) {
    /**
     * @internal
     */
    ModifyVpcEndpointServiceConfigurationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointServiceConfigurationRequest = exports.ModifyVpcEndpointServiceConfigurationRequest || (exports.ModifyVpcEndpointServiceConfigurationRequest = {}));
var ModifyVpcEndpointServiceConfigurationResult;
(function (ModifyVpcEndpointServiceConfigurationResult) {
    /**
     * @internal
     */
    ModifyVpcEndpointServiceConfigurationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointServiceConfigurationResult = exports.ModifyVpcEndpointServiceConfigurationResult || (exports.ModifyVpcEndpointServiceConfigurationResult = {}));
var ModifyVpcEndpointServicePermissionsRequest;
(function (ModifyVpcEndpointServicePermissionsRequest) {
    /**
     * @internal
     */
    ModifyVpcEndpointServicePermissionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointServicePermissionsRequest = exports.ModifyVpcEndpointServicePermissionsRequest || (exports.ModifyVpcEndpointServicePermissionsRequest = {}));
var ModifyVpcEndpointServicePermissionsResult;
(function (ModifyVpcEndpointServicePermissionsResult) {
    /**
     * @internal
     */
    ModifyVpcEndpointServicePermissionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcEndpointServicePermissionsResult = exports.ModifyVpcEndpointServicePermissionsResult || (exports.ModifyVpcEndpointServicePermissionsResult = {}));
var PeeringConnectionOptionsRequest;
(function (PeeringConnectionOptionsRequest) {
    /**
     * @internal
     */
    PeeringConnectionOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PeeringConnectionOptionsRequest = exports.PeeringConnectionOptionsRequest || (exports.PeeringConnectionOptionsRequest = {}));
var ModifyVpcPeeringConnectionOptionsRequest;
(function (ModifyVpcPeeringConnectionOptionsRequest) {
    /**
     * @internal
     */
    ModifyVpcPeeringConnectionOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcPeeringConnectionOptionsRequest = exports.ModifyVpcPeeringConnectionOptionsRequest || (exports.ModifyVpcPeeringConnectionOptionsRequest = {}));
var PeeringConnectionOptions;
(function (PeeringConnectionOptions) {
    /**
     * @internal
     */
    PeeringConnectionOptions.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PeeringConnectionOptions = exports.PeeringConnectionOptions || (exports.PeeringConnectionOptions = {}));
var ModifyVpcPeeringConnectionOptionsResult;
(function (ModifyVpcPeeringConnectionOptionsResult) {
    /**
     * @internal
     */
    ModifyVpcPeeringConnectionOptionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcPeeringConnectionOptionsResult = exports.ModifyVpcPeeringConnectionOptionsResult || (exports.ModifyVpcPeeringConnectionOptionsResult = {}));
var ModifyVpcTenancyRequest;
(function (ModifyVpcTenancyRequest) {
    /**
     * @internal
     */
    ModifyVpcTenancyRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcTenancyRequest = exports.ModifyVpcTenancyRequest || (exports.ModifyVpcTenancyRequest = {}));
var ModifyVpcTenancyResult;
(function (ModifyVpcTenancyResult) {
    /**
     * @internal
     */
    ModifyVpcTenancyResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpcTenancyResult = exports.ModifyVpcTenancyResult || (exports.ModifyVpcTenancyResult = {}));
var ModifyVpnConnectionRequest;
(function (ModifyVpnConnectionRequest) {
    /**
     * @internal
     */
    ModifyVpnConnectionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnConnectionRequest = exports.ModifyVpnConnectionRequest || (exports.ModifyVpnConnectionRequest = {}));
var ModifyVpnConnectionResult;
(function (ModifyVpnConnectionResult) {
    /**
     * @internal
     */
    ModifyVpnConnectionResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnConnectionResult = exports.ModifyVpnConnectionResult || (exports.ModifyVpnConnectionResult = {}));
var ModifyVpnConnectionOptionsRequest;
(function (ModifyVpnConnectionOptionsRequest) {
    /**
     * @internal
     */
    ModifyVpnConnectionOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnConnectionOptionsRequest = exports.ModifyVpnConnectionOptionsRequest || (exports.ModifyVpnConnectionOptionsRequest = {}));
var ModifyVpnConnectionOptionsResult;
(function (ModifyVpnConnectionOptionsResult) {
    /**
     * @internal
     */
    ModifyVpnConnectionOptionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnConnectionOptionsResult = exports.ModifyVpnConnectionOptionsResult || (exports.ModifyVpnConnectionOptionsResult = {}));
var ModifyVpnTunnelCertificateRequest;
(function (ModifyVpnTunnelCertificateRequest) {
    /**
     * @internal
     */
    ModifyVpnTunnelCertificateRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnTunnelCertificateRequest = exports.ModifyVpnTunnelCertificateRequest || (exports.ModifyVpnTunnelCertificateRequest = {}));
var ModifyVpnTunnelCertificateResult;
(function (ModifyVpnTunnelCertificateResult) {
    /**
     * @internal
     */
    ModifyVpnTunnelCertificateResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnTunnelCertificateResult = exports.ModifyVpnTunnelCertificateResult || (exports.ModifyVpnTunnelCertificateResult = {}));
var ModifyVpnTunnelOptionsSpecification;
(function (ModifyVpnTunnelOptionsSpecification) {
    /**
     * @internal
     */
    ModifyVpnTunnelOptionsSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnTunnelOptionsSpecification = exports.ModifyVpnTunnelOptionsSpecification || (exports.ModifyVpnTunnelOptionsSpecification = {}));
var ModifyVpnTunnelOptionsRequest;
(function (ModifyVpnTunnelOptionsRequest) {
    /**
     * @internal
     */
    ModifyVpnTunnelOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnTunnelOptionsRequest = exports.ModifyVpnTunnelOptionsRequest || (exports.ModifyVpnTunnelOptionsRequest = {}));
var ModifyVpnTunnelOptionsResult;
(function (ModifyVpnTunnelOptionsResult) {
    /**
     * @internal
     */
    ModifyVpnTunnelOptionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ModifyVpnTunnelOptionsResult = exports.ModifyVpnTunnelOptionsResult || (exports.ModifyVpnTunnelOptionsResult = {}));
var MonitorInstancesRequest;
(function (MonitorInstancesRequest) {
    /**
     * @internal
     */
    MonitorInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(MonitorInstancesRequest = exports.MonitorInstancesRequest || (exports.MonitorInstancesRequest = {}));
var InstanceMonitoring;
(function (InstanceMonitoring) {
    /**
     * @internal
     */
    InstanceMonitoring.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceMonitoring = exports.InstanceMonitoring || (exports.InstanceMonitoring = {}));
var MonitorInstancesResult;
(function (MonitorInstancesResult) {
    /**
     * @internal
     */
    MonitorInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(MonitorInstancesResult = exports.MonitorInstancesResult || (exports.MonitorInstancesResult = {}));
var MoveAddressToVpcRequest;
(function (MoveAddressToVpcRequest) {
    /**
     * @internal
     */
    MoveAddressToVpcRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(MoveAddressToVpcRequest = exports.MoveAddressToVpcRequest || (exports.MoveAddressToVpcRequest = {}));
var Status;
(function (Status) {
    Status["inClassic"] = "InClassic";
    Status["inVpc"] = "InVpc";
    Status["moveInProgress"] = "MoveInProgress";
})(Status = exports.Status || (exports.Status = {}));
var MoveAddressToVpcResult;
(function (MoveAddressToVpcResult) {
    /**
     * @internal
     */
    MoveAddressToVpcResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(MoveAddressToVpcResult = exports.MoveAddressToVpcResult || (exports.MoveAddressToVpcResult = {}));
var CidrAuthorizationContext;
(function (CidrAuthorizationContext) {
    /**
     * @internal
     */
    CidrAuthorizationContext.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(CidrAuthorizationContext = exports.CidrAuthorizationContext || (exports.CidrAuthorizationContext = {}));
var ProvisionByoipCidrRequest;
(function (ProvisionByoipCidrRequest) {
    /**
     * @internal
     */
    ProvisionByoipCidrRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ProvisionByoipCidrRequest = exports.ProvisionByoipCidrRequest || (exports.ProvisionByoipCidrRequest = {}));
var ProvisionByoipCidrResult;
(function (ProvisionByoipCidrResult) {
    /**
     * @internal
     */
    ProvisionByoipCidrResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ProvisionByoipCidrResult = exports.ProvisionByoipCidrResult || (exports.ProvisionByoipCidrResult = {}));
var PurchaseHostReservationRequest;
(function (PurchaseHostReservationRequest) {
    /**
     * @internal
     */
    PurchaseHostReservationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseHostReservationRequest = exports.PurchaseHostReservationRequest || (exports.PurchaseHostReservationRequest = {}));
var PurchaseHostReservationResult;
(function (PurchaseHostReservationResult) {
    /**
     * @internal
     */
    PurchaseHostReservationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseHostReservationResult = exports.PurchaseHostReservationResult || (exports.PurchaseHostReservationResult = {}));
var ReservedInstanceLimitPrice;
(function (ReservedInstanceLimitPrice) {
    /**
     * @internal
     */
    ReservedInstanceLimitPrice.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReservedInstanceLimitPrice = exports.ReservedInstanceLimitPrice || (exports.ReservedInstanceLimitPrice = {}));
var PurchaseReservedInstancesOfferingRequest;
(function (PurchaseReservedInstancesOfferingRequest) {
    /**
     * @internal
     */
    PurchaseReservedInstancesOfferingRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseReservedInstancesOfferingRequest = exports.PurchaseReservedInstancesOfferingRequest || (exports.PurchaseReservedInstancesOfferingRequest = {}));
var PurchaseReservedInstancesOfferingResult;
(function (PurchaseReservedInstancesOfferingResult) {
    /**
     * @internal
     */
    PurchaseReservedInstancesOfferingResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseReservedInstancesOfferingResult = exports.PurchaseReservedInstancesOfferingResult || (exports.PurchaseReservedInstancesOfferingResult = {}));
var PurchaseRequest;
(function (PurchaseRequest) {
    /**
     * @internal
     */
    PurchaseRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseRequest = exports.PurchaseRequest || (exports.PurchaseRequest = {}));
var PurchaseScheduledInstancesRequest;
(function (PurchaseScheduledInstancesRequest) {
    /**
     * @internal
     */
    PurchaseScheduledInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseScheduledInstancesRequest = exports.PurchaseScheduledInstancesRequest || (exports.PurchaseScheduledInstancesRequest = {}));
var PurchaseScheduledInstancesResult;
(function (PurchaseScheduledInstancesResult) {
    /**
     * @internal
     */
    PurchaseScheduledInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PurchaseScheduledInstancesResult = exports.PurchaseScheduledInstancesResult || (exports.PurchaseScheduledInstancesResult = {}));
var RebootInstancesRequest;
(function (RebootInstancesRequest) {
    /**
     * @internal
     */
    RebootInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RebootInstancesRequest = exports.RebootInstancesRequest || (exports.RebootInstancesRequest = {}));
var RegisterImageRequest;
(function (RegisterImageRequest) {
    /**
     * @internal
     */
    RegisterImageRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterImageRequest = exports.RegisterImageRequest || (exports.RegisterImageRequest = {}));
var RegisterImageResult;
(function (RegisterImageResult) {
    /**
     * @internal
     */
    RegisterImageResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterImageResult = exports.RegisterImageResult || (exports.RegisterImageResult = {}));
var RegisterInstanceTagAttributeRequest;
(function (RegisterInstanceTagAttributeRequest) {
    /**
     * @internal
     */
    RegisterInstanceTagAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterInstanceTagAttributeRequest = exports.RegisterInstanceTagAttributeRequest || (exports.RegisterInstanceTagAttributeRequest = {}));
var RegisterInstanceEventNotificationAttributesRequest;
(function (RegisterInstanceEventNotificationAttributesRequest) {
    /**
     * @internal
     */
    RegisterInstanceEventNotificationAttributesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterInstanceEventNotificationAttributesRequest = exports.RegisterInstanceEventNotificationAttributesRequest || (exports.RegisterInstanceEventNotificationAttributesRequest = {}));
var RegisterInstanceEventNotificationAttributesResult;
(function (RegisterInstanceEventNotificationAttributesResult) {
    /**
     * @internal
     */
    RegisterInstanceEventNotificationAttributesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterInstanceEventNotificationAttributesResult = exports.RegisterInstanceEventNotificationAttributesResult || (exports.RegisterInstanceEventNotificationAttributesResult = {}));
var RegisterTransitGatewayMulticastGroupMembersRequest;
(function (RegisterTransitGatewayMulticastGroupMembersRequest) {
    /**
     * @internal
     */
    RegisterTransitGatewayMulticastGroupMembersRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterTransitGatewayMulticastGroupMembersRequest = exports.RegisterTransitGatewayMulticastGroupMembersRequest || (exports.RegisterTransitGatewayMulticastGroupMembersRequest = {}));
var TransitGatewayMulticastRegisteredGroupMembers;
(function (TransitGatewayMulticastRegisteredGroupMembers) {
    /**
     * @internal
     */
    TransitGatewayMulticastRegisteredGroupMembers.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayMulticastRegisteredGroupMembers = exports.TransitGatewayMulticastRegisteredGroupMembers || (exports.TransitGatewayMulticastRegisteredGroupMembers = {}));
var RegisterTransitGatewayMulticastGroupMembersResult;
(function (RegisterTransitGatewayMulticastGroupMembersResult) {
    /**
     * @internal
     */
    RegisterTransitGatewayMulticastGroupMembersResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterTransitGatewayMulticastGroupMembersResult = exports.RegisterTransitGatewayMulticastGroupMembersResult || (exports.RegisterTransitGatewayMulticastGroupMembersResult = {}));
var RegisterTransitGatewayMulticastGroupSourcesRequest;
(function (RegisterTransitGatewayMulticastGroupSourcesRequest) {
    /**
     * @internal
     */
    RegisterTransitGatewayMulticastGroupSourcesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterTransitGatewayMulticastGroupSourcesRequest = exports.RegisterTransitGatewayMulticastGroupSourcesRequest || (exports.RegisterTransitGatewayMulticastGroupSourcesRequest = {}));
var TransitGatewayMulticastRegisteredGroupSources;
(function (TransitGatewayMulticastRegisteredGroupSources) {
    /**
     * @internal
     */
    TransitGatewayMulticastRegisteredGroupSources.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayMulticastRegisteredGroupSources = exports.TransitGatewayMulticastRegisteredGroupSources || (exports.TransitGatewayMulticastRegisteredGroupSources = {}));
var RegisterTransitGatewayMulticastGroupSourcesResult;
(function (RegisterTransitGatewayMulticastGroupSourcesResult) {
    /**
     * @internal
     */
    RegisterTransitGatewayMulticastGroupSourcesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RegisterTransitGatewayMulticastGroupSourcesResult = exports.RegisterTransitGatewayMulticastGroupSourcesResult || (exports.RegisterTransitGatewayMulticastGroupSourcesResult = {}));
var RejectTransitGatewayMulticastDomainAssociationsRequest;
(function (RejectTransitGatewayMulticastDomainAssociationsRequest) {
    /**
     * @internal
     */
    RejectTransitGatewayMulticastDomainAssociationsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectTransitGatewayMulticastDomainAssociationsRequest = exports.RejectTransitGatewayMulticastDomainAssociationsRequest || (exports.RejectTransitGatewayMulticastDomainAssociationsRequest = {}));
var RejectTransitGatewayMulticastDomainAssociationsResult;
(function (RejectTransitGatewayMulticastDomainAssociationsResult) {
    /**
     * @internal
     */
    RejectTransitGatewayMulticastDomainAssociationsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectTransitGatewayMulticastDomainAssociationsResult = exports.RejectTransitGatewayMulticastDomainAssociationsResult || (exports.RejectTransitGatewayMulticastDomainAssociationsResult = {}));
var RejectTransitGatewayPeeringAttachmentRequest;
(function (RejectTransitGatewayPeeringAttachmentRequest) {
    /**
     * @internal
     */
    RejectTransitGatewayPeeringAttachmentRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectTransitGatewayPeeringAttachmentRequest = exports.RejectTransitGatewayPeeringAttachmentRequest || (exports.RejectTransitGatewayPeeringAttachmentRequest = {}));
var RejectTransitGatewayPeeringAttachmentResult;
(function (RejectTransitGatewayPeeringAttachmentResult) {
    /**
     * @internal
     */
    RejectTransitGatewayPeeringAttachmentResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectTransitGatewayPeeringAttachmentResult = exports.RejectTransitGatewayPeeringAttachmentResult || (exports.RejectTransitGatewayPeeringAttachmentResult = {}));
var RejectTransitGatewayVpcAttachmentRequest;
(function (RejectTransitGatewayVpcAttachmentRequest) {
    /**
     * @internal
     */
    RejectTransitGatewayVpcAttachmentRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectTransitGatewayVpcAttachmentRequest = exports.RejectTransitGatewayVpcAttachmentRequest || (exports.RejectTransitGatewayVpcAttachmentRequest = {}));
var RejectTransitGatewayVpcAttachmentResult;
(function (RejectTransitGatewayVpcAttachmentResult) {
    /**
     * @internal
     */
    RejectTransitGatewayVpcAttachmentResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectTransitGatewayVpcAttachmentResult = exports.RejectTransitGatewayVpcAttachmentResult || (exports.RejectTransitGatewayVpcAttachmentResult = {}));
var RejectVpcEndpointConnectionsRequest;
(function (RejectVpcEndpointConnectionsRequest) {
    /**
     * @internal
     */
    RejectVpcEndpointConnectionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectVpcEndpointConnectionsRequest = exports.RejectVpcEndpointConnectionsRequest || (exports.RejectVpcEndpointConnectionsRequest = {}));
var RejectVpcEndpointConnectionsResult;
(function (RejectVpcEndpointConnectionsResult) {
    /**
     * @internal
     */
    RejectVpcEndpointConnectionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectVpcEndpointConnectionsResult = exports.RejectVpcEndpointConnectionsResult || (exports.RejectVpcEndpointConnectionsResult = {}));
var RejectVpcPeeringConnectionRequest;
(function (RejectVpcPeeringConnectionRequest) {
    /**
     * @internal
     */
    RejectVpcPeeringConnectionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectVpcPeeringConnectionRequest = exports.RejectVpcPeeringConnectionRequest || (exports.RejectVpcPeeringConnectionRequest = {}));
var RejectVpcPeeringConnectionResult;
(function (RejectVpcPeeringConnectionResult) {
    /**
     * @internal
     */
    RejectVpcPeeringConnectionResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RejectVpcPeeringConnectionResult = exports.RejectVpcPeeringConnectionResult || (exports.RejectVpcPeeringConnectionResult = {}));
var ReleaseAddressRequest;
(function (ReleaseAddressRequest) {
    /**
     * @internal
     */
    ReleaseAddressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReleaseAddressRequest = exports.ReleaseAddressRequest || (exports.ReleaseAddressRequest = {}));
var ReleaseHostsRequest;
(function (ReleaseHostsRequest) {
    /**
     * @internal
     */
    ReleaseHostsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReleaseHostsRequest = exports.ReleaseHostsRequest || (exports.ReleaseHostsRequest = {}));
var ReleaseHostsResult;
(function (ReleaseHostsResult) {
    /**
     * @internal
     */
    ReleaseHostsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReleaseHostsResult = exports.ReleaseHostsResult || (exports.ReleaseHostsResult = {}));
var ReplaceIamInstanceProfileAssociationRequest;
(function (ReplaceIamInstanceProfileAssociationRequest) {
    /**
     * @internal
     */
    ReplaceIamInstanceProfileAssociationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceIamInstanceProfileAssociationRequest = exports.ReplaceIamInstanceProfileAssociationRequest || (exports.ReplaceIamInstanceProfileAssociationRequest = {}));
var ReplaceIamInstanceProfileAssociationResult;
(function (ReplaceIamInstanceProfileAssociationResult) {
    /**
     * @internal
     */
    ReplaceIamInstanceProfileAssociationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceIamInstanceProfileAssociationResult = exports.ReplaceIamInstanceProfileAssociationResult || (exports.ReplaceIamInstanceProfileAssociationResult = {}));
var ReplaceNetworkAclAssociationRequest;
(function (ReplaceNetworkAclAssociationRequest) {
    /**
     * @internal
     */
    ReplaceNetworkAclAssociationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceNetworkAclAssociationRequest = exports.ReplaceNetworkAclAssociationRequest || (exports.ReplaceNetworkAclAssociationRequest = {}));
var ReplaceNetworkAclAssociationResult;
(function (ReplaceNetworkAclAssociationResult) {
    /**
     * @internal
     */
    ReplaceNetworkAclAssociationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceNetworkAclAssociationResult = exports.ReplaceNetworkAclAssociationResult || (exports.ReplaceNetworkAclAssociationResult = {}));
var ReplaceNetworkAclEntryRequest;
(function (ReplaceNetworkAclEntryRequest) {
    /**
     * @internal
     */
    ReplaceNetworkAclEntryRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceNetworkAclEntryRequest = exports.ReplaceNetworkAclEntryRequest || (exports.ReplaceNetworkAclEntryRequest = {}));
var ReplaceRouteRequest;
(function (ReplaceRouteRequest) {
    /**
     * @internal
     */
    ReplaceRouteRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceRouteRequest = exports.ReplaceRouteRequest || (exports.ReplaceRouteRequest = {}));
var ReplaceRouteTableAssociationRequest;
(function (ReplaceRouteTableAssociationRequest) {
    /**
     * @internal
     */
    ReplaceRouteTableAssociationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceRouteTableAssociationRequest = exports.ReplaceRouteTableAssociationRequest || (exports.ReplaceRouteTableAssociationRequest = {}));
var ReplaceRouteTableAssociationResult;
(function (ReplaceRouteTableAssociationResult) {
    /**
     * @internal
     */
    ReplaceRouteTableAssociationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceRouteTableAssociationResult = exports.ReplaceRouteTableAssociationResult || (exports.ReplaceRouteTableAssociationResult = {}));
var ReplaceTransitGatewayRouteRequest;
(function (ReplaceTransitGatewayRouteRequest) {
    /**
     * @internal
     */
    ReplaceTransitGatewayRouteRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceTransitGatewayRouteRequest = exports.ReplaceTransitGatewayRouteRequest || (exports.ReplaceTransitGatewayRouteRequest = {}));
var ReplaceTransitGatewayRouteResult;
(function (ReplaceTransitGatewayRouteResult) {
    /**
     * @internal
     */
    ReplaceTransitGatewayRouteResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReplaceTransitGatewayRouteResult = exports.ReplaceTransitGatewayRouteResult || (exports.ReplaceTransitGatewayRouteResult = {}));
var ReportInstanceStatusRequest;
(function (ReportInstanceStatusRequest) {
    /**
     * @internal
     */
    ReportInstanceStatusRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ReportInstanceStatusRequest = exports.ReportInstanceStatusRequest || (exports.ReportInstanceStatusRequest = {}));
var RequestSpotFleetRequest;
(function (RequestSpotFleetRequest) {
    /**
     * @internal
     */
    RequestSpotFleetRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RequestSpotFleetRequest = exports.RequestSpotFleetRequest || (exports.RequestSpotFleetRequest = {}));
var RequestSpotFleetResponse;
(function (RequestSpotFleetResponse) {
    /**
     * @internal
     */
    RequestSpotFleetResponse.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RequestSpotFleetResponse = exports.RequestSpotFleetResponse || (exports.RequestSpotFleetResponse = {}));
var RequestSpotLaunchSpecification;
(function (RequestSpotLaunchSpecification) {
    /**
     * @internal
     */
    RequestSpotLaunchSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RequestSpotLaunchSpecification = exports.RequestSpotLaunchSpecification || (exports.RequestSpotLaunchSpecification = {}));
var RequestSpotInstancesRequest;
(function (RequestSpotInstancesRequest) {
    /**
     * @internal
     */
    RequestSpotInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RequestSpotInstancesRequest = exports.RequestSpotInstancesRequest || (exports.RequestSpotInstancesRequest = {}));
var RequestSpotInstancesResult;
(function (RequestSpotInstancesResult) {
    /**
     * @internal
     */
    RequestSpotInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RequestSpotInstancesResult = exports.RequestSpotInstancesResult || (exports.RequestSpotInstancesResult = {}));
var ResetAddressAttributeRequest;
(function (ResetAddressAttributeRequest) {
    /**
     * @internal
     */
    ResetAddressAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetAddressAttributeRequest = exports.ResetAddressAttributeRequest || (exports.ResetAddressAttributeRequest = {}));
var ResetAddressAttributeResult;
(function (ResetAddressAttributeResult) {
    /**
     * @internal
     */
    ResetAddressAttributeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetAddressAttributeResult = exports.ResetAddressAttributeResult || (exports.ResetAddressAttributeResult = {}));
var ResetEbsDefaultKmsKeyIdRequest;
(function (ResetEbsDefaultKmsKeyIdRequest) {
    /**
     * @internal
     */
    ResetEbsDefaultKmsKeyIdRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetEbsDefaultKmsKeyIdRequest = exports.ResetEbsDefaultKmsKeyIdRequest || (exports.ResetEbsDefaultKmsKeyIdRequest = {}));
var ResetEbsDefaultKmsKeyIdResult;
(function (ResetEbsDefaultKmsKeyIdResult) {
    /**
     * @internal
     */
    ResetEbsDefaultKmsKeyIdResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetEbsDefaultKmsKeyIdResult = exports.ResetEbsDefaultKmsKeyIdResult || (exports.ResetEbsDefaultKmsKeyIdResult = {}));
var ResetFpgaImageAttributeRequest;
(function (ResetFpgaImageAttributeRequest) {
    /**
     * @internal
     */
    ResetFpgaImageAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetFpgaImageAttributeRequest = exports.ResetFpgaImageAttributeRequest || (exports.ResetFpgaImageAttributeRequest = {}));
var ResetFpgaImageAttributeResult;
(function (ResetFpgaImageAttributeResult) {
    /**
     * @internal
     */
    ResetFpgaImageAttributeResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetFpgaImageAttributeResult = exports.ResetFpgaImageAttributeResult || (exports.ResetFpgaImageAttributeResult = {}));
var ResetImageAttributeRequest;
(function (ResetImageAttributeRequest) {
    /**
     * @internal
     */
    ResetImageAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetImageAttributeRequest = exports.ResetImageAttributeRequest || (exports.ResetImageAttributeRequest = {}));
var ResetInstanceAttributeRequest;
(function (ResetInstanceAttributeRequest) {
    /**
     * @internal
     */
    ResetInstanceAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetInstanceAttributeRequest = exports.ResetInstanceAttributeRequest || (exports.ResetInstanceAttributeRequest = {}));
var ResetNetworkInterfaceAttributeRequest;
(function (ResetNetworkInterfaceAttributeRequest) {
    /**
     * @internal
     */
    ResetNetworkInterfaceAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetNetworkInterfaceAttributeRequest = exports.ResetNetworkInterfaceAttributeRequest || (exports.ResetNetworkInterfaceAttributeRequest = {}));
var ResetSnapshotAttributeRequest;
(function (ResetSnapshotAttributeRequest) {
    /**
     * @internal
     */
    ResetSnapshotAttributeRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ResetSnapshotAttributeRequest = exports.ResetSnapshotAttributeRequest || (exports.ResetSnapshotAttributeRequest = {}));
var RestoreAddressToClassicRequest;
(function (RestoreAddressToClassicRequest) {
    /**
     * @internal
     */
    RestoreAddressToClassicRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RestoreAddressToClassicRequest = exports.RestoreAddressToClassicRequest || (exports.RestoreAddressToClassicRequest = {}));
var RestoreAddressToClassicResult;
(function (RestoreAddressToClassicResult) {
    /**
     * @internal
     */
    RestoreAddressToClassicResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RestoreAddressToClassicResult = exports.RestoreAddressToClassicResult || (exports.RestoreAddressToClassicResult = {}));
var RestoreManagedPrefixListVersionRequest;
(function (RestoreManagedPrefixListVersionRequest) {
    /**
     * @internal
     */
    RestoreManagedPrefixListVersionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RestoreManagedPrefixListVersionRequest = exports.RestoreManagedPrefixListVersionRequest || (exports.RestoreManagedPrefixListVersionRequest = {}));
var RestoreManagedPrefixListVersionResult;
(function (RestoreManagedPrefixListVersionResult) {
    /**
     * @internal
     */
    RestoreManagedPrefixListVersionResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RestoreManagedPrefixListVersionResult = exports.RestoreManagedPrefixListVersionResult || (exports.RestoreManagedPrefixListVersionResult = {}));
var RevokeClientVpnIngressRequest;
(function (RevokeClientVpnIngressRequest) {
    /**
     * @internal
     */
    RevokeClientVpnIngressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RevokeClientVpnIngressRequest = exports.RevokeClientVpnIngressRequest || (exports.RevokeClientVpnIngressRequest = {}));
var RevokeClientVpnIngressResult;
(function (RevokeClientVpnIngressResult) {
    /**
     * @internal
     */
    RevokeClientVpnIngressResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RevokeClientVpnIngressResult = exports.RevokeClientVpnIngressResult || (exports.RevokeClientVpnIngressResult = {}));
var RevokeSecurityGroupEgressRequest;
(function (RevokeSecurityGroupEgressRequest) {
    /**
     * @internal
     */
    RevokeSecurityGroupEgressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RevokeSecurityGroupEgressRequest = exports.RevokeSecurityGroupEgressRequest || (exports.RevokeSecurityGroupEgressRequest = {}));
var RevokeSecurityGroupEgressResult;
(function (RevokeSecurityGroupEgressResult) {
    /**
     * @internal
     */
    RevokeSecurityGroupEgressResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RevokeSecurityGroupEgressResult = exports.RevokeSecurityGroupEgressResult || (exports.RevokeSecurityGroupEgressResult = {}));
var RevokeSecurityGroupIngressRequest;
(function (RevokeSecurityGroupIngressRequest) {
    /**
     * @internal
     */
    RevokeSecurityGroupIngressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RevokeSecurityGroupIngressRequest = exports.RevokeSecurityGroupIngressRequest || (exports.RevokeSecurityGroupIngressRequest = {}));
var RevokeSecurityGroupIngressResult;
(function (RevokeSecurityGroupIngressResult) {
    /**
     * @internal
     */
    RevokeSecurityGroupIngressResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RevokeSecurityGroupIngressResult = exports.RevokeSecurityGroupIngressResult || (exports.RevokeSecurityGroupIngressResult = {}));
var CpuOptionsRequest;
(function (CpuOptionsRequest) {
    /**
     * @internal
     */
    CpuOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(CpuOptionsRequest = exports.CpuOptionsRequest || (exports.CpuOptionsRequest = {}));
var ElasticInferenceAccelerator;
(function (ElasticInferenceAccelerator) {
    /**
     * @internal
     */
    ElasticInferenceAccelerator.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ElasticInferenceAccelerator = exports.ElasticInferenceAccelerator || (exports.ElasticInferenceAccelerator = {}));
var EnclaveOptionsRequest;
(function (EnclaveOptionsRequest) {
    /**
     * @internal
     */
    EnclaveOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(EnclaveOptionsRequest = exports.EnclaveOptionsRequest || (exports.EnclaveOptionsRequest = {}));
var HibernationOptionsRequest;
(function (HibernationOptionsRequest) {
    /**
     * @internal
     */
    HibernationOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(HibernationOptionsRequest = exports.HibernationOptionsRequest || (exports.HibernationOptionsRequest = {}));
var SpotMarketOptions;
(function (SpotMarketOptions) {
    /**
     * @internal
     */
    SpotMarketOptions.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SpotMarketOptions = exports.SpotMarketOptions || (exports.SpotMarketOptions = {}));
var InstanceMarketOptionsRequest;
(function (InstanceMarketOptionsRequest) {
    /**
     * @internal
     */
    InstanceMarketOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceMarketOptionsRequest = exports.InstanceMarketOptionsRequest || (exports.InstanceMarketOptionsRequest = {}));
var LaunchTemplateSpecification;
(function (LaunchTemplateSpecification) {
    /**
     * @internal
     */
    LaunchTemplateSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(LaunchTemplateSpecification = exports.LaunchTemplateSpecification || (exports.LaunchTemplateSpecification = {}));
var LicenseConfigurationRequest;
(function (LicenseConfigurationRequest) {
    /**
     * @internal
     */
    LicenseConfigurationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(LicenseConfigurationRequest = exports.LicenseConfigurationRequest || (exports.LicenseConfigurationRequest = {}));
var InstanceMetadataOptionsRequest;
(function (InstanceMetadataOptionsRequest) {
    /**
     * @internal
     */
    InstanceMetadataOptionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceMetadataOptionsRequest = exports.InstanceMetadataOptionsRequest || (exports.InstanceMetadataOptionsRequest = {}));
var RunInstancesRequest;
(function (RunInstancesRequest) {
    /**
     * @internal
     */
    RunInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RunInstancesRequest = exports.RunInstancesRequest || (exports.RunInstancesRequest = {}));
var ScheduledInstancesEbs;
(function (ScheduledInstancesEbs) {
    /**
     * @internal
     */
    ScheduledInstancesEbs.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesEbs = exports.ScheduledInstancesEbs || (exports.ScheduledInstancesEbs = {}));
var ScheduledInstancesBlockDeviceMapping;
(function (ScheduledInstancesBlockDeviceMapping) {
    /**
     * @internal
     */
    ScheduledInstancesBlockDeviceMapping.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesBlockDeviceMapping = exports.ScheduledInstancesBlockDeviceMapping || (exports.ScheduledInstancesBlockDeviceMapping = {}));
var ScheduledInstancesIamInstanceProfile;
(function (ScheduledInstancesIamInstanceProfile) {
    /**
     * @internal
     */
    ScheduledInstancesIamInstanceProfile.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesIamInstanceProfile = exports.ScheduledInstancesIamInstanceProfile || (exports.ScheduledInstancesIamInstanceProfile = {}));
var ScheduledInstancesMonitoring;
(function (ScheduledInstancesMonitoring) {
    /**
     * @internal
     */
    ScheduledInstancesMonitoring.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesMonitoring = exports.ScheduledInstancesMonitoring || (exports.ScheduledInstancesMonitoring = {}));
var ScheduledInstancesIpv6Address;
(function (ScheduledInstancesIpv6Address) {
    /**
     * @internal
     */
    ScheduledInstancesIpv6Address.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesIpv6Address = exports.ScheduledInstancesIpv6Address || (exports.ScheduledInstancesIpv6Address = {}));
var ScheduledInstancesPrivateIpAddressConfig;
(function (ScheduledInstancesPrivateIpAddressConfig) {
    /**
     * @internal
     */
    ScheduledInstancesPrivateIpAddressConfig.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesPrivateIpAddressConfig = exports.ScheduledInstancesPrivateIpAddressConfig || (exports.ScheduledInstancesPrivateIpAddressConfig = {}));
var ScheduledInstancesNetworkInterface;
(function (ScheduledInstancesNetworkInterface) {
    /**
     * @internal
     */
    ScheduledInstancesNetworkInterface.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesNetworkInterface = exports.ScheduledInstancesNetworkInterface || (exports.ScheduledInstancesNetworkInterface = {}));
var ScheduledInstancesPlacement;
(function (ScheduledInstancesPlacement) {
    /**
     * @internal
     */
    ScheduledInstancesPlacement.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesPlacement = exports.ScheduledInstancesPlacement || (exports.ScheduledInstancesPlacement = {}));
var ScheduledInstancesLaunchSpecification;
(function (ScheduledInstancesLaunchSpecification) {
    /**
     * @internal
     */
    ScheduledInstancesLaunchSpecification.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ScheduledInstancesLaunchSpecification = exports.ScheduledInstancesLaunchSpecification || (exports.ScheduledInstancesLaunchSpecification = {}));
var RunScheduledInstancesRequest;
(function (RunScheduledInstancesRequest) {
    /**
     * @internal
     */
    RunScheduledInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RunScheduledInstancesRequest = exports.RunScheduledInstancesRequest || (exports.RunScheduledInstancesRequest = {}));
var RunScheduledInstancesResult;
(function (RunScheduledInstancesResult) {
    /**
     * @internal
     */
    RunScheduledInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(RunScheduledInstancesResult = exports.RunScheduledInstancesResult || (exports.RunScheduledInstancesResult = {}));
var SearchLocalGatewayRoutesRequest;
(function (SearchLocalGatewayRoutesRequest) {
    /**
     * @internal
     */
    SearchLocalGatewayRoutesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SearchLocalGatewayRoutesRequest = exports.SearchLocalGatewayRoutesRequest || (exports.SearchLocalGatewayRoutesRequest = {}));
var SearchLocalGatewayRoutesResult;
(function (SearchLocalGatewayRoutesResult) {
    /**
     * @internal
     */
    SearchLocalGatewayRoutesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SearchLocalGatewayRoutesResult = exports.SearchLocalGatewayRoutesResult || (exports.SearchLocalGatewayRoutesResult = {}));
var SearchTransitGatewayMulticastGroupsRequest;
(function (SearchTransitGatewayMulticastGroupsRequest) {
    /**
     * @internal
     */
    SearchTransitGatewayMulticastGroupsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SearchTransitGatewayMulticastGroupsRequest = exports.SearchTransitGatewayMulticastGroupsRequest || (exports.SearchTransitGatewayMulticastGroupsRequest = {}));
var TransitGatewayMulticastGroup;
(function (TransitGatewayMulticastGroup) {
    /**
     * @internal
     */
    TransitGatewayMulticastGroup.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TransitGatewayMulticastGroup = exports.TransitGatewayMulticastGroup || (exports.TransitGatewayMulticastGroup = {}));
var SearchTransitGatewayMulticastGroupsResult;
(function (SearchTransitGatewayMulticastGroupsResult) {
    /**
     * @internal
     */
    SearchTransitGatewayMulticastGroupsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SearchTransitGatewayMulticastGroupsResult = exports.SearchTransitGatewayMulticastGroupsResult || (exports.SearchTransitGatewayMulticastGroupsResult = {}));
var SearchTransitGatewayRoutesRequest;
(function (SearchTransitGatewayRoutesRequest) {
    /**
     * @internal
     */
    SearchTransitGatewayRoutesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SearchTransitGatewayRoutesRequest = exports.SearchTransitGatewayRoutesRequest || (exports.SearchTransitGatewayRoutesRequest = {}));
var SearchTransitGatewayRoutesResult;
(function (SearchTransitGatewayRoutesResult) {
    /**
     * @internal
     */
    SearchTransitGatewayRoutesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SearchTransitGatewayRoutesResult = exports.SearchTransitGatewayRoutesResult || (exports.SearchTransitGatewayRoutesResult = {}));
var SendDiagnosticInterruptRequest;
(function (SendDiagnosticInterruptRequest) {
    /**
     * @internal
     */
    SendDiagnosticInterruptRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SendDiagnosticInterruptRequest = exports.SendDiagnosticInterruptRequest || (exports.SendDiagnosticInterruptRequest = {}));
var StartInstancesRequest;
(function (StartInstancesRequest) {
    /**
     * @internal
     */
    StartInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StartInstancesRequest = exports.StartInstancesRequest || (exports.StartInstancesRequest = {}));
var InstanceStateChange;
(function (InstanceStateChange) {
    /**
     * @internal
     */
    InstanceStateChange.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(InstanceStateChange = exports.InstanceStateChange || (exports.InstanceStateChange = {}));
var StartInstancesResult;
(function (StartInstancesResult) {
    /**
     * @internal
     */
    StartInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StartInstancesResult = exports.StartInstancesResult || (exports.StartInstancesResult = {}));
var StartNetworkInsightsAnalysisRequest;
(function (StartNetworkInsightsAnalysisRequest) {
    /**
     * @internal
     */
    StartNetworkInsightsAnalysisRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StartNetworkInsightsAnalysisRequest = exports.StartNetworkInsightsAnalysisRequest || (exports.StartNetworkInsightsAnalysisRequest = {}));
var StartNetworkInsightsAnalysisResult;
(function (StartNetworkInsightsAnalysisResult) {
    /**
     * @internal
     */
    StartNetworkInsightsAnalysisResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StartNetworkInsightsAnalysisResult = exports.StartNetworkInsightsAnalysisResult || (exports.StartNetworkInsightsAnalysisResult = {}));
var StartVpcEndpointServicePrivateDnsVerificationRequest;
(function (StartVpcEndpointServicePrivateDnsVerificationRequest) {
    /**
     * @internal
     */
    StartVpcEndpointServicePrivateDnsVerificationRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StartVpcEndpointServicePrivateDnsVerificationRequest = exports.StartVpcEndpointServicePrivateDnsVerificationRequest || (exports.StartVpcEndpointServicePrivateDnsVerificationRequest = {}));
var StartVpcEndpointServicePrivateDnsVerificationResult;
(function (StartVpcEndpointServicePrivateDnsVerificationResult) {
    /**
     * @internal
     */
    StartVpcEndpointServicePrivateDnsVerificationResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StartVpcEndpointServicePrivateDnsVerificationResult = exports.StartVpcEndpointServicePrivateDnsVerificationResult || (exports.StartVpcEndpointServicePrivateDnsVerificationResult = {}));
var StopInstancesRequest;
(function (StopInstancesRequest) {
    /**
     * @internal
     */
    StopInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StopInstancesRequest = exports.StopInstancesRequest || (exports.StopInstancesRequest = {}));
var StopInstancesResult;
(function (StopInstancesResult) {
    /**
     * @internal
     */
    StopInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(StopInstancesResult = exports.StopInstancesResult || (exports.StopInstancesResult = {}));
var TerminateClientVpnConnectionsRequest;
(function (TerminateClientVpnConnectionsRequest) {
    /**
     * @internal
     */
    TerminateClientVpnConnectionsRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TerminateClientVpnConnectionsRequest = exports.TerminateClientVpnConnectionsRequest || (exports.TerminateClientVpnConnectionsRequest = {}));
var TerminateConnectionStatus;
(function (TerminateConnectionStatus) {
    /**
     * @internal
     */
    TerminateConnectionStatus.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TerminateConnectionStatus = exports.TerminateConnectionStatus || (exports.TerminateConnectionStatus = {}));
var TerminateClientVpnConnectionsResult;
(function (TerminateClientVpnConnectionsResult) {
    /**
     * @internal
     */
    TerminateClientVpnConnectionsResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TerminateClientVpnConnectionsResult = exports.TerminateClientVpnConnectionsResult || (exports.TerminateClientVpnConnectionsResult = {}));
var TerminateInstancesRequest;
(function (TerminateInstancesRequest) {
    /**
     * @internal
     */
    TerminateInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TerminateInstancesRequest = exports.TerminateInstancesRequest || (exports.TerminateInstancesRequest = {}));
var TerminateInstancesResult;
(function (TerminateInstancesResult) {
    /**
     * @internal
     */
    TerminateInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(TerminateInstancesResult = exports.TerminateInstancesResult || (exports.TerminateInstancesResult = {}));
var UnassignIpv6AddressesRequest;
(function (UnassignIpv6AddressesRequest) {
    /**
     * @internal
     */
    UnassignIpv6AddressesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnassignIpv6AddressesRequest = exports.UnassignIpv6AddressesRequest || (exports.UnassignIpv6AddressesRequest = {}));
var UnassignIpv6AddressesResult;
(function (UnassignIpv6AddressesResult) {
    /**
     * @internal
     */
    UnassignIpv6AddressesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnassignIpv6AddressesResult = exports.UnassignIpv6AddressesResult || (exports.UnassignIpv6AddressesResult = {}));
var UnassignPrivateIpAddressesRequest;
(function (UnassignPrivateIpAddressesRequest) {
    /**
     * @internal
     */
    UnassignPrivateIpAddressesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnassignPrivateIpAddressesRequest = exports.UnassignPrivateIpAddressesRequest || (exports.UnassignPrivateIpAddressesRequest = {}));
var UnmonitorInstancesRequest;
(function (UnmonitorInstancesRequest) {
    /**
     * @internal
     */
    UnmonitorInstancesRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnmonitorInstancesRequest = exports.UnmonitorInstancesRequest || (exports.UnmonitorInstancesRequest = {}));
var UnmonitorInstancesResult;
(function (UnmonitorInstancesResult) {
    /**
     * @internal
     */
    UnmonitorInstancesResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UnmonitorInstancesResult = exports.UnmonitorInstancesResult || (exports.UnmonitorInstancesResult = {}));
var SecurityGroupRuleDescription;
(function (SecurityGroupRuleDescription) {
    /**
     * @internal
     */
    SecurityGroupRuleDescription.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(SecurityGroupRuleDescription = exports.SecurityGroupRuleDescription || (exports.SecurityGroupRuleDescription = {}));
var UpdateSecurityGroupRuleDescriptionsEgressRequest;
(function (UpdateSecurityGroupRuleDescriptionsEgressRequest) {
    /**
     * @internal
     */
    UpdateSecurityGroupRuleDescriptionsEgressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UpdateSecurityGroupRuleDescriptionsEgressRequest = exports.UpdateSecurityGroupRuleDescriptionsEgressRequest || (exports.UpdateSecurityGroupRuleDescriptionsEgressRequest = {}));
var UpdateSecurityGroupRuleDescriptionsEgressResult;
(function (UpdateSecurityGroupRuleDescriptionsEgressResult) {
    /**
     * @internal
     */
    UpdateSecurityGroupRuleDescriptionsEgressResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UpdateSecurityGroupRuleDescriptionsEgressResult = exports.UpdateSecurityGroupRuleDescriptionsEgressResult || (exports.UpdateSecurityGroupRuleDescriptionsEgressResult = {}));
var UpdateSecurityGroupRuleDescriptionsIngressRequest;
(function (UpdateSecurityGroupRuleDescriptionsIngressRequest) {
    /**
     * @internal
     */
    UpdateSecurityGroupRuleDescriptionsIngressRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UpdateSecurityGroupRuleDescriptionsIngressRequest = exports.UpdateSecurityGroupRuleDescriptionsIngressRequest || (exports.UpdateSecurityGroupRuleDescriptionsIngressRequest = {}));
var UpdateSecurityGroupRuleDescriptionsIngressResult;
(function (UpdateSecurityGroupRuleDescriptionsIngressResult) {
    /**
     * @internal
     */
    UpdateSecurityGroupRuleDescriptionsIngressResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(UpdateSecurityGroupRuleDescriptionsIngressResult = exports.UpdateSecurityGroupRuleDescriptionsIngressResult || (exports.UpdateSecurityGroupRuleDescriptionsIngressResult = {}));
var WithdrawByoipCidrRequest;
(function (WithdrawByoipCidrRequest) {
    /**
     * @internal
     */
    WithdrawByoipCidrRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(WithdrawByoipCidrRequest = exports.WithdrawByoipCidrRequest || (exports.WithdrawByoipCidrRequest = {}));
var WithdrawByoipCidrResult;
(function (WithdrawByoipCidrResult) {
    /**
     * @internal
     */
    WithdrawByoipCidrResult.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(WithdrawByoipCidrResult = exports.WithdrawByoipCidrResult || (exports.WithdrawByoipCidrResult = {}));
