// renovate/dist/types/skip-reason.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkipReason = void 0;
var SkipReason;
(function (SkipReason) {
    SkipReason["AnyVersion"] = "any-version";
    SkipReason["ContainsVariable"] = "contains-variable";
    SkipReason["Disabled"] = "disabled";
    SkipReason["Empty"] = "empty";
    SkipReason["FileDependency"] = "file-dependency";
    SkipReason["File"] = "file";
    SkipReason["GitDependency"] = "git-dependency";
    SkipReason["GitPlugin"] = "git-plugin";
    SkipReason["Ignored"] = "ignored";
    SkipReason["InternalError"] = "internal-error";
    SkipReason["InternalPackage"] = "internal-package";
    SkipReason["InvalidConfig"] = "invalid-config";
    SkipReason["InvalidDependencySpecification"] = "invalid-dependency-specification";
    SkipReason["InvalidName"] = "invalid-name";
    SkipReason["InvalidSha256"] = "invalid-sha256";
    SkipReason["InvalidUrl"] = "invalid-url";
    SkipReason["InvalidValue"] = "invalid-value";
    SkipReason["InvalidVersion"] = "invalid-version";
    SkipReason["LocalChart"] = "local-chart";
    SkipReason["LocalDependency"] = "local-dependency";
    SkipReason["Local"] = "local";
    SkipReason["MultipleConstraintDep"] = "multiple-constraint-dep";
    SkipReason["NamePlaceholder"] = "name-placeholder";
    SkipReason["NoRepository"] = "no-repository";
    SkipReason["NoSourceMatch"] = "no-source-match";
    SkipReason["NoSource"] = "no-source";
    SkipReason["NoVersion"] = "no-version";
    SkipReason["NonHexDeptypes"] = "non-hex depTypes";
    SkipReason["NotAVersion"] = "not-a-version";
    SkipReason["PathDependency"] = "path-dependency";
    SkipReason["PlaceholderUrl"] = "placeholder-url";
    SkipReason["UnknownEngines"] = "unknown-engines";
    SkipReason["UnknownRegistry"] = "unknown-registry";
    SkipReason["UnknownVersion"] = "unknown-version";
    SkipReason["UnknownVolta"] = "unknown-volta";
    SkipReason["UnsupportedChartType"] = "unsupported-chart-type";
    SkipReason["UnsupportedDatasource"] = "unsupported-datasource";
    SkipReason["UnsupportedRemote"] = "unsupported-remote";
    SkipReason["UnsupportedUrl"] = "unsupported-url";
    SkipReason["UnsupportedVersion"] = "unsupported-version";
    SkipReason["Unsupported"] = "unsupported";
    SkipReason["UnversionedReference"] = "unversioned-reference";
    SkipReason["VersionPlaceholder"] = "version-placeholder";
    SkipReason["IsPinned"] = "is-pinned";
})(SkipReason = exports.SkipReason || (exports.SkipReason = {}));
