// renovate/dist/datasource/bitbucket-tags/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitBucketTagsDatasource = void 0;
const tslib_1 = require("tslib");
const decorator_1 = require("../../util/cache/package/decorator");
const bitbucket_1 = require("../../util/http/bitbucket");
const url_1 = require("../../util/url");
const datasource_1 = require("../datasource");
class BitBucketTagsDatasource extends datasource_1.Datasource {
    constructor() {
        super(BitBucketTagsDatasource.id);
        this.bitbucketHttp = new bitbucket_1.BitbucketHttp();
    }
    static getRegistryURL(registryUrl) {
        // fallback to default API endpoint if custom not provided
        return registryUrl !== null && registryUrl !== void 0 ? registryUrl : this.defaultRegistryUrls[0];
    }
    static getCacheKey(registryUrl, repo, type) {
        return `${BitBucketTagsDatasource.getRegistryURL(registryUrl)}:${repo}:${type}`;
    }
    static getSourceUrl(lookupName, registryUrl) {
        const url = BitBucketTagsDatasource.getRegistryURL(registryUrl);
        const normalizedUrl = (0, url_1.ensureTrailingSlash)(url);
        return `${normalizedUrl}${lookupName}`;
    }
    // getReleases fetches list of tags for the repository
    async getReleases({ registryUrl, lookupName: repo, }) {
        const url = `/2.0/repositories/${repo}/refs/tags`;
        const bitbucketTags = (await this.bitbucketHttp.getJson(url)).body;
        const dependency = {
            sourceUrl: BitBucketTagsDatasource.getSourceUrl(repo, registryUrl),
            registryUrl: BitBucketTagsDatasource.getRegistryURL(registryUrl),
            releases: null,
        };
        dependency.releases = bitbucketTags.values.map(({ name, target }) => ({
            version: name,
            gitRef: name,
            releaseTimestamp: target === null || target === void 0 ? void 0 : target.date,
        }));
        return dependency;
    }
    // getTagCommit fetched the commit has for specified tag
    async getTagCommit(registryUrl, repo, tag) {
        const url = `/2.0/repositories/${repo}/refs/tags/${tag}`;
        const bitbucketTag = (await this.bitbucketHttp.getJson(url))
            .body;
        return bitbucketTag.target.hash;
    }
    async getMainBranch(repo) {
        return (await this.bitbucketHttp.getJson(`/2.0/repositories/${repo}`)).body.mainbranch.name;
    }
    // getDigest fetched the latest commit for repository main branch
    // however, if newValue is provided, then getTagCommit is called
    async getDigest({ lookupName: repo, registryUrl }, newValue) {
        if (newValue === null || newValue === void 0 ? void 0 : newValue.length) {
            return this.getTagCommit(registryUrl, repo, newValue);
        }
        const mainBranch = await this.getMainBranch(repo);
        const url = `/2.0/repositories/${repo}/commits/${mainBranch}`;
        const bitbucketCommits = (await this.bitbucketHttp.getJson(url)).body;
        if (bitbucketCommits.values.length === 0) {
            return null;
        }
        return bitbucketCommits.values[0].hash;
    }
}
BitBucketTagsDatasource.id = 'bitbucket-tags';
BitBucketTagsDatasource.customRegistrySupport = true;
BitBucketTagsDatasource.registryStrategy = 'first';
BitBucketTagsDatasource.defaultRegistryUrls = ['https://bitbucket.org'];
BitBucketTagsDatasource.cacheNamespace = `datasource-${BitBucketTagsDatasource.id}`;
(0, tslib_1.__decorate)([
    (0, decorator_1.cache)({
        namespace: BitBucketTagsDatasource.cacheNamespace,
        key: ({ registryUrl, lookupName }) => BitBucketTagsDatasource.getCacheKey(registryUrl, lookupName, 'tags'),
    })
], BitBucketTagsDatasource.prototype, "getReleases", null);
(0, tslib_1.__decorate)([
    (0, decorator_1.cache)({
        namespace: BitBucketTagsDatasource.cacheNamespace,
        key: (registryUrl, repo, tag) => BitBucketTagsDatasource.getCacheKey(registryUrl, repo, `tag-${tag}`),
    })
], BitBucketTagsDatasource.prototype, "getTagCommit", null);
(0, tslib_1.__decorate)([
    (0, decorator_1.cache)({
        namespace: BitBucketTagsDatasource.cacheNamespace,
        key: (registryUrl, repo) => BitBucketTagsDatasource.getCacheKey(registryUrl, repo, 'mainbranch'),
        ttlMinutes: 60,
    })
], BitBucketTagsDatasource.prototype, "getMainBranch", null);
(0, tslib_1.__decorate)([
    (0, decorator_1.cache)({
        namespace: BitBucketTagsDatasource.cacheNamespace,
        key: ({ registryUrl, lookupName }) => BitBucketTagsDatasource.getCacheKey(registryUrl, lookupName, 'digest'),
    })
], BitBucketTagsDatasource.prototype, "getDigest", null);
exports.BitBucketTagsDatasource = BitBucketTagsDatasource;
