// renovate/dist/datasource/dart/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DartDatasource = void 0;
const datasource_1 = require("../datasource");
class DartDatasource extends datasource_1.Datasource {
    constructor() {
        super(DartDatasource.id);
        this.customRegistrySupport = false;
        this.defaultRegistryUrls = ['https://pub.dartlang.org/'];
    }
    async getReleases({ lookupName, registryUrl, }) {
        let result = null;
        const pkgUrl = `${registryUrl}api/packages/${lookupName}`;
        let raw = null;
        try {
            raw = await this.http.getJson(pkgUrl);
        }
        catch (err) {
            this.handleGenericErrors(err);
        }
        const body = raw === null || raw === void 0 ? void 0 : raw.body;
        if (body) {
            const { versions, latest } = body;
            if (versions && latest) {
                result = {
                    releases: body.versions.map(({ version, published }) => ({
                        version,
                        releaseTimestamp: published,
                    })),
                };
                const pubspec = latest.pubspec;
                if (pubspec) {
                    if (pubspec.homepage) {
                        result.homepage = pubspec.homepage;
                    }
                    if (pubspec.repository) {
                        result.sourceUrl = pubspec.repository;
                    }
                }
            }
        }
        return result;
    }
}
exports.DartDatasource = DartDatasource;
DartDatasource.id = 'dart';
