// renovate/dist/manager/gradle-wrapper/artifacts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateArtifacts = void 0;
const shlex_1 = require("shlex");
const upath_1 = require("upath");
const global_1 = require("../../config/global");
const error_messages_1 = require("../../constants/error-messages");
const logger_1 = require("../../logger");
const exec_1 = require("../../util/exec");
const fs_1 = require("../../util/fs");
const git_1 = require("../../util/git");
const http_1 = require("../../util/http");
const utils_1 = require("./utils");
const http = new http_1.Http('gradle-wrapper');
async function addIfUpdated(status, fileProjectPath) {
    if (status.modified.includes(fileProjectPath)) {
        return {
            file: {
                name: fileProjectPath,
                contents: await (0, fs_1.readLocalFile)(fileProjectPath),
            },
        };
    }
    return null;
}
function getDistributionUrl(newPackageFileContent) {
    const distributionUrlLine = newPackageFileContent
        .split('\n')
        .find((line) => line.startsWith('distributionUrl='));
    if (distributionUrlLine) {
        return distributionUrlLine
            .replace('distributionUrl=', '')
            .replace('https\\:', 'https:');
    }
    return null;
}
async function getDistributionChecksum(url) {
    const { body } = await http.get(`${url}.sha256`);
    return body;
}
async function updateArtifacts({ packageFileName, newPackageFileContent, updatedDeps, config, }) {
    var _a, _b;
    try {
        const projectDir = global_1.GlobalConfig.get('localDir');
        logger_1.logger.debug({ updatedDeps }, 'gradle-wrapper.updateArtifacts()');
        const gradlew = (0, utils_1.gradleWrapperFileName)();
        const gradlewPath = (0, upath_1.resolve)(projectDir, `./${gradlew}`);
        let cmd = await (0, utils_1.prepareGradleCommand)(gradlew, projectDir, await (0, fs_1.stat)(gradlewPath).catch(() => null), `wrapper`);
        if (!cmd) {
            logger_1.logger.info('No gradlew found - skipping Artifacts update');
            return null;
        }
        const distributionUrl = getDistributionUrl(newPackageFileContent);
        if (distributionUrl) {
            cmd += ` --gradle-distribution-url ${distributionUrl}`;
            if (newPackageFileContent.includes('distributionSha256Sum=')) {
                // need to reset version, otherwise we have a checksum mismatch
                await (0, fs_1.writeLocalFile)(packageFileName, newPackageFileContent.replace(config.newValue, config.currentValue));
                const checksum = await getDistributionChecksum(distributionUrl);
                cmd += ` --gradle-distribution-sha256-sum ${(0, shlex_1.quote)(checksum)}`;
            }
        }
        else {
            cmd += ` --gradle-version ${(0, shlex_1.quote)(config.newValue)}`;
        }
        logger_1.logger.debug(`Updating gradle wrapper: "${cmd}"`);
        const execOptions = {
            docker: {
                image: 'java',
                tagConstraint: (_b = (_a = config.constraints) === null || _a === void 0 ? void 0 : _a.java) !== null && _b !== void 0 ? _b : (0, utils_1.getJavaContraint)(config.currentValue),
                tagScheme: (0, utils_1.getJavaVersioning)(),
            },
            extraEnv: utils_1.extraEnv,
        };
        try {
            await (0, exec_1.exec)(cmd, execOptions);
        }
        catch (err) {
            // istanbul ignore if
            if (err.message === error_messages_1.TEMPORARY_ERROR) {
                throw err;
            }
            logger_1.logger.warn({ err }, 'Error executing gradle wrapper update command. It can be not a critical one though.');
        }
        const status = await (0, git_1.getRepoStatus)();
        const artifactFileNames = [
            'gradle/wrapper/gradle-wrapper.properties',
            'gradle/wrapper/gradle-wrapper.jar',
            'gradlew',
            'gradlew.bat',
        ].map((filename) => packageFileName
            .replace('gradle/wrapper/', '')
            .replace('gradle-wrapper.properties', '') + filename);
        const updateArtifactsResult = (await Promise.all(artifactFileNames.map((fileProjectPath) => addIfUpdated(status, fileProjectPath)))).filter(Boolean);
        logger_1.logger.debug({ files: updateArtifactsResult.map((r) => r.file.name) }, `Returning updated gradle-wrapper files`);
        return updateArtifactsResult;
    }
    catch (err) {
        logger_1.logger.debug({ err }, 'Error setting new Gradle Wrapper release value');
        return [
            {
                artifactError: {
                    lockFile: packageFileName,
                    stderr: err.message,
                },
            },
        ];
    }
}
exports.updateArtifacts = updateArtifacts;
