// renovate/dist/datasource/github-releases/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGithubRelease = exports.getSourceUrl = exports.getApiBaseUrl = exports.getSourceUrlBase = exports.http = exports.cacheNamespace = exports.id = void 0;
const github_1 = require("../../util/http/github");
const url_1 = require("../../util/url");
const defaultSourceUrlBase = 'https://github.com/';
exports.id = 'github-releases';
exports.cacheNamespace = 'datasource-github-releases';
exports.http = new github_1.GithubHttp(exports.id);
function getSourceUrlBase(registryUrl) {
    // default to GitHub.com if no GHE host is specified.
    return (0, url_1.ensureTrailingSlash)(registryUrl !== null && registryUrl !== void 0 ? registryUrl : defaultSourceUrlBase);
}
exports.getSourceUrlBase = getSourceUrlBase;
function getApiBaseUrl(registryUrl) {
    const sourceUrlBase = getSourceUrlBase(registryUrl);
    return sourceUrlBase === defaultSourceUrlBase
        ? `https://api.github.com/`
        : `${sourceUrlBase}api/v3/`;
}
exports.getApiBaseUrl = getApiBaseUrl;
function getSourceUrl(lookupName, registryUrl) {
    const sourceUrlBase = getSourceUrlBase(registryUrl);
    return `${sourceUrlBase}${lookupName}`;
}
exports.getSourceUrl = getSourceUrl;
async function getGithubRelease(apiBaseUrl, repo, version) {
    const url = `${apiBaseUrl}repos/${repo}/releases/tags/${version}`;
    const res = await exports.http.getJson(url);
    return res.body;
}
exports.getGithubRelease = getGithubRelease;
