// renovate/dist/manager/travis/extract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractPackageFile = void 0;
const tslib_1 = require("tslib");
const is_1 = (0, tslib_1.__importDefault)(require("@sindresorhus/is"));
const js_yaml_1 = require("js-yaml");
const datasourceGithubTags = (0, tslib_1.__importStar)(require("../../datasource/github-tags"));
const logger_1 = require("../../logger");
function extractPackageFile(content) {
    var _a, _b;
    let doc;
    try {
        doc = (0, js_yaml_1.load)(content, {
            json: true,
        });
    }
    catch (err) {
        logger_1.logger.warn({ err, content }, 'Failed to parse .travis.yml file.');
        return null;
    }
    let deps = [];
    if (doc && is_1.default.array(doc.node_js)) {
        deps = doc.node_js.map((currentValue) => ({
            depName: 'node',
            datasource: datasourceGithubTags.id,
            lookupName: 'nodejs/node',
            currentValue: currentValue.toString(),
        }));
    }
    // Handle the matrix syntax
    let matrix_include;
    if ((_a = doc === null || doc === void 0 ? void 0 : doc.jobs) === null || _a === void 0 ? void 0 : _a.include) {
        matrix_include = doc.jobs.include;
    }
    else if ((_b = doc === null || doc === void 0 ? void 0 : doc.matrix) === null || _b === void 0 ? void 0 : _b.include) {
        matrix_include = doc.matrix.include;
    }
    if (!is_1.default.array(matrix_include)) {
        return deps.length ? { deps } : null;
    }
    for (const item of matrix_include) {
        if (item === null || item === void 0 ? void 0 : item.node_js) {
            if (is_1.default.array(item.node_js)) {
                item.node_js.forEach((currentValue) => {
                    deps.push({
                        depName: 'node',
                        datasource: datasourceGithubTags.id,
                        lookupName: 'nodejs/node',
                        currentValue: currentValue.toString(),
                    });
                });
            }
            else if (is_1.default.string(item.node_js)) {
                deps.push({
                    depName: 'node',
                    datasource: datasourceGithubTags.id,
                    lookupName: 'nodejs/node',
                    currentValue: item.node_js.toString(),
                });
            }
        }
    }
    if (!deps.length) {
        return null;
    }
    return { deps };
}
exports.extractPackageFile = extractPackageFile;
