// renovate/dist/config/presets/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchPreset = exports.PRESET_RENOVATE_CONFIG_NOT_FOUND = exports.PRESET_PROHIBITED_SUBPRESET = exports.PRESET_NOT_FOUND = exports.PRESET_INVALID_JSON = exports.PRESET_INVALID = exports.PRESET_DEP_NOT_FOUND = void 0;
const logger_1 = require("../../logger");
const url_1 = require("../../util/url");
exports.PRESET_DEP_NOT_FOUND = 'dep not found';
exports.PRESET_INVALID = 'invalid preset';
exports.PRESET_INVALID_JSON = 'invalid preset JSON';
exports.PRESET_NOT_FOUND = 'preset not found';
exports.PRESET_PROHIBITED_SUBPRESET = 'prohibited sub-preset';
exports.PRESET_RENOVATE_CONFIG_NOT_FOUND = 'preset renovate-config not found';
async function fetchPreset({ pkgName, filePreset, presetPath, endpoint: _endpoint, packageTag = null, fetch, }) {
    const endpoint = (0, url_1.ensureTrailingSlash)(_endpoint);
    const [fileName, presetName, subPresetName] = filePreset.split('/');
    const pathPrefix = presetPath ? `${presetPath}/` : '';
    const buildFilePath = (name) => `${pathPrefix}${name}`;
    let jsonContent;
    if (fileName === 'default') {
        try {
            jsonContent = await fetch(pkgName, buildFilePath('default.json'), endpoint, packageTag);
        }
        catch (err) {
            if (err.message !== exports.PRESET_DEP_NOT_FOUND) {
                throw err;
            }
            logger_1.logger.info('Fallback to renovate.json file as a preset is deprecated, please use a default.json file instead.');
            jsonContent = await fetch(pkgName, buildFilePath('renovate.json'), endpoint, packageTag);
        }
    }
    else {
        jsonContent = await fetch(pkgName, buildFilePath(`${fileName}.json`), endpoint, packageTag);
    }
    if (!jsonContent) {
        throw new Error(exports.PRESET_DEP_NOT_FOUND);
    }
    if (presetName) {
        const preset = jsonContent[presetName];
        if (!preset) {
            throw new Error(exports.PRESET_NOT_FOUND);
        }
        if (subPresetName) {
            const subPreset = preset[subPresetName];
            if (!subPreset) {
                throw new Error(exports.PRESET_NOT_FOUND);
            }
            return subPreset;
        }
        return preset;
    }
    return jsonContent;
}
exports.fetchPreset = fetchPreset;
