// renovate/dist/workers/pr/body/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPrBody = void 0;
const tslib_1 = require("tslib");
const platform_1 = require("../../../platform");
const regex_1 = require("../../../util/regex");
const template = (0, tslib_1.__importStar)(require("../../../util/template"));
const changelogs_1 = require("./changelogs");
const config_description_1 = require("./config-description");
const controls_1 = require("./controls");
const footer_1 = require("./footer");
const header_1 = require("./header");
const notes_1 = require("./notes");
const updates_table_1 = require("./updates-table");
function massageUpdateMetadata(config) {
    config.upgrades.forEach((upgrade) => {
        const { homepage, sourceUrl, sourceDirectory, changelogUrl, dependencyUrl, } = upgrade;
        let depNameLinked = upgrade.depName;
        const primaryLink = homepage || sourceUrl || dependencyUrl;
        if (primaryLink) {
            depNameLinked = `[${depNameLinked}](${primaryLink})`;
        }
        const otherLinks = [];
        if (homepage && sourceUrl) {
            otherLinks.push(`[source](${sourceUrl})`);
        }
        if (changelogUrl) {
            otherLinks.push(`[changelog](${changelogUrl})`);
        }
        if (otherLinks.length) {
            depNameLinked += ` (${otherLinks.join(', ')})`;
        }
        upgrade.depNameLinked = depNameLinked;
        const references = [];
        if (homepage) {
            references.push(`[homepage](${homepage})`);
        }
        if (sourceUrl) {
            let fullUrl = sourceUrl;
            if (sourceDirectory) {
                fullUrl =
                    sourceUrl.replace((0, regex_1.regEx)(/\/?$/), '/') +
                        'tree/HEAD/' +
                        sourceDirectory.replace('^/?/', '');
            }
            references.push(`[source](${fullUrl})`);
        }
        if (changelogUrl) {
            references.push(`[changelog](${changelogUrl})`);
        }
        upgrade.references = references.join(', ');
    });
}
async function getPrBody(config) {
    massageUpdateMetadata(config);
    const content = {
        header: (0, header_1.getPrHeader)(config),
        table: (0, updates_table_1.getPrUpdatesTable)(config),
        notes: (0, notes_1.getPrNotes)(config) + (0, notes_1.getPrExtraNotes)(config),
        changelogs: (0, changelogs_1.getChangelogs)(config),
        configDescription: await (0, config_description_1.getPrConfigDescription)(config),
        controls: await (0, controls_1.getControls)(config),
        footer: (0, footer_1.getPrFooter)(config),
    };
    const prBodyTemplate = config.prBodyTemplate;
    let prBody = template.compile(prBodyTemplate, content, false);
    prBody = prBody.trim();
    prBody = prBody.replace((0, regex_1.regEx)(/\n\n\n+/g), '\n\n');
    prBody = platform_1.platform.massageMarkdown(prBody);
    return prBody;
}
exports.getPrBody = getPrBody;
