// renovate/dist/versioning/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.api = exports.isStable = exports.isValid = exports.supportsRanges = exports.urls = exports.displayName = exports.id = void 0;
const tslib_1 = require("tslib");
const luxon_1 = require("luxon");
const npm_1 = (0, tslib_1.__importStar)(require("../npm"));
Object.defineProperty(exports, "isValid", { enumerable: true, get: function () { return npm_1.isValid; } });
const schedule_1 = require("./schedule");
exports.id = 'node';
exports.displayName = 'Node.js';
exports.urls = [];
exports.supportsRanges = false;
function getNewValue({ currentValue, rangeStrategy, currentVersion, newVersion, }) {
    const res = npm_1.default.getNewValue({
        currentValue,
        rangeStrategy,
        currentVersion,
        newVersion,
    });
    if ((0, npm_1.isVersion)(res)) {
        // normalize out any 'v' prefix
        return (0, npm_1.isVersion)(res);
    }
    return res;
}
function isStable(version) {
    if (npm_1.default.isStable(version)) {
        const major = npm_1.default.getMajor(version);
        const schedule = schedule_1.nodeSchedule[`v${major}`];
        if (schedule === null || schedule === void 0 ? void 0 : schedule.lts) {
            // TODO: use the exact release that started LTS (#9716)
            return luxon_1.DateTime.local() > luxon_1.DateTime.fromISO(schedule.lts);
        }
    }
    return false;
}
exports.isStable = isStable;
exports.api = {
    ...npm_1.default,
    isStable,
    getNewValue,
};
exports.default = exports.api;
