// renovate/dist/manager/gradle/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = exports.language = exports.updateDependency = exports.extractAllPackageFiles = void 0;
const tslib_1 = require("tslib");
const constants_1 = require("../../constants");
const gradleVersioning = (0, tslib_1.__importStar)(require("../../versioning/gradle"));
const deep = (0, tslib_1.__importStar)(require("./deep"));
const shallow = (0, tslib_1.__importStar)(require("./shallow"));
function extractAllPackageFiles(config, packageFiles) {
    return config.deepExtract
        ? deep.extractAllPackageFiles(config, packageFiles)
        : shallow.extractAllPackageFiles(config, packageFiles);
}
exports.extractAllPackageFiles = extractAllPackageFiles;
function updateDependency(params) {
    var _a;
    return ((_a = params.upgrade) === null || _a === void 0 ? void 0 : _a.deepExtract)
        ? deep.updateDependency(params)
        : shallow.updateDependency(params);
}
exports.updateDependency = updateDependency;
exports.language = constants_1.ProgrammingLanguage.Java;
exports.defaultConfig = {
    fileMatch: [
        '\\.gradle(\\.kts)?$',
        '(^|\\/)gradle\\.properties$',
        '(^|\\/)gradle\\/.+\\.toml$',
        '\\.versions\\.toml$',
    ],
    timeout: 600,
    versioning: gradleVersioning.id,
};
