// renovate/dist/util/fs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWriteStream = exports.readLocalDirectory = exports.findLocalSiblingOrParent = exports.localPathExists = exports.privateCacheDir = exports.ensureCacheDir = exports.ensureLocalDir = exports.ensureDir = exports.renameLocalFile = exports.deleteLocalFile = exports.writeLocalFile = exports.readLocalFile = exports.getSiblingFileName = exports.getSubDirectory = exports.pipeline = void 0;
const tslib_1 = require("tslib");
const stream_1 = (0, tslib_1.__importDefault)(require("stream"));
const util_1 = (0, tslib_1.__importDefault)(require("util"));
const is_1 = (0, tslib_1.__importDefault)(require("@sindresorhus/is"));
const fs = (0, tslib_1.__importStar)(require("fs-extra"));
const upath_1 = require("upath");
const global_1 = require("../../config/global");
const logger_1 = require("../../logger");
(0, tslib_1.__exportStar)(require("./proxies"), exports);
exports.pipeline = util_1.default.promisify(stream_1.default.pipeline);
function getSubDirectory(fileName) {
    return (0, upath_1.parse)(fileName).dir;
}
exports.getSubDirectory = getSubDirectory;
function getSiblingFileName(existingFileNameWithPath, otherFileName) {
    const subDirectory = getSubDirectory(existingFileNameWithPath);
    return (0, upath_1.join)(subDirectory, otherFileName);
}
exports.getSiblingFileName = getSiblingFileName;
async function readLocalFile(fileName, encoding) {
    const { localDir } = global_1.GlobalConfig.get();
    const localFileName = (0, upath_1.join)(localDir, fileName);
    try {
        const fileContent = encoding
            ? await fs.readFile(localFileName, encoding)
            : await fs.readFile(localFileName);
        return fileContent;
    }
    catch (err) {
        logger_1.logger.trace({ err }, 'Error reading local file');
        return null;
    }
}
exports.readLocalFile = readLocalFile;
async function writeLocalFile(fileName, fileContent) {
    const { localDir } = global_1.GlobalConfig.get();
    const localFileName = (0, upath_1.join)(localDir, fileName);
    await fs.outputFile(localFileName, fileContent);
}
exports.writeLocalFile = writeLocalFile;
async function deleteLocalFile(fileName) {
    const { localDir } = global_1.GlobalConfig.get();
    if (localDir) {
        const localFileName = (0, upath_1.join)(localDir, fileName);
        await fs.remove(localFileName);
    }
}
exports.deleteLocalFile = deleteLocalFile;
// istanbul ignore next
async function renameLocalFile(fromFile, toFile) {
    const { localDir } = global_1.GlobalConfig.get();
    await fs.move((0, upath_1.join)(localDir, fromFile), (0, upath_1.join)(localDir, toFile));
}
exports.renameLocalFile = renameLocalFile;
// istanbul ignore next
async function ensureDir(dirName) {
    if (is_1.default.nonEmptyString(dirName)) {
        await fs.ensureDir(dirName);
    }
}
exports.ensureDir = ensureDir;
// istanbul ignore next
async function ensureLocalDir(dirName) {
    const { localDir } = global_1.GlobalConfig.get();
    const localDirName = (0, upath_1.join)(localDir, dirName);
    await fs.ensureDir(localDirName);
}
exports.ensureLocalDir = ensureLocalDir;
async function ensureCacheDir(name) {
    const cacheDirName = (0, upath_1.join)(global_1.GlobalConfig.get('cacheDir'), `others/${name}`);
    await fs.ensureDir(cacheDirName);
    return cacheDirName;
}
exports.ensureCacheDir = ensureCacheDir;
/**
 * Return the path of the private cache directory. This directory is wiped
 * between repositories, so they can be used to store private registries' index
 * without risk of that information leaking to other repositories/users.
 */
function privateCacheDir() {
    const { cacheDir } = global_1.GlobalConfig.get();
    return (0, upath_1.join)(cacheDir, '__renovate-private-cache');
}
exports.privateCacheDir = privateCacheDir;
function localPathExists(pathName) {
    const { localDir } = global_1.GlobalConfig.get();
    // Works for both files as well as directories
    return fs
        .stat((0, upath_1.join)(localDir, pathName))
        .then((s) => !!s)
        .catch(() => false);
}
exports.localPathExists = localPathExists;
/**
 * Tries to find `otherFileName` in the directory where
 * `existingFileNameWithPath` is, then in its parent directory, then in the
 * grandparent, until we reach the top-level directory. All paths
 * must be relative to `localDir`.
 */
async function findLocalSiblingOrParent(existingFileNameWithPath, otherFileName) {
    if ((0, upath_1.isAbsolute)(existingFileNameWithPath)) {
        return null;
    }
    if ((0, upath_1.isAbsolute)(otherFileName)) {
        return null;
    }
    let current = existingFileNameWithPath;
    while (current !== '') {
        current = getSubDirectory(current);
        const candidate = (0, upath_1.join)(current, otherFileName);
        if (await localPathExists(candidate)) {
            return candidate;
        }
    }
    return null;
}
exports.findLocalSiblingOrParent = findLocalSiblingOrParent;
/**
 * Get files by name from directory
 */
async function readLocalDirectory(path) {
    const { localDir } = global_1.GlobalConfig.get();
    const localPath = (0, upath_1.join)(localDir, path);
    const fileList = await fs.readdir(localPath);
    return fileList;
}
exports.readLocalDirectory = readLocalDirectory;
function createWriteStream(path) {
    return fs.createWriteStream(path);
}
exports.createWriteStream = createWriteStream;
