// renovate/dist/util/fs/proxies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rm = exports.readdir = exports.move = exports.pathExists = exports.exists = exports.unlink = exports.remove = exports.outputFile = exports.writeFile = exports.readFile = exports.chmod = exports.stat = void 0;
const tslib_1 = require("tslib");
const fs = (0, tslib_1.__importStar)(require("fs-extra"));
// istanbul ignore next
function stat(path) {
    return fs.stat(path);
}
exports.stat = stat;
// istanbul ignore next
function chmod(path, mode) {
    return fs.chmod(path, mode);
}
exports.chmod = chmod;
function readFile(fileName, encoding) {
    return encoding ? fs.readFile(fileName, encoding) : fs.readFile(fileName);
}
exports.readFile = readFile;
// istanbul ignore next
function writeFile(fileName, fileContent) {
    return fs.writeFile(fileName, fileContent);
}
exports.writeFile = writeFile;
// istanbul ignore next
function outputFile(file, data, options) {
    return fs.outputFile(file, data, options !== null && options !== void 0 ? options : {});
}
exports.outputFile = outputFile;
function remove(dir) {
    return fs.remove(dir);
}
exports.remove = remove;
// istanbul ignore next
function unlink(path) {
    return fs.unlink(path);
}
exports.unlink = unlink;
// istanbul ignore next
function exists(path) {
    return fs.pathExists(path);
}
exports.exists = exists;
// istanbul ignore next
function pathExists(path) {
    return fs.pathExists(path);
}
exports.pathExists = pathExists;
// istanbul ignore next
function move(src, dest, options) {
    return fs.move(src, dest, options !== null && options !== void 0 ? options : {});
}
exports.move = move;
// istanbul ignore next
function readdir(path) {
    return fs.readdir(path);
}
exports.readdir = readdir;
// istanbul ignore next
function rm(path, options) {
    return fs.rm(path, options);
}
exports.rm = rm;
