// renovate/dist/datasource/terraform-module/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformModuleDatasource = void 0;
const tslib_1 = require("tslib");
const logger_1 = require("../../logger");
const external_host_error_1 = require("../../types/errors/external-host-error");
const decorator_1 = require("../../util/cache/package/decorator");
const regex_1 = require("../../util/regex");
const hashicorpVersioning = (0, tslib_1.__importStar)(require("../../versioning/hashicorp"));
const base_1 = require("./base");
class TerraformModuleDatasource extends base_1.TerraformDatasource {
    constructor() {
        super(TerraformModuleDatasource.id);
        this.defaultRegistryUrls = ['https://registry.terraform.io'];
        this.defaultVersioning = hashicorpVersioning.id;
    }
    /**
     * This function will fetch a package from the specified Terraform registry and return all semver versions.
     *  - `sourceUrl` is supported of "source" field is set
     *  - `homepage` is set to the Terraform registry's page if it's on the official main registry
     */
    async getReleases({ lookupName, registryUrl, }) {
        const { registry, repository } = TerraformModuleDatasource.getRegistryRepository(lookupName, registryUrl);
        logger_1.logger.trace({ registry, terraformRepository: repository }, 'terraform-module.getReleases()');
        let res;
        let pkgUrl;
        try {
            const serviceDiscovery = await this.getTerraformServiceDiscoveryResult(registryUrl);
            pkgUrl = `${registry}${serviceDiscovery['modules.v1']}${repository}`;
            res = (await this.http.getJson(pkgUrl)).body;
            const returnedName = res.namespace + '/' + res.name + '/' + res.provider;
            if (returnedName !== repository) {
                logger_1.logger.warn({ pkgUrl }, 'Terraform registry result mismatch');
                return null;
            }
        }
        catch (err) {
            this.handleGenericErrors(err);
        }
        // Simplify response before caching and returning
        const dep = {
            releases: null,
        };
        if (res.source) {
            dep.sourceUrl = res.source;
        }
        dep.releases = res.versions.map((version) => ({
            version,
        }));
        if (pkgUrl.startsWith('https://registry.terraform.io/')) {
            dep.homepage = `https://registry.terraform.io/modules/${repository}`;
        }
        // set published date for latest release
        const latestVersion = dep.releases.find((release) => res.version === release.version);
        if (latestVersion) {
            latestVersion.releaseTimestamp = res.published_at;
        }
        logger_1.logger.trace({ dep }, 'dep');
        return dep;
    }
    handleSpecificErrors(err) {
        const failureCodes = ['EAI_AGAIN'];
        // istanbul ignore if
        if (failureCodes.includes(err.code)) {
            throw new external_host_error_1.ExternalHostError(err);
        }
    }
    static getRegistryRepository(lookupName, registryUrl) {
        let registry;
        const split = lookupName.split('/');
        if (split.length > 3 && split[0].includes('.')) {
            [registry] = split;
            split.shift();
        }
        else {
            registry = registryUrl;
        }
        if (!(0, regex_1.regEx)(/^https?:\/\//).test(registry)) {
            registry = `https://${registry}`;
        }
        const repository = split.join('/');
        return {
            registry,
            repository,
        };
    }
    static getCacheKey({ lookupName, registryUrl, }) {
        const { registry, repository } = TerraformModuleDatasource.getRegistryRepository(lookupName, registryUrl);
        return `${registry}/${repository}`;
    }
}
TerraformModuleDatasource.id = 'terraform-module';
(0, tslib_1.__decorate)([
    (0, decorator_1.cache)({
        namespace: `datasource-${TerraformModuleDatasource.id}`,
        key: (getReleasesConfig) => TerraformModuleDatasource.getCacheKey(getReleasesConfig),
    })
], TerraformModuleDatasource.prototype, "getReleases", null);
exports.TerraformModuleDatasource = TerraformModuleDatasource;
