// renovate/dist/manager/npm/post-update/npm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateLockFile = void 0;
const semver_1 = require("semver");
const shlex_1 = require("shlex");
const upath_1 = require("upath");
const global_1 = require("../../../config/global");
const error_messages_1 = require("../../../constants/error-messages");
const logger_1 = require("../../../logger");
const exec_1 = require("../../../util/exec");
const fs_1 = require("../../../util/fs");
const node_version_1 = require("./node-version");
async function generateLockFile(cwd, env, filename, config = {}, upgrades = []) {
    var _a, _b, _c;
    logger_1.logger.debug(`Spawning npm install to create ${cwd}/${filename}`);
    const { skipInstalls, postUpdateOptions } = config;
    let lockFile = null;
    try {
        let installNpm = 'npm i -g npm';
        const npmCompatibility = (_a = config.constraints) === null || _a === void 0 ? void 0 : _a.npm;
        // istanbul ignore else
        if (npmCompatibility) {
            // istanbul ignore else
            if ((0, semver_1.validRange)(npmCompatibility)) {
                installNpm = `npm i -g ${(0, shlex_1.quote)(`npm@${npmCompatibility}`)} || true`;
            }
            else {
                logger_1.logger.debug({ npmCompatibility }, 'npm compatibility range is not valid - skipping');
            }
        }
        else {
            logger_1.logger.debug('No npm compatibility range found - installing npm latest');
        }
        const preCommands = [installNpm, 'hash -d npm'];
        const commands = [];
        let cmdOptions = '';
        if ((postUpdateOptions === null || postUpdateOptions === void 0 ? void 0 : postUpdateOptions.includes('npmDedupe')) || skipInstalls === false) {
            logger_1.logger.debug('Performing node_modules install');
            cmdOptions += '--no-audit';
        }
        else {
            logger_1.logger.debug('Updating lock file only');
            cmdOptions += '--package-lock-only --no-audit';
        }
        if (!global_1.GlobalConfig.get('allowScripts') || config.ignoreScripts) {
            cmdOptions += ' --ignore-scripts';
        }
        const tagConstraint = await (0, node_version_1.getNodeConstraint)(config);
        const execOptions = {
            cwd,
            extraEnv: {
                NPM_CONFIG_CACHE: env.NPM_CONFIG_CACHE,
                npm_config_store: env.npm_config_store,
            },
            docker: {
                image: 'node',
                tagScheme: 'node',
                tagConstraint,
                preCommands,
            },
        };
        // istanbul ignore if
        if (global_1.GlobalConfig.get('exposeAllEnv')) {
            execOptions.extraEnv.NPM_AUTH = env.NPM_AUTH;
            execOptions.extraEnv.NPM_EMAIL = env.NPM_EMAIL;
        }
        if (!upgrades.every((upgrade) => upgrade.isLockfileUpdate)) {
            // This command updates the lock file based on package.json
            commands.push(`npm install ${cmdOptions}`.trim());
        }
        // rangeStrategy = update-lockfile
        const lockUpdates = upgrades.filter((upgrade) => upgrade.isLockfileUpdate);
        if (lockUpdates.length) {
            logger_1.logger.debug('Performing lockfileUpdate (npm)');
            const updateCmd = `npm install ${cmdOptions}` +
                lockUpdates
                    .map((update) => ` ${update.depName}@${update.newVersion}`)
                    .join('');
            commands.push(updateCmd);
        }
        if (upgrades.some((upgrade) => upgrade.isRemediation)) {
            // We need to run twice to get the correct lock file
            commands.push(`npm install ${cmdOptions}`.trim());
        }
        // postUpdateOptions
        if ((_b = config.postUpdateOptions) === null || _b === void 0 ? void 0 : _b.includes('npmDedupe')) {
            logger_1.logger.debug('Performing npm dedupe');
            commands.push('npm dedupe');
        }
        // TODO: don't assume package-lock.json is in the same directory
        const lockFileName = (0, upath_1.join)(cwd, filename);
        if (upgrades.find((upgrade) => upgrade.isLockFileMaintenance)) {
            logger_1.logger.debug(`Removing ${lockFileName} first due to lock file maintenance upgrade`);
            try {
                await (0, fs_1.remove)(lockFileName);
            }
            catch (err) /* istanbul ignore next */ {
                logger_1.logger.debug({ err, lockFileName }, 'Error removing yarn.lock for lock file maintenance');
            }
        }
        // Run the commands
        await (0, exec_1.exec)(commands, execOptions);
        // massage to shrinkwrap if necessary
        if (filename === 'npm-shrinkwrap.json' &&
            (await (0, fs_1.pathExists)((0, upath_1.join)(cwd, 'package-lock.json')))) {
            await (0, fs_1.move)((0, upath_1.join)(cwd, 'package-lock.json'), (0, upath_1.join)(cwd, 'npm-shrinkwrap.json'));
        }
        // Read the result
        lockFile = await (0, fs_1.readFile)((0, upath_1.join)(cwd, filename), 'utf8');
    }
    catch (err) /* istanbul ignore next */ {
        if (err.message === error_messages_1.TEMPORARY_ERROR) {
            throw err;
        }
        logger_1.logger.debug({
            err,
            type: 'npm',
        }, 'lock file error');
        if ((_c = err.stderr) === null || _c === void 0 ? void 0 : _c.includes('ENOSPC: no space left on device')) {
            throw new Error(error_messages_1.SYSTEM_INSUFFICIENT_DISK_SPACE);
        }
        return { error: true, stderr: err.stderr };
    }
    return { lockFile };
}
exports.generateLockFile = generateLockFile;
