// renovate/dist/config/presets/gitea/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPreset = exports.getPresetFromEndpoint = exports.fetchJSONFile = exports.Endpoint = void 0;
const logger_1 = require("../../../logger");
const gitea_helper_1 = require("../../../platform/gitea/gitea-helper");
const external_host_error_1 = require("../../../types/errors/external-host-error");
const util_1 = require("../util");
exports.Endpoint = 'https://gitea.com/api/v1/';
async function fetchJSONFile(repo, fileName, endpoint, packageTag) {
    let res;
    try {
        res = await (0, gitea_helper_1.getRepoContents)(repo, fileName, packageTag, {
            baseUrl: endpoint,
        });
    }
    catch (err) {
        // istanbul ignore if: not testable with nock
        if (err instanceof external_host_error_1.ExternalHostError) {
            throw err;
        }
        logger_1.logger.debug({ statusCode: err.statusCode, repo, fileName }, `Failed to retrieve ${fileName} from repo`);
        throw new Error(util_1.PRESET_DEP_NOT_FOUND);
    }
    try {
        const content = Buffer.from(res.content, 'base64').toString();
        const parsed = JSON.parse(content);
        return parsed;
    }
    catch (err) {
        throw new Error(util_1.PRESET_INVALID_JSON);
    }
}
exports.fetchJSONFile = fetchJSONFile;
function getPresetFromEndpoint(pkgName, filePreset, presetPath, endpoint = exports.Endpoint, packageTag) {
    return (0, util_1.fetchPreset)({
        pkgName,
        filePreset,
        presetPath,
        endpoint,
        packageTag,
        fetch: fetchJSONFile,
    });
}
exports.getPresetFromEndpoint = getPresetFromEndpoint;
function getPreset({ packageName: pkgName, presetName = 'default', presetPath, packageTag = null, }) {
    return getPresetFromEndpoint(pkgName, presetName, presetPath, exports.Endpoint, packageTag);
}
exports.getPreset = getPreset;
