// renovate/dist/config/presets/gitlab/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPreset = exports.getPresetFromEndpoint = exports.fetchJSONFile = exports.Endpoint = void 0;
const tslib_1 = require("tslib");
const is_1 = (0, tslib_1.__importDefault)(require("@sindresorhus/is"));
const logger_1 = require("../../../logger");
const external_host_error_1 = require("../../../types/errors/external-host-error");
const gitlab_1 = require("../../../util/http/gitlab");
const util_1 = require("../util");
const gitlabApi = new gitlab_1.GitlabHttp();
exports.Endpoint = 'https://gitlab.com/api/v4/';
async function getDefaultBranchName(urlEncodedPkgName, endpoint) {
    const branchesUrl = `${endpoint}projects/${urlEncodedPkgName}/repository/branches`;
    const res = await gitlabApi.getJson(branchesUrl);
    const branches = res.body;
    let defaultBranchName = 'master';
    for (const branch of branches) {
        if (branch.default) {
            defaultBranchName = branch.name;
            break;
        }
    }
    return defaultBranchName;
}
async function fetchJSONFile(repo, fileName, endpoint, packageTag) {
    let url = endpoint;
    let ref = '';
    try {
        const urlEncodedRepo = encodeURIComponent(repo);
        const urlEncodedPkgName = encodeURIComponent(fileName);
        if (is_1.default.nonEmptyString(packageTag)) {
            ref = `?ref=${packageTag}`;
        }
        else {
            const defaultBranchName = await getDefaultBranchName(urlEncodedRepo, endpoint);
            ref = `?ref=${defaultBranchName}`;
        }
        url += `projects/${urlEncodedRepo}/repository/files/${urlEncodedPkgName}/raw${ref}`;
        logger_1.logger.trace({ url }, `Preset URL`);
        return (await gitlabApi.getJson(url)).body;
    }
    catch (err) {
        if (err instanceof external_host_error_1.ExternalHostError) {
            throw err;
        }
        logger_1.logger.debug({ statusCode: err.statusCode, url }, `Failed to retrieve ${fileName} from repo`);
        throw new Error(util_1.PRESET_DEP_NOT_FOUND);
    }
}
exports.fetchJSONFile = fetchJSONFile;
function getPresetFromEndpoint(pkgName, presetName, presetPath, endpoint = exports.Endpoint, packageTag) {
    return (0, util_1.fetchPreset)({
        pkgName,
        filePreset: presetName,
        presetPath,
        endpoint,
        packageTag,
        fetch: fetchJSONFile,
    });
}
exports.getPresetFromEndpoint = getPresetFromEndpoint;
function getPreset({ packageName: pkgName, presetPath, presetName = 'default', packageTag = null, }) {
    return getPresetFromEndpoint(pkgName, presetName, presetPath, exports.Endpoint, packageTag);
}
exports.getPreset = getPreset;
